{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this script extracts transcription start sites of promoters from Molly's dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##needs mysql installed\n",
    "#Need to be aware that fasta file sequence can be multiline or single line, databases provide it multilined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "from BCBio.GFF import GFFExaminer\n",
    "from BCBio import GFF\n",
    "import pprint\n",
    "import pymysql.cursors\n",
    "from pyfaidx import Fasta\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('Chr1',): 20692,\n",
      "            ('Chr2',): 11967,\n",
      "            ('Chr3',): 15930,\n",
      "            ('Chr4',): 12159,\n",
      "            ('Chr5',): 18540,\n",
      "            ('ChrC',): 235,\n",
      "            ('ChrM',): 183},\n",
      " 'gff_source': {('MegrawPEAT',): 79706},\n",
      " 'gff_source_type': {('MegrawPEAT', 'Peak'): 79706},\n",
      " 'gff_type': {('Peak',): 79706}}\n"
     ]
    }
   ],
   "source": [
    "#examine the gff3 file\n",
    "in_file = \"/home/witham/Documents/Pipeline/data/TSS_data/AnnotatedPEATPeaks.txt\"\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('Chr1',): 207497,\n",
      "            ('Chr2',): 125500,\n",
      "            ('Chr3',): 154731,\n",
      "            ('Chr4',): 121971,\n",
      "            ('Chr5',): 178800,\n",
      "            ('ChrC',): 620,\n",
      "            ('ChrM',): 771},\n",
      " 'gff_source': {('Araport11',): 789890},\n",
      " 'gff_source_type': {('Araport11', 'CDS'): 286355,\n",
      "                     ('Araport11', 'antisense_RNA'): 91,\n",
      "                     ('Araport11', 'antisense_lncRNA'): 1424,\n",
      "                     ('Araport11', 'exon'): 200542,\n",
      "                     ('Araport11', 'five_prime_UTR'): 46895,\n",
      "                     ('Araport11', 'gene'): 33341,\n",
      "                     ('Araport11', 'lnc_RNA'): 2455,\n",
      "                     ('Araport11', 'mRNA'): 52270,\n",
      "                     ('Araport11', 'miRNA'): 427,\n",
      "                     ('Araport11', 'miRNA_primary_transcript'): 325,\n",
      "                     ('Araport11', 'ncRNA'): 286,\n",
      "                     ('Araport11', 'protein'): 48359,\n",
      "                     ('Araport11', 'pseudogene'): 952,\n",
      "                     ('Araport11', 'pseudogenic_exon'): 2058,\n",
      "                     ('Araport11', 'pseudogenic_tRNA'): 27,\n",
      "                     ('Araport11', 'pseudogenic_transcript'): 1100,\n",
      "                     ('Araport11', 'rRNA'): 15,\n",
      "                     ('Araport11', 'snRNA'): 82,\n",
      "                     ('Araport11', 'snoRNA'): 287,\n",
      "                     ('Araport11', 'tRNA'): 689,\n",
      "                     ('Araport11', 'three_prime_UTR'): 41127,\n",
      "                     ('Araport11', 'transcript_region'): 726,\n",
      "                     ('Araport11', 'transposable_element'): 31189,\n",
      "                     ('Araport11', 'transposable_element_gene'): 3901,\n",
      "                     ('Araport11', 'transposon_fragment'): 34856,\n",
      "                     ('Araport11', 'uORF'): 111},\n",
      " 'gff_type': {('CDS',): 286355,\n",
      "              ('antisense_RNA',): 91,\n",
      "              ('antisense_lncRNA',): 1424,\n",
      "              ('exon',): 200542,\n",
      "              ('five_prime_UTR',): 46895,\n",
      "              ('gene',): 33341,\n",
      "              ('lnc_RNA',): 2455,\n",
      "              ('mRNA',): 52270,\n",
      "              ('miRNA',): 427,\n",
      "              ('miRNA_primary_transcript',): 325,\n",
      "              ('ncRNA',): 286,\n",
      "              ('protein',): 48359,\n",
      "              ('pseudogene',): 952,\n",
      "              ('pseudogenic_exon',): 2058,\n",
      "              ('pseudogenic_tRNA',): 27,\n",
      "              ('pseudogenic_transcript',): 1100,\n",
      "              ('rRNA',): 15,\n",
      "              ('snRNA',): 82,\n",
      "              ('snoRNA',): 287,\n",
      "              ('tRNA',): 689,\n",
      "              ('three_prime_UTR',): 41127,\n",
      "              ('transcript_region',): 726,\n",
      "              ('transposable_element',): 31189,\n",
      "              ('transposable_element_gene',): 3901,\n",
      "              ('transposon_fragment',): 34856,\n",
      "              ('uORF',): 111}}\n"
     ]
    }
   ],
   "source": [
    "#examine the gff3 file\n",
    "in_file = \"/home/witham/Documents/Pipeline/data/genomes/Araport11_GFF3_genes_transposons.201606.gff\"\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def download_chromosome_sizes(genome_assembly_id,output_file):\n",
    "#     \"\"\"download chromosome sizes from the UCSC Genome browser's mysql database, using the UCSC Genome Browser assembly ID\"\"\"\n",
    "#     #connect to the database\n",
    "#     connection = pymysql.connect(host='genome-euro-mysql.soe.ucsc.edu',\n",
    "#                                 user='genome',\n",
    "#                                 port=3306)\n",
    "#     output = open(output_file, 'w') #make output file with write capability\n",
    "#     try:\n",
    "#         with connection.cursor() as cursor:\n",
    "#             sql = f\"SELECT 'chrom', 'size' FROM {genome_assembly_id}.chromInfo\" #specify the columns and genome assembly\n",
    "#             cursor.execute(sql) #grab the columns and genome assembly\n",
    "#             result = cursor.fetchone()\n",
    "#             output.write(result)\n",
    "#     finally:\n",
    "#         connection.close()\n",
    "#     output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta_chromsizes(genome, output_file):\n",
    "    \"\"\"extracts chromosome sizes in a form compatible with BedTools.flank (as a dictionary)\"\"\"\n",
    "    #output = open(output_file, 'w')\n",
    "    the_genome = Fasta(genome) #using pyfaidx Fasta, parse the genome file as a fasta\n",
    "     \n",
    "    chromsizes = {} #make dictionary called chromsizes\n",
    "    for key in the_genome.keys():\n",
    "       \n",
    "        #chromsizes[key] = f'({0}, {len(the_genome[key])})' #add the chromosome name and length to dictionary\n",
    "        chromsizes[f'Chr{key}'] = f'({len(the_genome[key])})' #add the chromosome name and length to dictionary\n",
    "        #chromsizes[key] = f'({len(the_genome[key])})' #add the chromosome name and length to dictionary\n",
    "    #output_file.write(chrom)\n",
    "    chromsizes_string = ''\n",
    "    \n",
    "    for k,v in chromsizes.items():\n",
    "        chromsizes_string = chromsizes_string + f'{k}\\t{v}\\n'\n",
    "    \n",
    "    \n",
    "    with open(output_file, 'w') as output:            \n",
    "        output.write(chromsizes_string.replace('(','').replace(')',''))\n",
    "    #\n",
    "       # a,b = k.split(':'#turn dictionary string into columns\n",
    " \n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def extract_genes(gene_gff,output_file):\n",
    "#     \"\"\"This function extracts all whole genes from a gff3 file, ignoring gene features, and adds them to an output file\"\"\"\n",
    "#     stringToMatch = 'ID=gene'\n",
    "#     matchedLine = ''\n",
    "#     with open(gene_gff, 'r') as file:        \n",
    "#         for line in file:\n",
    "#             if stringToMatch in line:\n",
    "#                 matchedLine = matchedLine + line + '\\n'\n",
    "    \n",
    "    \n",
    "#     with open(output_file, 'w') as output:\n",
    "#         output.write(matchedLine)       \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_genes(gene_gff,temp,output_file):\n",
    "    \"\"\"This function extracts all whole genes from a gff3 file, ignoring gene features, and adds them to an output file\"\"\"\n",
    "    #limit dictionary to genes\n",
    "    limit_info = dict(gff_type = ['gene'])\n",
    "    #open temporary file\n",
    "    output = open(temp, 'w')\n",
    "    #open gff file, parse it, limiting to genes only. Save the file.\n",
    "    with open(gene_gff, 'r') as in_handle:                    \n",
    "            GFF.write(GFF.parse(in_handle, limit_info=limit_info),output)\n",
    "    output.close()\n",
    "    \n",
    "    #now remove the unwanted annotations that were added by GFF.write eg. Chr1\tannotation\tremark\t1\t30425192\t.\t.\t.\tgff-version=3\n",
    "    #remove lines beginning with ##\n",
    "    with open(temp, 'r') as tempfile, open(output_file, 'w') as newfile:        \n",
    "        for line in tempfile:\n",
    "            line = line.strip() # removes hidden characters/spaces\n",
    "            if line[0] == \"#\":\n",
    "                pass\n",
    "            else:                \n",
    "                #don't include lines beginning with ##\n",
    "                newfile.write(line + '\\n') #output to new file\n",
    "    #remove temporary file\n",
    "    os.remove(temp)       \n",
    "    #read in gff file\n",
    "    genes = pd.read_table(output_file, sep='\\t', header=0)\n",
    "    cols2 = ['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']\n",
    "    genes.columns = cols2\n",
    "    #remove all lines where source is annotation\n",
    "    no_annotation = genes[~genes.source.str.contains(\"annotation\")]\n",
    "    #remove all lines that are not protein coding\n",
    "    protein_coding = no_annotation[no_annotation.attributes.str.contains('locus_type=protein_coding')]\n",
    "    #create gff file containing no lines with annotation and no lines starting with ##\n",
    "    protein_coding.to_csv(output_file,index=False,sep='\\t',header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_promoter(genes_gff,chromsize,promoter_length,output_file):\n",
    "#     \"\"\"This function adds a promoter of a certain length to each gene in the input file and exports to an output file\"\"\"\n",
    "#     #output = open(output_location, 'w') #make output file with write capability\n",
    "#     genes = BedTool(genes_gff)\n",
    "#     promoters = genes.flank(g=chromsize, l=promoter_length, r=0, s=True)\n",
    "      \n",
    "\n",
    "#     #promoters = a.flank(genome=chromsize, l=1000)\n",
    "#     #open gff file as dataframe and separate by strand, and add column names\n",
    "#   #  df = pd.read_csv(genes_gff, sep='\\t', encoding='utf-8',\n",
    "#     #names=[\"chromosome\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"])\n",
    "        \n",
    "#     #positive = BedTool.from_dataframe(df[df.strand.isin(['+'])]) #add only positive strand genes to bedtools\n",
    "#     #pos_promoters = positive.flank(g=chromsize, l=promoter_length, r=0, s=True) #add promoter upstream of gene\n",
    "    \n",
    "#     #make negative signs positive\n",
    "#   #  negative = df[df.strand.isin(['-'])].copy()\n",
    "#     #negpos = negative.strand.replace('-','+')\n",
    "#    # negative2 = BedTool.from_dataframe(negpos) #add only negative strand genes to bedtools\n",
    "#    # neg_promoters = negative2.flank(g=chromsize, l=0, r=promoter_length) #add promoter after gene as it is on -ve strand\n",
    "    \n",
    "#     #merge the positive and negative promoters, with postmerge option so that promoters themselves aren't merged\n",
    "#     #promoters = pos_promoters.cat(neg_promoters, postmerge=False)     \n",
    "         \n",
    "#     with open(output_file,'w') as f:\n",
    "#         f.write(str(promoters))\n",
    "# #     count = 0\n",
    "# #     while count < 10:\n",
    "# #         for line in promoters:\n",
    "# #             print(line)\n",
    "# #             count = count + 1\n",
    "#   #  return negpos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_promoter(genes_gff,chromsize,promoter_length):\n",
    "    \"\"\"This function adds a promoter of a certain length to each gene in the input file and exports an output string\"\"\"\n",
    "    #output = open(output_location, 'w') #make output file with write capability\n",
    "    #parse gff file containing only genes.\n",
    "    genes = BedTool(genes_gff)\n",
    "    #extract promoters upsteam using chromsize file and specified promoter length. r, no. of bp to add to end coordinate. s, based on strand.\n",
    "    promoters = genes.flank(g=chromsize, l=promoter_length, r=0, s=True)\n",
    "    return promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def promoter_overlap(promoter_gff,allfeatures_gff,output_file):\n",
    "#     \"\"\"function to create file containing promoters which overlap other genome features\"\"\"\n",
    "#     promoters = BedTool(promoter_gff) #read in files using BedTools\n",
    "#     features = BedTool(allfeatures_gff)\n",
    "#     #report chromosome position of overlapping feature, along with the promoter which overlaps it (only reports the overlapping nucleotides, not the whole promoter length. Can use u=True to get whole promoter length)\n",
    "#     #f, the minimum overlap as fraction of A. F, nucleotide fraction of B (genes) that need to be overlapping with A (promoters)\n",
    "#     #wa, Write the original entry in A for each overlap.\n",
    "#     #u, write original A entry only once even if more than one overlap\n",
    "#     intersect = promoters.intersect(features,f=0.001, F=0.001, u=True, wa=True) #could add u=True which indicates we want to see the promoters that overlap features in the genome\n",
    "#     with open(output_file, 'w') as output:\n",
    "#         output.write(str(intersect))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_promoter_overlap(promoter_gff, all_genes_gff, output_file):\n",
    "    \"\"\"function to create file containing promoters which overlap other genome features. \n",
    "    Then create a df with only promoters which overlap other genes. Then shorten them so they are no longer overlapping. \n",
    "    Then merge back with all the extracted promoters, keeping only the shortened ones\"\"\"\n",
    "    all_proms = BedTool(promoter_gff) #read in files using BedTools\n",
    "    features = BedTool(all_genes_gff)\n",
    "    #report chromosome position of overlapping feature, along with the promoter which overlaps it (only reports the overlapping nucleotides, not the whole promoter length. Can use u=True to get whole promoter length)\n",
    "    #f, the minimum overlap as fraction of A. F, nucleotide fraction of B (genes) that need to be overlapping with A (promoters)\n",
    "    #wa, Write the original entry in A for each overlap.\n",
    "    #wo,  Write the original A and B entries plus the number of base pairs of overlap between the two features. Only A features with overlap are reported. \n",
    "    #u, write original A entry only once even if more than one overlap\n",
    "    intersect = all_proms.intersect(features, wo=True) #could add u=True which indicates we want to see the promoters that overlap features in the genome\n",
    "    #Write to output_file\n",
    "    with open(output_file, 'w') as output:\n",
    "        #Each line in the file contains gff entry a and gff entry b that it overlaps plus the number of bp in the overlap so 19 columns\n",
    "        output.write(str(intersect))    \n",
    "    #read in gff file\n",
    "    overlapping_proms = pd.read_table(output_file, sep='\\t', header=0)\n",
    "    cols = ['chrA', 'sourceA', 'typeA', 'startA','stopA','dot1A','strandA','dot2A','attributesA','chrB', 'sourceB', 'typeB', 'startB','stopB','dot1B','strandB','dot2B','attributesB','bp_overlap']\n",
    "    overlapping_proms.columns = cols\n",
    "    #make new columns for the new start and stop for feature A\n",
    "    overlapping_proms['new_startA'] = overlapping_proms.startA\n",
    "    overlapping_proms['new_stopA'] = overlapping_proms.stopA\n",
    "\n",
    "    #iterate over rows\n",
    "    for i,data in overlapping_proms.iterrows():\n",
    "        #if positive strand feature A, reduce length to the stop position of feature B + 1\n",
    "        if overlapping_proms.loc[i,'strandA'] == '+':\n",
    "            \n",
    "            #if overlapping feature is downstream of the promoter or equal to promoter end, reduce both start and stop to the stop of the feature + 1\n",
    "            if overlapping_proms.loc[i,'stopA'] <=  overlapping_proms.loc[i, 'stopB']:\n",
    "                overlapping_proms.loc[i, 'new_startA'] = overlapping_proms.loc[i, 'stopB'] + 1\n",
    "                overlapping_proms.loc[i, 'new_stopA'] = overlapping_proms.loc[i, 'stopB'] + 1\n",
    "                #overlapping_proms.drop(i, inplace=True)\n",
    "                \n",
    "            #else  reduce length of promoter to stop position of feature +1\n",
    "            else:\n",
    "                \n",
    "                overlapping_proms.loc[i, 'new_startA'] = overlapping_proms.loc[i, 'stopB'] + 1\n",
    "          \n",
    "                \n",
    "        #if negative strand feature A, reduce length of promoter to start position of feature -1\n",
    "        elif overlapping_proms.loc[i,'strandA'] == '-':\n",
    "            #if overlapping feature is downstream of the promoter, change promtoer start and stop to the start position of the feature -1\n",
    "            if overlapping_proms.loc[i,'startA'] >=  overlapping_proms.loc[i, 'startB']:\n",
    "                overlapping_proms.loc[i, 'new_stopA'] = overlapping_proms.loc[i, 'startB'] - 1\n",
    "                overlapping_proms.loc[i, 'new_startA'] = overlapping_proms.loc[i, 'startB'] - 1\n",
    "                \n",
    "                #overlapping_proms.drop(i, inplace=True)\n",
    "            else:           \n",
    "                overlapping_proms.loc[i, 'new_stopA'] = overlapping_proms.loc[i, 'startB'] - 1\n",
    "\n",
    "    #create df with just feature A containing new start and stop locations\n",
    "    new_feature_A = overlapping_proms[['chrA', 'sourceA', 'typeA', 'new_startA', 'new_stopA', 'dot1A', 'strandA', 'dot2A', 'attributesA']]\n",
    "   \n",
    "    #rename columns\n",
    "    cols2 = ['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']\n",
    "    new_feature_A.columns = cols2\n",
    "    #create a buffer of promoter_gff\n",
    "    promoter_gff_buffer = io.StringIO()\n",
    "    promoter_gff_buffer.write(str(promoter_gff))\n",
    "    #go back to beginning of the buffer\n",
    "    promoter_gff_buffer.seek(0)\n",
    "    #read in all_proms as df\n",
    "    all_proms_df = pd.read_table(promoter_gff_buffer, sep='\\t', header=0)\n",
    "    all_proms_df.columns = cols2\n",
    "    #remove overlapping promoters from all_proms_df\n",
    "    #first add AGI columns\n",
    "    #add AGI column to promoters\n",
    "    all_proms_df_agi = all_proms_df.assign(AGI=all_proms_df.attributes.str.extract(r'ID=(.*?)\\;'))\n",
    "    new_feature_A_agi = new_feature_A.assign(AGI=new_feature_A.attributes.str.extract(r'ID=(.*?)\\;'))\n",
    "    overlapping_proms_agi = overlapping_proms.assign(AGI=overlapping_proms.attributesA.str.extract(r'ID=(.*?)\\;'))\n",
    "    #remove any promoters in all_prom_df that overlap other genes\n",
    "    all_prom_removed_overlaps = all_proms_df_agi[~all_proms_df_agi.AGI.isin(overlapping_proms_agi.AGI)]\n",
    "       \n",
    "    #merge all_prom_removed_overlaps with new_feature_A_agi\n",
    "    #outer merge adding suffix to new_feature_A\n",
    "    merged = pd.merge(all_prom_removed_overlaps, new_feature_A_agi, how='outer')\n",
    "    #remove unwanted columns\n",
    "    merged = merged[cols2]\n",
    "\n",
    "    return merged\n",
    "#     #if start_reduced column from new_feature_A is NaN, pass\n",
    "#     #replace NaN in start_reduced with string 'NAN'\n",
    "#     merged.fillna('NAN', inplace=True)\n",
    "#     for index,row in merged.iterrows():\n",
    "#         #if NAN is in start_reduced column, pass\n",
    "#         if merged.loc[index, 'start_reduced'] == 'NAN':\n",
    "#         #merged.isnull(row.start_reduced):\n",
    "#             pass\n",
    "#         #if not NAN, replace start and stop from all_proms_df with those from new_feature_A\n",
    "#         else:\n",
    "#             merged.loc[index, 'start'] = merged.loc[index, 'start_reduced']\n",
    "#             merged.loc[index, 'stop'] = merged.loc[index, 'stop_reduced']\n",
    "    \n",
    "\n",
    "# #         merged.loc[merged.start] == merged.loc[merged.start_reduced]\n",
    "# #         merged.loc[merged.stop] == merged.loc[merged.stop_reduced]\n",
    "   \n",
    "#     #merged.columns = cols2\n",
    "#     promoter_gff_buffer.close()\n",
    "#     new_merged = merged[cols2]\n",
    "#     #remove trailing decimal .0 from start and stop\n",
    "#     new_merged = new_merged.astype({'start': 'int'})\n",
    "#     new_merged = new_merged.astype({'stop': 'int'})\n",
    "#     #remove promoters which have a start equal to or higher than the stop\n",
    "#     #new_merged = new_feature_A[~(new_feature_A.new_startA >= new_feature_A.new_stopA)]\n",
    "#     return new_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_5UTR(promoter_gff, all_features_gff):\n",
    "    \"\"\"Function to extend the promoters to include the 5'UTR region until the start codon of the first CDS feature of the same gene.\n",
    "    Also remove mitochondira and chloroplast features.\"\"\"\n",
    "    promoters = pd.read_table(promoter_gff, sep='\\t', header=0)\n",
    "    \n",
    "    cols = ['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']\n",
    "    promoters.columns = cols\n",
    "    #add AGI column to promoters\n",
    "    promoters = promoters.assign(AGI=promoters.attributes.str.extract(r'ID=(.*?)\\;'))\n",
    "    \n",
    "    #remove lines beginning with ##\n",
    "    #create a buffer for feature output\n",
    "    feature_buffer = io.StringIO()\n",
    "    with open(all_features_gff, 'r') as f:        \n",
    "        for line in f:\n",
    "            line = line.strip() # removes hidden characters/spaces\n",
    "            if line[0] == \"#\":\n",
    "                pass\n",
    "            else:                \n",
    "                #don't include lines beginning with ##\n",
    "                feature_buffer.write(line + '\\n') #output to buffer\n",
    "    #go back to beginning of the buffer\n",
    "    feature_buffer.seek(0)\n",
    "\n",
    "    #read in feature buffer to df\n",
    "    features = pd.read_table(feature_buffer, sep='\\t', header=0)\n",
    "    \n",
    "    features.columns = cols\n",
    "    #filter features to contain only cds\n",
    "    cds = features[features.type == 'CDS']\n",
    "    #filter to only contain cds 1 in parent .1\n",
    "    #remove mitochondria and chloroplast features\n",
    "    cds_1 = cds[cds['chr'].isin(['Chr1', 'Chr2', 'Chr3', 'Chr4', 'Chr5'])]\n",
    "    #remove all features whose parent isnt Parent=AGI.1\n",
    "    #cds_1 = cds_1[cds_1.attributes.str.contains('Parent=[A-Z]{2}\\d{1}[A-Z]{1}\\d{5}.1')]\n",
    "    # keep only CDS:1;\n",
    "    cds_1 = cds_1[cds_1.attributes.str.contains(':CDS:1;')]\n",
    "\n",
    "    #add AGI column to CDSs\n",
    "    cds2 = cds_1.assign(AGI=cds_1.attributes.str.extract(r'ID=(.*?)\\:'))\n",
    "    #remove mitochondria and chloroplast features from promoters\n",
    "    promoters = promoters[promoters['chr'].isin(['Chr1', 'Chr2', 'Chr3', 'Chr4', 'Chr5'])]\n",
    "\n",
    "    #merge on AGI\n",
    "    merged = pd.merge(promoters, cds2, on='AGI', how='left',suffixes=('','_cds'))\n",
    "\n",
    "    for i, v in merged.iterrows():\n",
    "        #if positive strand feature A, reduce length to the stop position of feature B + 1\n",
    "        if merged.loc[i,'strand'] == '+':\n",
    "            #if overlapping feature is upstream or equal to start of the promoter, drop row\n",
    "            if merged.loc[i,'start'] >=  merged.loc[i, 'start_cds']:\n",
    "                merged.drop(i, inplace=True)\n",
    "            else:                    \n",
    "                #extend promoter to include 5'UTR region up until the start of the cds\n",
    "                merged.loc[i, 'stop'] = merged.loc[i, 'start_cds'] - 1\n",
    "        elif merged.loc[i,'strand'] == '-':            \n",
    "            #if overlapping feature is upstream or equal to start of the promoter, drop row\n",
    "            if merged.loc[i,'stop'] <=  merged.loc[i, 'stop_cds']:\n",
    "                merged.drop(i, inplace=True)\n",
    "            else:            \n",
    "                #extend promoter to include 5'UTR region up until the start of the cds\n",
    "                merged.loc[i, 'start'] = merged.loc[i, 'stop_cds'] + 1     \n",
    "            \n",
    "    #remove unwanted columns\n",
    "    merged_new = merged[['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']]\n",
    "    return merged_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_bad_proms(proms):\n",
    "    \"\"\"function to remove promoters whose stop is higher or equal to the start. \n",
    "    Then calculate lengths and sort by length. Drop duplicates keeping the shortest promoter.\"\"\"\n",
    "    #Remove promoters with stop higher than start\n",
    "    for i, v in proms.iterrows():\n",
    "        if proms.loc[i,'start'] >=  proms.loc[i, 'stop']:\n",
    "                proms.drop(i, inplace=True)          \n",
    "    #sort based on length then keep the shortest, and remove lengths that are 0\n",
    "    #Make promoter length column\n",
    "    proms_length =  proms.assign(length=(proms.start - proms.stop).abs())\n",
    "    #remove promoters with length 0\n",
    "    removed= proms_length.loc[proms_length.length != 0]\n",
    "    #remove duplicates keeping the shortest promoter\n",
    "    no_dups = removed.sort_values('length',ascending=True).drop_duplicates('attributes',keep='first')\n",
    "    proms_dropped = no_dups[['chr', 'source', 'type', 'start', 'stop', 'dot1', 'strand', 'dot2', 'attributes']]\n",
    "    #sort by chr then start\n",
    "    proms_dropped = proms_dropped.sort_values(['chr','start']).reset_index(drop=True)\n",
    "    return proms_dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>12941</td>\n",
       "      <td>14714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:11118137,PMID:15010618,PMID:166036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>22121</td>\n",
       "      <td>23518</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>32671</td>\n",
       "      <td>34171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>37062</td>\n",
       "      <td>38443</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=LHY1,LATE ELONGATED HYPOCOTYL 1;Dbxref=P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>40878</td>\n",
       "      <td>42017</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15703057,locus:2200990;ID=AT1G0107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27416</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26960227</td>\n",
       "      <td>26960903</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12805598,PMID:18055584,PMID:216589...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27417</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26963615</td>\n",
       "      <td>26964550</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2158641;ID=AT5G67610;Name=AT5G676...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27418</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26965721</td>\n",
       "      <td>26967010</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12912986,PMID:15469496,PMID:200320...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27419</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26969307</td>\n",
       "      <td>26969515</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12062092,PMID:17675552,PMID:186504...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27420</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26970549</td>\n",
       "      <td>26971668</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2158661;ID=AT5G67640;Name=AT5G676...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27421 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1     2         3         4  5  6  7  \\\n",
       "0      Chr1  Araport11  gene     12941     14714  .  -  .   \n",
       "1      Chr1  Araport11  gene     22121     23518  .  +  .   \n",
       "2      Chr1  Araport11  gene     32671     34171  .  -  .   \n",
       "3      Chr1  Araport11  gene     37062     38443  .  -  .   \n",
       "4      Chr1  Araport11  gene     40878     42017  .  -  .   \n",
       "...     ...        ...   ...       ...       ... .. .. ..   \n",
       "27416  Chr5  Araport11  gene  26960227  26960903  .  -  .   \n",
       "27417  Chr5  Araport11  gene  26963615  26964550  .  -  .   \n",
       "27418  Chr5  Araport11  gene  26965721  26967010  .  -  .   \n",
       "27419  Chr5  Araport11  gene  26969307  26969515  .  -  .   \n",
       "27420  Chr5  Araport11  gene  26970549  26971668  .  -  .   \n",
       "\n",
       "                                                       8  \n",
       "0      Dbxref=PMID:11118137,PMID:15010618,PMID:166036...  \n",
       "1      Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...  \n",
       "2      Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...  \n",
       "3      Alias=LHY1,LATE ELONGATED HYPOCOTYL 1;Dbxref=P...  \n",
       "4      Dbxref=PMID:15703057,locus:2200990;ID=AT1G0107...  \n",
       "...                                                  ...  \n",
       "27416  Dbxref=PMID:12805598,PMID:18055584,PMID:216589...  \n",
       "27417  Dbxref=locus:2158641;ID=AT5G67610;Name=AT5G676...  \n",
       "27418  Dbxref=PMID:12912986,PMID:15469496,PMID:200320...  \n",
       "27419  Dbxref=PMID:12062092,PMID:17675552,PMID:186504...  \n",
       "27420  Dbxref=locus:2158661;ID=AT5G67640;Name=AT5G676...  \n",
       "\n",
       "[27421 rows x 9 columns]"
      ]
     },
     "execution_count": 456,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#will come back to this later once pipline is finished\n",
    "# def rename_promoter_feature(promoter_gff, output_file):\n",
    "#     \"\"\"This function renames the 3rd column of the promoter gff file to promoter\"\"\"\n",
    "#     output =  open(output_file, 'w')\n",
    "#     with open(promoter_gff, 'r') as in_handle:\n",
    "#         GFF.write(GFF.parse(in_handle, ??????????????????),output)\n",
    "#     output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def find_closest_TSS(gene_gff,TSS_gff,output_location):\n",
    "#     \"\"\"this reads in the genes gff file and TSS gff file, finds the closest gene each TSS belongs to\"\"\"\n",
    "#     output = open(output_location, 'w') #make output file with write capability\n",
    "#     genes = BedTool(gene_gff)\n",
    "#     TSS = BedTool(TSS_gff)\n",
    "#     non_overlapping_TSS = TSS.subtract(genes)\n",
    "#     #intersect = TSS.intersect(genes)\n",
    "    \n",
    "#    # for gene in intersect:\n",
    "#        # print(gene.name)\n",
    "   \n",
    "#     nearby = genes.closest(TSS, d=True, stream=True)   \n",
    "  \n",
    "#     for gene in nearby:             \n",
    "#         if int(gene[-1]) < 1000:\n",
    "#             output.write(gene.name + '\\n')      \n",
    "#     output.close()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_closest_TSS(gene_gff,TSS_gff,output_location):\n",
    "    \"\"\"this reads in the genes gff file and TSS gff file, finds the closest gene each TSS belongs to\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gff columns:\n",
    "   # 1:seqname 2:source 3:feature 4:start 5:end 6:score 7:filter 8:strand 9:group 10:attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_characters_linestart(input_location,output_location,oldcharacters,newcharacters):\n",
    "    \"\"\"this function removes characters from the start of each line in the input file and sends modified lines to output\"\"\"\n",
    "    output = open(output_location, 'w') #make output file with write capability\n",
    "    with open(input_location, 'r') as infile:  \n",
    "        \n",
    "        for line in infile:\n",
    "            line = line.strip() # removes hidden characters/spaces\n",
    "            if line[0] == \"C\":\n",
    "                                 \n",
    "                line = line.replace(oldcharacters, newcharacters) #remove characters from start of line, replace with new characters        \n",
    "                output.write(line + '\\n') #output to new file\n",
    "    output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_empty_lines(input_file):\n",
    "    \"\"\"function to strip empty line from end of file\"\"\"\n",
    "    #make temporary buffer file\n",
    "    temp_buffer = io.StringIO()\n",
    "    with open(input_file) as f_input:\n",
    "        data = f_input.read().rstrip('\\n')\n",
    "    temp_buffer.write(data)\n",
    "    temp_buffer.seek(0)\n",
    "    data2 = temp_buffer.read().rstrip('\\n')\n",
    "\n",
    "    with open(input_file, 'w') as f_output:    \n",
    "        f_output.write(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bidirectional_proms(in_file, out_file):\n",
    "    \"\"\"this function create a file containing all promoters with an upstream gene going in the other direction ie. potential bidirectional promoters\"\"\"\n",
    "    #read in gff file\n",
    "    promoters = pd.read_table(in_file, sep='\\t', header=0)\n",
    "    cols2 = ['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']\n",
    "    promoters.columns = cols2\n",
    "    #make sure lines are sorted\n",
    "  \n",
    "    promoters = promoters.sort_values(['chr','start']).reset_index(drop=True)\n",
    "    #if bidirectional\n",
    "    promoters['bidirectional'] = 'no'\n",
    "#     promoters['shift_f_start']\n",
    "#     promoters['shift_f_stop']\n",
    "    #print(promoters)\n",
    "    \n",
    "    for i,data in promoters.iterrows():\n",
    "        if i-1 >= 0:\n",
    "            #print(i,data)\n",
    "            if promoters.loc[i, 'strand'] == '+' and promoters.loc[i-1, 'strand'] == '-' and promoters.loc[i, 'start'] - promoters.loc[i-1, 'stop'] < 2000:\n",
    "                promoters.loc[i, 'bidirectional'] = 'yes'\n",
    "                promoters.loc[i-1, 'bidirectional'] = 'yes'\n",
    "    #write file\n",
    "    with open(out_file, 'w') as output:  \n",
    "        promoters[promoters.bidirectional == 'no'][['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']].to_csv(out_file,index=False,sep='\\t',header=0)\n",
    "       \n",
    "\n",
    "    \n",
    "#     for i,data in promoters.iterrows():\n",
    "#         print(i, data)\n",
    "#         if i-1 > 0:\n",
    "#             if promoters.loc[i, 'strand'] == '+' and promoters.loc[i-1, 'strand'] == '-' and promoters.loc[i, 'start'] - promoters.loc[i-1, 'stop'] < 2000:\n",
    "#                 promoters.loc[i, 'bidirectional'] = 'yes'\n",
    "#                 promoters.loc[i-1, 'bidirectional'] = 'yes'\n",
    "\n",
    "    #print(len(promoters.bidirectional[promoters.bidirectional == 'yes']))     \n",
    "    #print(promoters)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TSS_raw = '/home/witham/Documents/Pipeline/data/TSS_data/AnnotatedPEATPeaks.txt'\n",
    "#TSS_renamedChr = '/home/witham/Documents/Pipeline/data/TSS_data/AnnotatedPEATPeaks_renamedChr.gff'\n",
    "#genome2 = \"/home/witham/Documents/Pipeline/data/genomes/Arabidopsis_thaliana.TAIR10.dna.toplevel_sl_simpleChr.fasta\"\n",
    "genome = '/home/witham/Documents/Pipeline/data/genomes/TAIR10_chr_all.fas'\n",
    "nonbidirectional_promoters = '../../data/genomes/nonbidirectional_proms.gff3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = \"../../data/genomes/Araport11_GFF3_genes_transposons.201606.gff\"\n",
    "#genes_renamedChr = \"/home/witham/Documents/Pipeline/data/genomes/Araport11_GFF3_genes_transposons.201606_renamedChr.gff\"\n",
    "test_genes = \"/home/witham/Documents/Pipeline/data/genomes/test_genes.gff3\"\n",
    "testgenesonly_gff = \"/home/witham/Documents/Pipeline/data/genomes/testgenes_only.gff3\"\n",
    "temp = \"/home/witham/Documents/Pipeline/data/TSS_data/temp.txt\"\n",
    "#TSS = \"/home/witham/Documents/Pipeline/data/TSS_data/AnnotatedPEATPeaks_renamedcol.gff\"\n",
    "#TSS = \"/home/witham/Documents/Pipeline/data/TSS_data/TSStest.txt\"\n",
    "#find_closest_TSS(genes,output,temp)\n",
    "genesonly_gff = \"../../data/genomes/genesonly.gff3\"\n",
    "promoters = \"../../data/genomes/promoters.gff3\"\n",
    "promoters_5UTR = '../../data/genomes/promoters_5UTR.gff3'\n",
    "#genes_bed = \"/home/witham/Documents/Pipeline/data/genomes/genes.bed\"\n",
    "overlapping_promoters = \"../../data/genomes/promoters_overlapping.gff3\"\n",
    "promoterandgenes_only_overlap = \"../../data/genomes/promoterandgenes_only_overlap.gff3\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove 'Chr' from seqname column in TSS gff file so that it matches the naming in the gene gff file\n",
    "#also remove Chr from araport gff file\n",
    "#remove_characters_linestart(TSS_raw, TSS_renamedChr, 'Chr')\n",
    "#remove_characters_linestart(genes, genes_renamedChr, 'Chr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#num = len([1 for line in open(\"/home/witham/Documents/Pipeline/data/genomes/Arabidopsis_thaliana.TAIR10.dna.toplevel_sl_simpleChr.fasta\") if line.startswith(\">\")])\n",
    "#print(num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromsizes_file = '/home/witham/Documents/Pipeline/data/genomes/chromsizes.chr'\n",
    "chromsizes_file_renamedChr_temp = '/home/witham/Documents/Pipeline/data/genomes/chromsizes_renamedChr_temp.chr'\n",
    "chromsizes_file_renamedChr = '/home/witham/Documents/Pipeline/data/genomes/chromsizes_renamedChr.chr'\n",
    "#chromsizes_file2 = '/home/witham/Documents/Pipeline/data/genomes/chromsizes2.chr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_chromsizes(genome, chromsizes_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename mitochondria and chloroplast to M and C\n",
    "remove_characters_linestart(chromsizes_file, chromsizes_file_renamedChr_temp, 'mitochondria','M')\n",
    "remove_characters_linestart(chromsizes_file_renamedChr_temp, chromsizes_file_renamedChr, 'chloroplast','C')\n",
    "os.remove(chromsizes_file_renamedChr_temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fasta_chromsizes(genome2, chromsizes_file2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "extract_genes() missing 1 required positional argument: 'output_file'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-32d685699172>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#extract_genes(genes,genesonly_gff)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mextract_genes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenesonly_gff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;31m#this changes chromosome no. to 1 rather than Chr1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: extract_genes() missing 1 required positional argument: 'output_file'"
     ]
    }
   ],
   "source": [
    "#extract_genes(genes,genesonly_gff)\n",
    "#extract_genes(genes,genesonly_gff)\n",
    "#this changes chromosome no. to 1 rather than Chr1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove genes which are bidirectional (ie. upsteam promoter of another gene)\n",
    "#bidirectional_proms(genesonly_gff, nonbidirectional_promoters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add_promoter(nonbidirectional_promoters,chromsizes_file_renamedChr,1000,promoters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(genesonly_gff, sep='\\t', encoding='utf-8',\n",
    "# names=[\"chromosome\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"])\n",
    "        \n",
    "# positive = BedTool.from_dataframe(df[df.strand.isin(['+'])]) #add only positive strand genes to bedtools\n",
    "# pos_promoters = positive.flank(g=chromosome_sizes_dict, l=1000, r=0) #add promoter upstream of gene\n",
    "# with open(\"/home/witham/Documents/Pipeline/data/genomes/promoters.gff3\",'w') as f:\n",
    "#         f.write(str(pos_promoters))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find_closest_TSS(genesonly_gff,TSS_raw,'/home/witham/Documents/Pipeline/data/genomes/closest.test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create file containing only promoters which overlap other genome features\n",
    "#promoter_overlap(promoters,genes,overlapping_promoters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('Chr1',): 6785,\n",
      "            ('Chr2',): 4064,\n",
      "            ('Chr3',): 5115,\n",
      "            ('Chr4',): 3857,\n",
      "            ('Chr5',): 5714,\n",
      "            ('ChrC',): 122,\n",
      "            ('ChrM',): 87},\n",
      " 'gff_source': {('Araport11',): 25744},\n",
      " 'gff_source_type': {('Araport11', 'gene'): 25744},\n",
      " 'gff_type': {('gene',): 25744}}\n"
     ]
    }
   ],
   "source": [
    "#examine this gff3 promoter file and compare to all promoters\n",
    "in_file = overlapping_promoters\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('Chr1',): 8771,\n",
      "            ('Chr2',): 5265,\n",
      "            ('Chr3',): 6544,\n",
      "            ('Chr4',): 5007,\n",
      "            ('Chr5',): 7469,\n",
      "            ('ChrC',): 133,\n",
      "            ('ChrM',): 152},\n",
      " 'gff_source': {('Araport11',): 33341},\n",
      " 'gff_source_type': {('Araport11', 'gene'): 33341},\n",
      " 'gff_type': {('gene',): 33341}}\n"
     ]
    }
   ],
   "source": [
    "# all promoters\n",
    "in_file = promoters\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #promoters overlapping only genes\n",
    "# promoter_overlap(promoters,genesonly_gff,promoterandgenes_only_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gff_id': {('Chr1',): 5328,\n",
      "            ('Chr2',): 2902,\n",
      "            ('Chr3',): 4005,\n",
      "            ('Chr4',): 2920,\n",
      "            ('Chr5',): 4387,\n",
      "            ('ChrC',): 133,\n",
      "            ('ChrM',): 151},\n",
      " 'gff_source': {('Araport11',): 19826},\n",
      " 'gff_source_type': {('Araport11', 'gene'): 19826},\n",
      " 'gff_type': {('gene',): 19826}}\n"
     ]
    }
   ],
   "source": [
    "# promoters overlapping genes only\n",
    "in_file = promoterandgenes_only_overlap\n",
    "examiner = GFFExaminer()\n",
    "in_handle = open(in_file)\n",
    "pprint.pprint(examiner.available_limits(in_handle))\n",
    "in_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_gff = \"../../data/genomes/genesonly_no_annotation_temp.gff3\"\n",
    "extract_genes(genes,temp_gff,genesonly_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "selected_genes = genesonly_gff\n",
    "\n",
    "promoters_incl_overlap = add_promoter(selected_genes,chromsizes_file_renamedChr,1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/tmp/pybedtools.3tdy2106.tmp)>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoters_incl_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "subtracted = remove_promoter_overlap(promoters_incl_overlap,selected_genes,promoterandgenes_only_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot1</th>\n",
       "      <th>strand</th>\n",
       "      <th>dot2</th>\n",
       "      <th>attributes</th>\n",
       "      <th>AGI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>9131</td>\n",
       "      <td>10130</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...</td>\n",
       "      <td>AT1G01020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>13715</td>\n",
       "      <td>14714</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:11118137,PMID:15010618,PMID:166036...</td>\n",
       "      <td>AT1G01030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>22121</td>\n",
       "      <td>23120</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...</td>\n",
       "      <td>AT1G01040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>33172</td>\n",
       "      <td>34171</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...</td>\n",
       "      <td>AT1G01050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>41018</td>\n",
       "      <td>42017</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15703057,locus:2200990;ID=AT1G0107...</td>\n",
       "      <td>AT1G01070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28246</th>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>309641</td>\n",
       "      <td>310513</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954608;ID=ATMG01250;Name=ATMG0...</td>\n",
       "      <td>ATMG01250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28247</th>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>318391</td>\n",
       "      <td>318587</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=NAD1,NADH DEHYDROGENASE 1,ND1,NAD%28P%29...</td>\n",
       "      <td>ATMG01275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28248</th>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>324595</td>\n",
       "      <td>325299</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954613;ID=ATMG01300;Name=ATMG0...</td>\n",
       "      <td>ATMG01300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28249</th>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>325711</td>\n",
       "      <td>326098</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954614;ID=ATMG01310;Name=ATMG0...</td>\n",
       "      <td>ATMG01310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28250</th>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>327889</td>\n",
       "      <td>327889</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954616;ID=ATMG01330;Name=ATMG0...</td>\n",
       "      <td>ATMG01330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28251 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     source  type   start    stop dot1 strand dot2  \\\n",
       "0      Chr1  Araport11  gene    9131   10130    .      -    .   \n",
       "1      Chr1  Araport11  gene   13715   14714    .      -    .   \n",
       "2      Chr1  Araport11  gene   22121   23120    .      +    .   \n",
       "3      Chr1  Araport11  gene   33172   34171    .      -    .   \n",
       "4      Chr1  Araport11  gene   41018   42017    .      -    .   \n",
       "...     ...        ...   ...     ...     ...  ...    ...  ...   \n",
       "28246  ChrM  Araport11  gene  309641  310513    .      +    .   \n",
       "28247  ChrM  Araport11  gene  318391  318587    .      -    .   \n",
       "28248  ChrM  Araport11  gene  324595  325299    .      +    .   \n",
       "28249  ChrM  Araport11  gene  325711  326098    .      +    .   \n",
       "28250  ChrM  Araport11  gene  327889  327889    .      -    .   \n",
       "\n",
       "                                              attributes        AGI  \n",
       "0      Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...  AT1G01020  \n",
       "1      Dbxref=PMID:11118137,PMID:15010618,PMID:166036...  AT1G01030  \n",
       "2      Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...  AT1G01040  \n",
       "3      Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...  AT1G01050  \n",
       "4      Dbxref=PMID:15703057,locus:2200990;ID=AT1G0107...  AT1G01070  \n",
       "...                                                  ...        ...  \n",
       "28246  Dbxref=locus:504954608;ID=ATMG01250;Name=ATMG0...  ATMG01250  \n",
       "28247  Alias=NAD1,NADH DEHYDROGENASE 1,ND1,NAD%28P%29...  ATMG01275  \n",
       "28248  Dbxref=locus:504954613;ID=ATMG01300;Name=ATMG0...  ATMG01300  \n",
       "28249  Dbxref=locus:504954614;ID=ATMG01310;Name=ATMG0...  ATMG01310  \n",
       "28250  Dbxref=locus:504954616;ID=ATMG01330;Name=ATMG0...  ATMG01330  \n",
       "\n",
       "[28251 rows x 10 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subtracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "subtracted = subtracted.sort_values(['chr','start']).reset_index(drop=True)\n",
    "cleaned_proms = filter_bad_proms(subtracted)\n",
    "with open(promoters,'w') as f:\n",
    "    cleaned_proms.to_csv(f,index=False,sep='\\t',header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds = add_5UTR(promoters, genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_proms_fiveUTR = filter_bad_proms(cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot1</th>\n",
       "      <th>strand</th>\n",
       "      <th>dot2</th>\n",
       "      <th>attributes</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2780</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>10426948</td>\n",
       "      <td>10426949</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18156295,locus:2019352;ID=AT1G2978...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2779</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>10426948</td>\n",
       "      <td>10426949</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G29775;Name=AT1G29775;Note=hypothetical ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>3817421</td>\n",
       "      <td>3817423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:14973164,PMID:15105442,PMID:170611...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25978</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>26583427</td>\n",
       "      <td>26583430</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT5G66595;Name=AT5G66595;Note=hypothetical ...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16834</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>6139712</td>\n",
       "      <td>6139716</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:17395691,locus:4515103369;ID=AT4G0...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22015</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>7316395</td>\n",
       "      <td>7322743</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504956461;ID=AT5G22090;Name=AT5G2...</td>\n",
       "      <td>6348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18217</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>11820136</td>\n",
       "      <td>11827143</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18210371,locus:2127490;ID=AT4G2241...</td>\n",
       "      <td>7007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9469</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>14391551</td>\n",
       "      <td>14402499</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2055476;ID=AT2G34110;Name=AT2G341...</td>\n",
       "      <td>10948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9472</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>14401656</td>\n",
       "      <td>14413693</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2055466;ID=AT2G34100;Name=AT2G341...</td>\n",
       "      <td>12037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4624</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>21745354</td>\n",
       "      <td>21760166</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15111722,PMID:17956627,PMID:215999...</td>\n",
       "      <td>14812</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26092 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     source  type     start      stop dot1 strand dot2  \\\n",
       "2780   Chr1  Araport11  gene  10426948  10426949    .      +    .   \n",
       "2779   Chr1  Araport11  gene  10426948  10426949    .      -    .   \n",
       "1038   Chr1  Araport11  gene   3817421   3817423    .      -    .   \n",
       "25978  Chr5  Araport11  gene  26583427  26583430    .      -    .   \n",
       "16834  Chr4  Araport11  gene   6139712   6139716    .      +    .   \n",
       "...     ...        ...   ...       ...       ...  ...    ...  ...   \n",
       "22015  Chr5  Araport11  gene   7316395   7322743    .      -    .   \n",
       "18217  Chr4  Araport11  gene  11820136  11827143    .      -    .   \n",
       "9469   Chr2  Araport11  gene  14391551  14402499    .      +    .   \n",
       "9472   Chr2  Araport11  gene  14401656  14413693    .      -    .   \n",
       "4624   Chr1  Araport11  gene  21745354  21760166    .      +    .   \n",
       "\n",
       "                                              attributes  length  \n",
       "2780   Dbxref=PMID:18156295,locus:2019352;ID=AT1G2978...       1  \n",
       "2779   ID=AT1G29775;Name=AT1G29775;Note=hypothetical ...       1  \n",
       "1038   Dbxref=PMID:14973164,PMID:15105442,PMID:170611...       2  \n",
       "25978  ID=AT5G66595;Name=AT5G66595;Note=hypothetical ...       3  \n",
       "16834  Dbxref=PMID:17395691,locus:4515103369;ID=AT4G0...       4  \n",
       "...                                                  ...     ...  \n",
       "22015  Dbxref=locus:504956461;ID=AT5G22090;Name=AT5G2...    6348  \n",
       "18217  Dbxref=PMID:18210371,locus:2127490;ID=AT4G2241...    7007  \n",
       "9469   Dbxref=locus:2055476;ID=AT2G34110;Name=AT2G341...   10948  \n",
       "9472   Dbxref=locus:2055466;ID=AT2G34100;Name=AT2G341...   12037  \n",
       "4624   Dbxref=PMID:15111722,PMID:17956627,PMID:215999...   14812  \n",
       "\n",
       "[26092 rows x 10 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_proms_fiveUTR = cleaned_proms_fiveUTR[['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']]\n",
    "with open(promoters_5UTR,'w') as f:\n",
    "    cleaned_proms_fiveUTR.to_csv(f,index=False,sep='\\t',header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_proms = cleaned_proms[['chr', 'source', 'type', 'start','stop','dot1','strand','dot2','attributes']]\n",
    "with open(promoters,'w') as f:\n",
    "    cleaned_proms.to_csv(f,index=False,sep='\\t',header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'rstrip'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-62-7f0d6cbd914c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcleaned_proms_fiveUTR\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrstrip\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5272\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5273\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5274\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5275\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5276\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'rstrip'"
     ]
    }
   ],
   "source": [
    "cleaned_proms_fiveUTR.rstrip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAEMCAYAAAA4S+qsAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3dfVRU950/8PfcOw88COigwBAQN6QxJEiaxd3UbtxfahBoMgrriXKWpk1PWtxuSOxJTnPW7DkBWfNQun+06Uby4Glt3PzaXw+nbSyEIqWmm7BtY6QeDSFqoygqAygjKk/zcO/9/THMyMjDXGbgzp30/TonEeZ779zPDI5vvt/vvd9rUBRFARERkQpCtAsgIqLYwdAgIiLVGBpERKQaQ4OIiFRjaBARkWrGaBewWGRZxujoKEwmEwwGQ7TLISKKCYqiwOPxIDExEYIwvV/xmQ2N0dFRnDp1KtplEBHFpNtvvx1JSUnTHv/MhobJZALge+Fmszni5+vq6kJ+fn7Ez7NYWF9kWF9kWF/k9FKj2+3GqVOnAv+G3uwzGxr+ISmz2QyLxbIgz7lQz7NYWF9kWF9kWF/k9FTjbMP6qibCe3p6UFFRgZKSElRUVODs2bPTtpEkCXV1dSgqKsLGjRvR2Nioqq2jowNbtmxBfn4+6uvrZzz+mTNncPfdd8/aTkRE2lDV06itrUVlZSXKyspw4MAB1NTUYP/+/UHbNDU1obe3F21tbRgeHkZ5eTnWrVuHrKysOduys7Px/PPP4+DBg3C73dOOLUkSamtrUVRUtDCvmIiIwhaypzE0NITu7m7Y7XYAgN1uR3d3N5xOZ9B2LS0t2Lp1KwRBgNVqRVFREVpbW0O25eTk4M4774TROHN+vfHGG7j//vuxatWqSF4nEREtgJA9DYfDgfT0dIiiCAAQRRFpaWlwOBywWq1B22VmZga+t9ls6O/vD9k2lxMnTqCjowP79+9HQ0OD+lc1RVdXV1j7zaSzs3PBnmsxsL7IsL7IsL7IxUKNup0I93g8eO655/DSSy8FAisc+fn5CzK51NnZicLCwoifZ7GwvsiwvsiwvsjppUaXyzXnL9shQ8Nms2FgYACSJEEURUiShMHBQdhstmnb9fX1oaCgAEBw72KuttlcunQJvb292L59OwDg2rVrUBQFIyMj2L17d6iyiYhoEYSc00hNTUVeXh6am5sBAM3NzcjLywsamgKA0tJSNDY2QpZlOJ1OtLe3o6SkJGTbbDIzM/HBBx/g0KFDOHToEB599FFs27btMxEYl66M43/+fCHaZRARzZuq4aldu3Zh586daGhoQHJycuDU16qqKuzYsQNr1qxBWVkZjh07huLiYgBAdXU1srOzAWDOtiNHjuDpp5/GyMgIFEXBO++8gxdeeAHr169f8BerF7870ov/23oC+bmpSE2Jj3Y5RESqqQqN3NzcoGsr/Pbu3Rv4WhRF1NXVzbj/XG1r167Fe++9F7KGJ598Uk2pMcHjlQEA3WecWH/PLVGuhohIPa5yGwWS5AuNrjOXo1wJEdH8MDSiQJJ9t2X/+MxQlCshIpofhkYU+EPjXP91XB1xRbkaIiL1GBpR4J0cngKA7h7nHFsSEekLQyMKZFlBUoIJZqPAISoiiim6vSL8s8wrybCYjVhlS8HHnAwnohjCnkYUSLICUTDgrltTcebiVYxNeKJdEhGRKgyNKJAlX2jk35oKWeG8BhHFDoZGFHhlGaIoYFVmMgDAcXk0yhUREanD0IgCabKnEWfxTSlNuL1RroiISB2GRhRIsgJRNMBsFCAYgAm3FO2SiIhUYWhEgSTJMAoCDAZfb2PCxZ4GEcUGhkYUSLICQTAAAOLMInsaRBQzGBpRIMkKjKLvrY8zs6dBRLGDoREFkiRDDPQ0jOxpEFHMYGhEgSQrEMTJ0LCIPHuKiGIGQyMKJEmBUZgcnrIYMc7hKSKKEQyNKJBkGaLIiXAiij0MjSgIPnvKyOEpIooZDI0oCBqeMouYcLGnQUSxgaERBVOHp+It7GkQUexQFRo9PT2oqKhASUkJKioqcPbs2WnbSJKEuro6FBUVYePGjWhsbFTV1tHRgS1btiA/Px/19fVBz7lnzx489NBD2Lx5M7Zs2YL3338/zJepL97JtacAwGI2wuOVIU25mx8RkV6puglTbW0tKisrUVZWhgMHDqCmpgb79+8P2qapqQm9vb1oa2vD8PAwysvLsW7dOmRlZc3Zlp2djeeffx4HDx6E2+0Oes6CggI89thjiI+Px4kTJ/DII4+go6MDcXFxC/cORIEsKxAnL+6Lt4gAfOtPJcaz40dE+hbyX6mhoSF0d3fDbrcDAOx2O7q7u+F0Bt8DoqWlBVu3boUgCLBarSgqKkJra2vItpycHNx5550wGqfn1/r16xEfHw8AWL16NRRFwfDwcGSvWAckOfjiPoAr3RJRbAjZ03A4HEhPT4co+n4jFkURaWlpcDgcsFqtQdtlZmYGvrfZbOjv7w/Zptbbb7+NlStXIiMjY177dXV1zWv7uXR2dkb8HHGJS+H2SLgyfA3Huk9j4JIvBD/6pAdJpvGo17eYWF9kWF9k9F4fEBs1xsQ9wg8fPoyXX34ZP/7xj+e9b35+PiwWS8Q1dHZ2orCwMOLnGXSOQZI/xphHgOOqiFGP70eQmGRF4V3zC8TFqG+xsL7IsL7I6L0+QD81ulyuOX/ZDjk8ZbPZMDAwAEnynRYqSRIGBwdhs9mmbdfX1xf43uFwBHoFc7WFcvToUTzzzDPYs2cPbr31VlX76J0sA4LBNzxlmpzbcHl42i0R6V/I0EhNTUVeXh6am5sBAM3NzcjLywsamgKA0tJSNDY2QpZlOJ1OtLe3o6SkJGTbXI4fP46nnnoKP/zhD3HXXXeF8/p0R1EUyMqNi/tMxsnQ4FXhRBQDVA1P7dq1Czt37kRDQwOSk5MDp8ZWVVVhx44dWLNmDcrKynDs2DEUFxcDAKqrq5GdnQ0Ac7YdOXIETz/9NEZGRqAoCt555x288MILWL9+Perq6jAxMYGamppALd/73vewevXqhXsHNKYovj8nMyMQGhPsaRBRDFAVGrm5uUHXVvjt3bs38LUoiqirq5tx/7na1q5di/fee2/Gtl/84hdqyospkuy7HoM9DSKKRbwwQGOS7OtqGPxzGkbfWWkunnJLRDGAoaExeTI0/BPhRk6EE1EMYWhoTJImQ0Pwh4bvTw5PEVEsYGhoTAr0NHzfGwwGmIwCexpEFBMYGhqTleCeBuCbDGdPg4hiAUNDY/7VbP1zGgDY0yCimMHQ0Fjg7KkpPQ2jyJ4GEcUGhobGbj57CgDMRoEX9xFRTGBoaCwwET7lnTcaBV6nQUQxgaGhMWmGnobJKHJOg4hiAkNDY4FlRG6eCOecBhHFAIaGxgJzGlNPuRV59hQRxQaGhsZuXnsKYE+DiGIHQ0NjN5YRufGY0SjA45UDgUJEpFcMDY3NdMrtjeXReQYVEekbQ0Nj0kxzGpPLo4+7GBpEpG8MDY3NePaUyBsxEVFsYGhoTJ5lIhxgT4OI9I+hoTH/8JQoTA+NCfY0iEjnGBoau3HK7Y3HboQGexpEpG8MDY3NdHGf/5avEy72NIhI31SFRk9PDyoqKlBSUoKKigqcPXt22jaSJKGurg5FRUXYuHEjGhsbVbV1dHRgy5YtyM/PR319vernjFWz3U8DYE+DiPTPqGaj2tpaVFZWoqysDAcOHEBNTQ32798ftE1TUxN6e3vR1taG4eFhlJeXY926dcjKypqzLTs7G88//zwOHjwIt9ut+jlj1cyn3Pp7GgwNItK3kD2NoaEhdHd3w263AwDsdju6u7vhdDqDtmtpacHWrVshCAKsViuKiorQ2toasi0nJwd33nknjMbp+TXXfrEqcLvXGXsaHJ4iIn0LGRoOhwPp6ekQRd8FaKIoIi0tDQ6HY9p2mZmZge9tNhv6+/tDtoU6djj76Zl/GRHD1GVERAEGAOMcniIinVM1PBXLurq6Fuy5Ojs7I36O4WseAMCFC+ennXZ7rrcPnZ1jUa1vMbG+yLC+yOi9PiA2agwZGjabDQMDA5AkCaIoQpIkDA4OwmazTduur68PBQUFAIJ7CXO1hTp2OPtNlZ+fD4vFMq99ZtLZ2YnCwsKIn+fohY8AAKtycoKGqOI+/gQpy1JRWPj5qNa3WFhfZFhfZPReH6CfGl0u15y/bIccnkpNTUVeXh6am5sBAM3NzcjLy4PVag3arrS0FI2NjZBlGU6nE+3t7SgpKQnZNpdw99OzwBXhNz0eZxZ5yi0R6Z6q4aldu3Zh586daGhoQHJycuDU2KqqKuzYsQNr1qxBWVkZjh07huLiYgBAdXU1srOzAWDOtiNHjuDpp5/GyMgIFEXBO++8gxdeeAHr16+fc79YJckKBEPwMiIAYDaJPOWWiHRPVWjk5ubOeI3E3r17A1+Looi6uroZ95+rbe3atXjvvffmvV+skmQl6HRbvzgzQ4OI9I9XhGtMlpVpvQxgsqfB4Ski0jmGhsZm62mYTSLvE05EusfQ0JgkyUFnTfmZjQJDg4h0j6GhMXlyIvxmJqMIN0ODiHSOoaGx2YenBLg9chQqIiJSj6GhMWmWiXCTUYTby54GEekbQ0Nj8mw9DaMAt0eCMrmgIRGRHjE0NOa7uG/ms6cUBfBKHKIiIv1iaGhMkmUIM7zrxsnl0V2c1yAiHWNoaEyeracxGRo8g4qI9IyhobHZhqdMRt/9ShgaRKRnDA2NSbICwyyn3ALgBX5EpGsMDY2FGp7ycE6DiHSMoaExSVJmnAj3D0+xp0FEesbQ0Jgkz7L2lIkT4USkfwwNjc22jAgnwokoFjA0NDbbnIYpcMot5zSISL8YGhqbdcFCI8+eIiL9Y2hobK4FCwFw0UIi0jWGhsZmu58GJ8KJKBYwNDTmW3tqrolwzmkQkX6pCo2enh5UVFSgpKQEFRUVOHv27LRtJElCXV0dioqKsHHjRjQ2NkbcNjQ0hO3bt2PTpk0oLS3Frl274PV6I3i50TfbMiJG0QCDgT0NItI3VaFRW1uLyspKHDx4EJWVlaipqZm2TVNTE3p7e9HW1oaf//zn+K//+i9cuHAhorbXXnsNubm5aGpqQlNTEz7++GO0tbUt1GuPCnmWZUQMBgPMJpET4USkayFDY2hoCN3d3bDb7QAAu92O7u5uOJ3OoO1aWlqwdetWCIIAq9WKoqIitLa2RtRmMBgwOjoKWZbhdrvh8XiQnp6+oG+A1iRp5p4GAJh5n3Ai0rmQoeFwOJCeng5R9I25i6KItLQ0OByOadtlZmYGvrfZbOjv74+o7fHHH0dPTw/uu+++wH+FhYXhvlZdkGaZCAd4n3Ai0j9jtAuYS2trK1avXo0333wTo6OjqKqqQmtrK0pLS1U/R1dX14LV09nZGfFzyLKM6yPXcK7XE/S4LUWCInnRP3gp7OMsRH2LifVFhvVFRu/1AbFRY8jQsNlsGBgYgCRJEEURkiRhcHAQNptt2nZ9fX0oKCgAENyDCLftrbfewosvvghBEJCUlIQNGzbggw8+mFdo5Ofnw2KxqN5+Np2dnRH3chRFgfzTC1iWshQ5KzOC2mwZaUhO6kXikoSwjrMQ9S0m1hcZ1hcZvdcH6KdGl8s15y/bIYenUlNTkZeXh+bmZgBAc3Mz8vLyYLVag7YrLS1FY2MjZFmG0+lEe3s7SkpKImrLysrCe++9BwBwu9344x//iM997nNhvA36IMkKAMx4yi3gH57inAYR6Zeq4aldu3Zh586daGhoQHJyMurr6wEAVVVV2LFjB9asWYOysjIcO3YMxcXFAIDq6mpkZ2cDQNht//7v/47a2lps2rQJkiTh3nvvxbZt2xbw5WsrEBqzTYSbRLi9nNMgIv1SFRq5ublB10/47d27N/C1KIqoq6ubcf9w21auXIl9+/apKTEmSJIvEAyz9O/MJhHXR90aVkREND+8IlxDoXoaFhNPuSUifWNoaMg72dOYdU7DKPKUWyLSNYaGhuSQcxoCrwgnIl1jaGhIkkKdPcXhKSLSN4aGhrzy5PDUrFeEMzSISN8YGhry9zRmugkTMHmdhleGoihalkVEpBpDQ0NyiIv7LCbf+l4eXqtBRDrF0NBQ4OypOS7uA3hPDSLSL4aGhkIvI+ILDZ5BRUR6xdDQUODsqdkmwo3++4RzeIqI9ImhoSFp8uyp2SfCOTxFRPrG0NCQf3hKDDERzuEpItIrhoaG1JxyC7CnQUT6xdDQkH94KtREOOc0iEivGBoaurHK7cztgdDwsqdBRPrE0NBQqLWnLJwIJyKdY2hoKDA8xbOniChGMTQ05B+eMsx6Pw3fj8PFOQ0i0imGhoYkLiNCRDGOoaEhtcuIMDSISK+M0S7gr4l3jmVEvJIM59VxGAzAlWsuDDrHAADxcUYkJZi1LJOIaFaqeho9PT2oqKhASUkJKioqcPbs2WnbSJKEuro6FBUVYePGjWhsbIy4DQBaWlqwadMm2O12bNq0CZcvXw7zpUbfXBPhLo+Eo6cuQRQEXLh0HX8+OYg/nxzE+IRX6zKJiGalqqdRW1uLyspKlJWV4cCBA6ipqcH+/fuDtmlqakJvby/a2towPDyM8vJyrFu3DllZWWG3ffTRR3jllVfw5ptvYsWKFbh+/TrM5tj9rTvU/TQAwCgaAqfmEhHpTciextDQELq7u2G32wEAdrsd3d3dcDqdQdu1tLRg69atEAQBVqsVRUVFaG1tjajtJz/5CR577DGsWLECAJCUlASLxbJwr15j3hDLiACAURQC990gItKbkKHhcDiQnp4OUfRN0oqiiLS0NDgcjmnbZWZmBr632Wzo7++PqO306dM4f/48vvKVr+Cf/umf0NDQENO3Qg21jAjgD43YfY1E9Nmm64lwSZJw8uRJ7Nu3D263G9/85jeRmZmJ8vJy1c/R1dW1YPV0dnZGtP+FC9cAAOfPn5vW27gl1YRzvecgSR5cH5FwrvccAMCWIuF8z7Am9S021hcZ1hcZvdcHxEaNIUPDZrNhYGAAkiRBFEVIkoTBwUHYbLZp2/X19aGgoABAcA8i3LbMzEyUlpbCbDbDbDbjgQcewPHjx+cVGvn5+QsypNXZ2YnCwsKInqN78BMYuq5hVc6qaW1LlixBzsocJHz6F5iNAnJW5gAAbBlpSLPmalLfYmJ9kWF9kdF7fYB+anS5XHP+sh1yeCo1NRV5eXlobm4GADQ3NyMvLw9WqzVou9LSUjQ2NkKWZTidTrS3t6OkpCSiNrvdjo6ODiiKAo/Hgz/96U+44447wnsndECSZIjC3G+5UTRweIqIdEvV8NSuXbuwc+dONDQ0IDk5GfX19QCAqqoq7NixA2vWrEFZWRmOHTuG4uJiAEB1dTWys7MBIOy2hx56CF1dXXjwwQchCALuu+8+PPzwwwv48rUlycqsN2DyE0UBbi9PsyUifVIVGrm5udOunwCAvXv3Br4WRRF1dXUz7h9umyAIePbZZ/Hss8+qKVP3JFmZcxIc4NlTRKRvXEYkAt/d/yEaf3dK9faSJEMUQ4UGr9MgIv1iaIRJURQc+WQALX84q/o0YElWIM5xjQbAngYR6RtDI0yj4x643BIuD4/jzMWrqvaRJAWiOPdbLooCexpEpFsMjTBdvjoR+PqDj/tV7SPJcug5DcHAngYR6RZDI0yXh8cBAPEWIz7oUhkakrqzpyRZiekr34nos4uhEaahq77QeGBtNs70XQ0sZT4XNafcGicnynmtBhHpEUMjTJeHJ2AwAF/+4ioAwOHu0L0Nr6RieGpyzkPiEBUR6RBDI0xDV8exLCkOKzOSkZW2RNUQlbqehu9HwnkNItIjhkaYLg+PY/nSOADAvXdl4KPTl3F9zD3nPrKsqFpGBAC8MoeniEh/GBphunx1Aqkp8QCA++6+BZKs4I8fOebcR83wlMjhKSLSMYZGmIaujmP50nhcH3MjKcGEtGXx+N2HvRh0jmHQOTZjr0OSlZBXhJuMvh+J28PQICL9YWiEYWzCg7EJL5anxGF8woujpy5hZXoSPulxouPYxVnv7S2rmNOIM/uWA5twc9FCItIfhkYY/Ndo+IenAOC27GVQAJy+MPvV4V5JhhBiGZE4s+8OiRNuKfJCiYgWGEMjDP6rwZcvvREaqSlxsCbH4S/nZ7/LnprhqXjLZE/DxZ4GEekPQyMMQ4GeRlzQ45/LXgrH0ChGZjmLSs1NmExGAYKBw1NEpE8MjTD4exo3h8Zt2UsBAKdnWcBQzf00DAYD4ixGDk8RkS4xNMIwdHUcS5dYYDKKQY8vXWJBcqIZfZdGZ9xPzcV9gG8yfMLF0CAi/WFohOHy8DhSl8bN2Ja5IhEXL49AnmHBQd/wlJrQEDHO4Ski0iGGRhiGrk5g+ZQzp6a6ZcUSuNwS+i6NBD3ed3kEV667kBBnCvn8vuEphgYR6Q9DIwyXh8enzWf4ZS5fAgA4ee5K4DFJVvD9n/4ZZpOIki/khHz+OLPI4Ski0iWGxjxNuLwYGfcEnW47VXKiGUkJZpyYEhq/fPcvOHHuCr61pQDW5JnDZqo4s6+nwXtqEJHeqAqNnp4eVFRUoKSkBBUVFTh79uy0bSRJQl1dHYqKirBx40Y0NjZG3OZ35swZ3H333aivrw/jJS6soWv+M6dmDg0AuGVFIk6dvwJZVnDm4lX89OAJ/MPdmfg/99yi6hjxFiMUhUuJEJH+GNVsVFtbi8rKSpSVleHAgQOoqanB/v37g7ZpampCb28v2traMDw8jPLycqxbtw5ZWVlhtwG+UKmtrUVRUdHCv/ow+G++NNvwFABkrliCE+eu4NT5K/jBz/6M5EQL/nVLAQwhrgb3u3FVOOc1iEhfQvY0hoaG0N3dDbvdDgCw2+3o7u6G0+kM2q6lpQVbt26FIAiwWq0oKipCa2trRG0A8MYbb+D+++/HqlWrFuo1R2R03APANww1m1tW+OY1Xtx3GH2XR/GdrxQiZYlF9TFurD/FeQ0i0peQoeFwOJCeng5R9P32K4oi0tLS4HA4pm2XmZkZ+N5ms6G/vz+ithMnTqCjowNf//rXw3x5C29sciFC/3IfM0lONMOaHIcr112oKFqNNbctn9cx4iy+93qcS4kQkc6oGp6KBo/Hg+eeew4vvfRSILDC0dXVtWA1dXZ24uRffKfSnjrxMS7GiTDGL8W53gvTti1cbcWFwevIyzLiWPfpwONGUxzO9Z6btv0tqabA4yMTvh7GRccAHP2JON8z+3pWN9enZ6wvMqwvMnqvD4iNGkOGhs1mw8DAACRJgiiKkCQJg4ODsNls07br6+tDQUEBgOAeRDhtly5dQm9vL7Zv3w4AuHbtGhRFwcjICHbv3q36Bebn58NiUT80NJvOzk4UFhbi0ysnAQxj3b2FMBlFDDrHkHN1eqitzlmGk+euYPD6zY8vQc7K6afdLlly4/EJtxe/P/4xEpcshS3DhjRrrur69Ir1RYb1RUbv9QH6qdHlcs35y3bI4anU1FTk5eWhubkZANDc3Iy8vDxYrdag7UpLS9HY2AhZluF0OtHe3o6SkpKw2zIzM/HBBx/g0KFDOHToEB599FFs27ZtXoGxGMYnvDAZhWlLiCwki0mEAZwIJyL9UTU8tWvXLuzcuRMNDQ1ITk4OnPpaVVWFHTt2YM2aNSgrK8OxY8dQXFwMAKiurkZ2djYAhN2mR2MTXiTELe6oHhctJCK9UvWvX25u7ozXT+zduzfwtSiKqKurm3H/cNumevLJJ9WUuujGJrxIsIReCiRSvqvC2dMgIn3R7US43sQlLsWgcwzD1ydgMgoYdI4BAFyexekN+K4KZ0+DiPSFoaGSVwb+fHIQlyZvwPTnk4MAfBPeiyHOIuLa6Mw3cyIiihauPTVPbq8Es2nxJsH9fPfU4PAUEekLQ2Oe3B4ZZtPiv22+e2pIXLSQiHSFoTFPbo8E8yKebusXZzFClpVFmzMhIgoHQ2Oe3F7tehoAMDLmWfRjERGpxdCYB0mSIcvKol7Y5xc3ubbVyDhDg4j0g6ExD26v7/4WWkyEx0+udDsyxjOoiEg/GBrz4J6cX9B0eIo9DSLSEYbGPAR6GhoOT40yNIhIRxga8+DRsKdhmexpXOdEOBHpCENjHvz37NaipyEYDLCYRVznnAYR6QhDYx7cXn9PY/FDAwCSEkxwXpvQ5FhERGowNObBPxFu0mB4CgCSEy24dGVck2MREanB0JgHLSfCAd+9xi8Nj0OWuZQIEekDQ2Me3B4JBgNgFA2aHC8l0QyvJOPKdQ5REZE+MDTmwe2VYTaKMBi0CY3kRN+9zR2XRzU5HhFRKAyNefB4JE1Ot/VLXmIGAPQPjWl2TCKiuTA05sG3LLo28xkAkJRghsEA9DvZ0yAifWBozIPbK8Fs1O4tEwUDUpPjMMCeBhHpBENjHtweCSYNexoAsHxZAhxD7GkQkT6oCo2enh5UVFSgpKQEFRUVOHv27LRtJElCXV0dioqKsHHjRjQ2NkbctmfPHjz00EPYvHkztmzZgvfffz+Clxo5/0S4llYsjWdPg4h0w6hmo9raWlRWVqKsrAwHDhxATU0N9u/fH7RNU1MTent70dbWhuHhYZSXl2PdunXIysoKu62goACPPfYY4uPjceLECTzyyCPo6OhAXFzcorwZobg1nggHgLRl8RgecWHc5UW8RdWPi4ho0YT8F3BoaAjd3d2w2+0AALvdju7ubjidzqDtWlpasHXrVgiCAKvViqKiIrS2tkbUtn79esTHxwMAVq9eDUVRMDw8vHCvfp7cHlmTGzBNtWKp7/UPONnbIKLoCxkaDocD6enpEEXfP5aiKCItLQ0Oh2PadpmZmYHvbTYb+vv7I2qb6u2338bKlSuRkZExn9e3YGRZgVfS5lavU/lDg9dqEJEexMR4x+HDh/Hyyy/jxz/+8bz37erqWpAaPEISAGB05CrO9boCj9+SasK53nPTtl+ox1Nu94XkkWMnYXb3zVljZ2fn3C8iylhfZFhfZPReHxAbNYYMDZvNhoGBAUiSBFEUIUkSBgcHYbPZpm3X19eHgoICAME9iHDbAODo0aN45pln0NDQgFtvvXXeLzA/Px8Wi2Xe+93sfwvVbwkAABAcSURBVA5/AgBIX7EcOStTA48vWbIEOStzpm2/UI/n5qQhMf40TAlWFBYWzFpfZ2cnCgsLVb2WaGB9kWF9kdF7fYB+anS5XHP+sh1yrCU1NRV5eXlobm4GADQ3NyMvLw9WqzVou9LSUjQ2NkKWZTidTrS3t6OkpCSituPHj+Opp57CD3/4Q9x1113hvQMLxOXR7v7gN8tITUA/T7slIh1QNTy1a9cu7Ny5Ew0NDUhOTkZ9fT0AoKqqCjt27MCaNWtQVlaGY8eOobi4GABQXV2N7OxsAAi7ra6uDhMTE6ipqQnU8r3vfQ+rV69eiNc+Ly63f4Vb7S9tybAm4qzjqubHJSK6marQyM3NDbp+wm/v3r2Br0VRRF1d3Yz7h9v2i1/8Qk15mnB5tL0B01S25Yn44GMHXB4Jligcn4jIj1eEq+QfntL6lFuvJCMnIwleScG7R85j0DmGQecYbwNLRFERE2dP6cGE29/T0DZnXR4JoxNeWMwifnv4xtlVf7s6DUkJZk1rISJiT0OlaE6Ei4IBt2am4GzfNUiSrPnxiYj8GBoq+SfCTVGYCAeAW29Jgdsr4/zgSFSOT0QEMDRUc3kkmIwCBI3u2nez7LQlMJsEnL4YvWVUiIgYGipNuOWonG7rJ4oCVtlS0HPxGiRZiVodRPTXjaGh0oRbisp8xlS5WSlweST0XeIQFRFFB0NDpeERD5YkmKJaw8r0JBhFAZ9e4BAVEUUHQ0OlK9fdSE6MfA2rSBhFAatsyejpuwaZQ1REFAUMDRVGxz0Yc0lISYz+dRG5WSkYd3lx6vyVaJdCRH+FGBoq+BcLTF4S/dBYmZEEo2hA54nBaJdCRH+FGBoq9E/eNS/aw1MAYDaKWJmejM4TAxyiIiLNMTRUGPD3NHQwPAX4hqiujrhx8hyHqIhIWwwNFfqHxpBgEXWzwuwqWzKMogEdxy9GuxQi+ivD0FChf2gUy5L00csAfOtfff72NDR39KD9cG+0yyGivyJc5VaFfucYlusoNADg6w/dCa8k4+WfH4Xz2gRyl0W7IiL6a8CeRgiSrGDQOYZlydG9sO9mJqOAf91SgC/k2/Dfv/kEn14SeJ8NIlp0DI0QhobHIckKrDrrabg8Eo5/ehmfv30FUpaYcaDjAo6cGMD4hDfapRHRZxhDIwTH5JlTeprTmEoUDPhCvg3Xx2WeTUVEi46hEUL/kO8aDb31NKbKvSUFSxNFHP64H+7Je5kTES0GhkYIA85RiIIByYn6mtOYymAwIC87HiPjHtT/9xG8/qvj+ONHfdEui4g+gxgaIfQPjSFtWQJEITo3X1IrNdmIdfk2CAYD2g/34sWffIi9Bz7iVeNEtKBUnXLb09ODnTt3Ynh4GEuXLkV9fT1WrVoVtI0kSXj++efx/vvvw2AwYPv27di6deuitWmlf2gU6akJmh4zXH97RxoKblsOwWDA/2s/iV+/dwYXB0fw+c+tgEeS8TeZKSi8Iw2GKN19kIhin6rQqK2tRWVlJcrKynDgwAHU1NRg//79Qds0NTWht7cXbW1tGB4eRnl5OdatW4esrKxFadOCLCvoHxrDP9ydqcnxFoLLI+HkuStYvXIZxie8+MNHjqDFDW/LSsGm9bfi2qgHFwavY+kSC+5YZUXm8kRcH3NjdMKLnIwkpKbEAwAURYHLLSHOwkt6iEhFaAwNDaG7uxv79u0DANjtduzevRtOpxNWqzWwXUtLC7Zu3QpBEGC1WlFUVITW1lZ885vfXJS2UBTFNyzjds//uoXzA9fxo6YuDFweg2iQkb0iHooszTiWJ3k9EDB98lnrx81GAwRINx43AIWrU3FHTjJkWcFdf2PFyd6r+N2RXuz79XEAQEKcEeMuCQf/eHra8y9fGo94ixEDzjG4PRIyUhPxueylEAQDrl53we2VsDTJgqVLLPDKCiZcXhgMBsRbRJiNIibcEsYmPBBFARazCOfQNXzUdxwerwSTUUScWYSiKBh1eeHxSIgzG5FgMcJgMECSZQCAySTCJAjwKjK8XhkGGGA0ChANBnhlGV5JhmDwPWaAAW6PBI8kwSSKMJl8Py23R4IkKzCJAsxGETAAigIoUHx/KgoUAAP913HG2Q2DYIDBABjg+1OAATAA/h0VBO8rK4DB4DuLTTAYIMkKPB4JCnzX0hhFXx3+YwKA7PsG/oHDqcczwDB5OAP8GykAHP3Xce7qCd9zBHacaT///5SgY/rrReCYvv18W/u/Dj6mMvU55jymAf0DI7hw7eS0/fyH9D+9MHnQ2d5XWZncV/Y9ZgACP5PgfYNfY9B+im8fo2iAAAPcXgkXHdfwl8sfw2wSIRoM8MgyJEmGKAowCgIUKPB6ZciyAtEowCQIkBQFHq8ERZn8WQoCvLIMt0eCwWCAySTAaBDg9kpweyUYBQFms2+pIbdbgluSYRZ9j8myggm3F5IEWEwCLGYjPF4J4y4JiqIgziLCOXQdpy51YdTlhWgwICHOCKNRwNi4FyPjblhMRixJMMEA4OqoG6PjHiQmmJCSaIHb48Xl4QmMu7xITYlD2rIE3F+YjTjz/Jc+8v+bOfXvy1QhQ8PhcCA9PR2i6Du4KIpIS0uDw+EICg2Hw4HMzBu/kdtsNvT39y9aWygejwcAcOrUKVXb3+zhLyQBSJr87jokF5A6wwlUVy8N6+Lx1FWJAIanPz759eiVMWQlAY9+ab6Xji+Z8rU8+acRN/7quG7a/uaAk3z/pScD8Ex5Hs9N23kn/4uOv1maBGA8ascPJTs5CcBotMuYVVbSEgD6vQ3xrdZkABPRLmMOHiA1CcGfp8nPQ8rU7ycfswK+KWkJgO8MT6QJAMzwfb5G8OmpTyKryONBXFzctMc/s2MOiYmJuP3222EymTiGT0SkkqIo8Hg8SExMnLE9ZGjYbDYMDAxAkiSIoghJkjA4OAibzTZtu76+PhQUFAAI7iUsRlsogiAgKSkp9IZERBRkph6GX8hTblNTU5GXl4fm5mYAQHNzM/Ly8oKGpgCgtLQUjY2NkGUZTqcT7e3tKCkpWbQ2IiLSnqrhqV27dmHnzp1oaGhAcnIy6uvrAQBVVVXYsWMH1qxZg7KyMhw7dgzFxcUAgOrqamRnZwPAorQREZH2DMpsU+REREQ34RXhRESkGkODiIhUY2gQEZFqDA0iIlKNoRFCT08PKioqUFJSgoqKCpw9e3bRj3nlyhVUVVWhpKQEmzZtwhNPPAGn0xmynnDbIvHKK69g9erVgSvv9VKfy+VCbW0tiouLsWnTJjz33HO6qu/dd99FeXk5ysrKsGnTJrS1tUW1vvr6emzYsCHoZ7lY9YRT60z1zfU50UN9U938OdG6vgWl0Jy++tWvKm+//baiKIry9ttvK1/96lcX/ZhXrlxR/vSnPwW+/+53v6s8++yzIesJty1cXV1dyje+8Q3l/vvvV06ePKmr+nbv3q288MILiizLiqIoyqVLl3RTnyzLytq1awPv2SeffKJ8/vOfVyRJilp9H374odLX16d86UtfCtQVyTEXutaZ6pvrc6KH+vxm+pxoXd9CYmjM4fLly0phYaHi9XoVRVEUr9erFBYWKkNDQ5rW0draqjz66KNz1hNuW7hcLpeybds2pbe3N/BB0Ut9IyMjSmFhoTIyMhL0uF7qk2VZ+fu//3vlyJEjiqIoyuHDh5Xi4mJd1Df1H73FqCfSWmf6R9nP/zlZrNrDqW+mz0k061sIn9m1pxaC2sUaF5Msy/jZz36GDRs2zFmPoihhtYX7Ol5++WVs3rw56GJLvdR3/vx5LF26FK+88go++OADJCYm4tvf/jbi4uJ0UZ/BYMAPfvADPP7440hISMDo6Chef/113bx/fotRz2LVOvVzsli1h1PfTJ8TPdUXDs5p6Nzu3buRkJCARx55JNqlBBw9ehQfffQRKisro13KjLxeL86fP48777wTv/zlL/Gd73wHTz75JMbGxqJdGgBffa+//joaGhrw7rvv4tVXX8VTTz2lm/piET8n2mFPYw5qF2tcLPX19Th37hxee+01CIIwZz2KooTVFo4PP/wQZ86cwQMPPAAA6O/vxze+8Q08++yzuqgvMzMTRqMRdrsdAHD33Xdj2bJliIuL00V9n3zyCQYHB1FYWAgAKCwsRHx8PCwWiy7q81uMv2+LUevNn5PFqn2+ZvucvPTSS8jLy4t6feFiT2MOahdrXAzf//730dXVhT179sBsNoesJ9y2cGzfvh0dHR04dOgQDh06hIyMDPzoRz/Cgw8+qIv6rFYr7r33Xvzv//4vAN/ZJkNDQ1i1apUu6svIyEB/fz/OnDkDADh9+jQuX76MnJwcXdTntxh/3xa61pk+J4tV+3zN9jm57777dFFf2LSYOIlln376qfLwww8rxcXFysMPP6ycPn160Y956tQp5fbbb1eKi4uVzZs3K5s3b1Yef/zxkPWE2xapqRN8eqmvt7dXeeSRRxS73a6Ul5crv//973VV34EDBxS73a5s2rRJ2bRpk/Lb3/42qvXt3r1bWb9+vZKXl6d88YtfVB588MFFqyecWmeqb67PiR7qu9nNk+R6+bs4X1ywkIiIVOPwFBERqcbQICIi1RgaRESkGkODiIhUY2gQEZFqDA2iediwYQP+8Ic/aHrMCxcuYPXq1fB6vZoel2gmDA0inYlGMBGpxdAgIiLVGBpEYZBlGW+88QaKiopw77334tvf/jaGh4cB3BhO+tWvfoX7778f9957L1599dXAvhMTE/i3f/s3/N3f/R2+/OUvY+/evfjHf/xHAMAzzzyDvr4+fOtb38I999yDvXv3BvZramqa8fmItMTQIArD/v370d7ejrfeegvvv/8+UlJS8B//8R9B23R2dqK1tRVvvvkm9uzZg9OnTwPw3cXt4sWLaG9vx759+/DrX/86sM9//ud/IjMzE6+99hqOHj2KqqqqkM9HpCWGBlEYfv7zn+Opp55CRkYGzGYznnjiCRw8eDBosvqJJ55AXFwc7rjjDtxxxx04ceIEAOA3v/kN/uVf/gUpKSnIyMjA1772NVXHnO35iLTEpdGJwtDX14fq6urAUtwAIAgChoaGAt8vX7488HV8fHzgfhk3L2WdkZGh6pizPR+RlhgaRGHIyMjAiy++GLgnxlQXLlyYc98VK1agv78ft912GwDffRaIYgWHp4jC8M///M/4wQ9+gIsXLwIAnE4n2tvbVe375S9/Ga+//jquXr2KgYEBvPXWW0Hty5cvx/nz5xe8ZqKFwNAgCsPXvvY1bNiwAY899hjuuecebNu2DcePH1e1b3V1NTIyMvDAAw/g61//OkpKSoJuILR9+3a8+uqrWLt2LX70ox8t1ksgCgvvp0EUZT/96U/R0tIyrcdBpEfsaRBpbHBwEJ2dnZBlGWfOnMG+fftQVFQU7bKIVOFEOJHGPB4PamtrceHCBSQlJeGhhx5CZWVltMsiUoXDU0REpBqHp4iISDWGBhERqcbQICIi1RgaRESkGkODiIhUY2gQEZFq/x+DipKojf6IlwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "graph = cleaned_proms_fiveUTR.copy()\n",
    "graph['length'] = (graph.start - graph.stop).abs()\n",
    "#graph.sort_values('length',ascending=True)\n",
    "\n",
    "sns.set(color_codes=True)\n",
    "sns.set_style(\"whitegrid\")\n",
    "dist_plot = graph['length']\n",
    "#create figure with no transparency\n",
    "dist_plot_fig = sns.distplot(dist_plot).get_figure()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
