{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_genes_file = '../../data/genomes/ara_housekeeping_list.out'\n",
    "all_promoters_bedfile = '../../data/FIMO/promoters_renamedChr.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_genes = pd.read_table(select_genes_file, sep='\\t', header=None)\n",
    "cols = ['gene','gene_type']\n",
    "select_genes.columns = cols\n",
    "all_promoters = pd.read_table(all_promoters_bedfile, sep='\\t', header=None)\n",
    "cols = ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']\n",
    "all_promoters.columns = cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT3G32260</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>AT1G65490</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>AT2G30360</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>AT5G56270</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>AT5G57785</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>AT2G46240</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type\n",
       "0    AT4G34270  housekeeping\n",
       "1    AT3G32260  housekeeping\n",
       "2    AT1G59830  housekeeping\n",
       "3    AT4G33380  housekeeping\n",
       "4    AT2G28390  housekeeping\n",
       "..         ...           ...\n",
       "401  AT1G65490      randCont\n",
       "402  AT2G30360      randCont\n",
       "403  AT5G56270      randCont\n",
       "404  AT5G57785      randCont\n",
       "405  AT2G46240      randCont\n",
       "\n",
       "[406 rows x 2 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2630</td>\n",
       "      <td>3630</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ANAC001,NAC domain containing protein 1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9130</td>\n",
       "      <td>10130</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10100</td>\n",
       "      <td>11100</td>\n",
       "      <td>AT1G03987</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G03987;Name=AT1G03987;locus_type=long_no...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>13714</td>\n",
       "      <td>14714</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:11118137,PMID:15010618,PMID:166036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>22120</td>\n",
       "      <td>23120</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33336</th>\n",
       "      <td>mitochondria</td>\n",
       "      <td>361052</td>\n",
       "      <td>362052</td>\n",
       "      <td>ATmitochondriaG01370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954620;ID=ATmitochondriaG01370...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33337</th>\n",
       "      <td>mitochondria</td>\n",
       "      <td>361179</td>\n",
       "      <td>362179</td>\n",
       "      <td>ATmitochondriaG01380</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PmitochondriaID:15067383,PmitochondriaI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33338</th>\n",
       "      <td>mitochondria</td>\n",
       "      <td>362724</td>\n",
       "      <td>363724</td>\n",
       "      <td>ATmitochondriaG01400</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954623;ID=ATmitochondriaG01400...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33339</th>\n",
       "      <td>mitochondria</td>\n",
       "      <td>363284</td>\n",
       "      <td>364284</td>\n",
       "      <td>ATmitochondriaG01390</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PmitochondriaID:14586555,PmitochondriaI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33340</th>\n",
       "      <td>mitochondria</td>\n",
       "      <td>366700</td>\n",
       "      <td>366924</td>\n",
       "      <td>ATmitochondriaG01410</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:504954624;ID=ATmitochondriaG01410...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33341 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr   start    stop                  gene dot strand  \\\n",
       "0                 1    2630    3630             AT1G01010   .      +   \n",
       "1                 1    9130   10130             AT1G01020   .      -   \n",
       "2                 1   10100   11100             AT1G03987   .      +   \n",
       "3                 1   13714   14714             AT1G01030   .      -   \n",
       "4                 1   22120   23120             AT1G01040   .      +   \n",
       "...             ...     ...     ...                   ...  ..    ...   \n",
       "33336  mitochondria  361052  362052  ATmitochondriaG01370   .      -   \n",
       "33337  mitochondria  361179  362179  ATmitochondriaG01380   .      -   \n",
       "33338  mitochondria  362724  363724  ATmitochondriaG01400   .      +   \n",
       "33339  mitochondria  363284  364284  ATmitochondriaG01390   .      -   \n",
       "33340  mitochondria  366700  366924  ATmitochondriaG01410   .      -   \n",
       "\n",
       "          source  type dot2                                            details  \n",
       "0      Araport11  gene    .  Alias=ANAC001,NAC domain containing protein 1;...  \n",
       "1      Araport11  gene    .  Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...  \n",
       "2      Araport11  gene    .  ID=AT1G03987;Name=AT1G03987;locus_type=long_no...  \n",
       "3      Araport11  gene    .  Dbxref=PMID:11118137,PMID:15010618,PMID:166036...  \n",
       "4      Araport11  gene    .  Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...  \n",
       "...          ...   ...  ...                                                ...  \n",
       "33336  Araport11  gene    .  Dbxref=locus:504954620;ID=ATmitochondriaG01370...  \n",
       "33337  Araport11  gene    .  Dbxref=PmitochondriaID:15067383,PmitochondriaI...  \n",
       "33338  Araport11  gene    .  Dbxref=locus:504954623;ID=ATmitochondriaG01400...  \n",
       "33339  Araport11  gene    .  Dbxref=PmitochondriaID:14586555,PmitochondriaI...  \n",
       "33340  Araport11  gene    .  Dbxref=locus:504954624;ID=ATmitochondriaG01410...  \n",
       "\n",
       "[33341 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_promoters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filter out unwanted genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(select_genes, all_promoters, on='gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>4</td>\n",
       "      <td>16406303</td>\n",
       "      <td>16407303</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT3G32260</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>3</td>\n",
       "      <td>13218167</td>\n",
       "      <td>13219167</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16166256,locus:2084673;ID=AT3G3226...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>1</td>\n",
       "      <td>22022503</td>\n",
       "      <td>22023503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:10091592,PMID:10080713,PMID:986939...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>4</td>\n",
       "      <td>16070739</td>\n",
       "      <td>16071739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16166256,PMID:17085508,PMID:179083...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>2</td>\n",
       "      <td>12143665</td>\n",
       "      <td>12144665</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=MON1,MONENSIN SENSITIVITY1;Dbxref=PMID:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>AT1G65490</td>\n",
       "      <td>randCont</td>\n",
       "      <td>1</td>\n",
       "      <td>24353645</td>\n",
       "      <td>24354645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15894741,PMID:18650403,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>AT2G30360</td>\n",
       "      <td>randCont</td>\n",
       "      <td>2</td>\n",
       "      <td>12938834</td>\n",
       "      <td>12939834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>AT5G56270</td>\n",
       "      <td>randCont</td>\n",
       "      <td>5</td>\n",
       "      <td>22778613</td>\n",
       "      <td>22779613</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATWRKY2,ARABIDOPSIS THALIANA WRKY DNA-BI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>AT5G57785</td>\n",
       "      <td>randCont</td>\n",
       "      <td>5</td>\n",
       "      <td>23409715</td>\n",
       "      <td>23410715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16258012,PMID:16792696,PMID:203512...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>AT2G46240</td>\n",
       "      <td>randCont</td>\n",
       "      <td>2</td>\n",
       "      <td>18985132</td>\n",
       "      <td>18986132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type chr     start      stop dot strand     source  \\\n",
       "0    AT4G34270  housekeeping   4  16406303  16407303   .      -  Araport11   \n",
       "1    AT3G32260  housekeeping   3  13218167  13219167   .      +  Araport11   \n",
       "2    AT1G59830  housekeeping   1  22022503  22023503   .      -  Araport11   \n",
       "3    AT4G33380  housekeeping   4  16070739  16071739   .      +  Araport11   \n",
       "4    AT2G28390  housekeeping   2  12143665  12144665   .      -  Araport11   \n",
       "..         ...           ...  ..       ...       ...  ..    ...        ...   \n",
       "401  AT1G65490      randCont   1  24353645  24354645   .      +  Araport11   \n",
       "402  AT2G30360      randCont   2  12938834  12939834   .      -  Araport11   \n",
       "403  AT5G56270      randCont   5  22778613  22779613   .      +  Araport11   \n",
       "404  AT5G57785      randCont   5  23409715  23410715   .      -  Araport11   \n",
       "405  AT2G46240      randCont   2  18985132  18986132   .      +  Araport11   \n",
       "\n",
       "     type dot2                                            details  \n",
       "0    gene    .  Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...  \n",
       "1    gene    .  Dbxref=PMID:16166256,locus:2084673;ID=AT3G3226...  \n",
       "2    gene    .  Dbxref=PMID:10091592,PMID:10080713,PMID:986939...  \n",
       "3    gene    .  Dbxref=PMID:16166256,PMID:17085508,PMID:179083...  \n",
       "4    gene    .  Alias=MON1,MONENSIN SENSITIVITY1;Dbxref=PMID:1...  \n",
       "..    ...  ...                                                ...  \n",
       "401  gene    .  Dbxref=PMID:15894741,PMID:18650403,PMID:207364...  \n",
       "402  gene    .  Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...  \n",
       "403  gene    .  Alias=ATWRKY2,ARABIDOPSIS THALIANA WRKY DNA-BI...  \n",
       "404  gene    .  Dbxref=PMID:16258012,PMID:16792696,PMID:203512...  \n",
       "405  gene    .  Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...  \n",
       "\n",
       "[406 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, gene_type, chr, start, stop, dot, strand, source, type, dot2, details]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if any NaN\n",
    "merged[pd.isna(merged.chr)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### change column order to bed file format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_proms = merged.loc[:, ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']]\n",
    "#sort the df by chromosome then by motif start position\n",
    "filtered_proms_sorted = filtered_proms.sort_values(['chr','start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>1</td>\n",
       "      <td>33171</td>\n",
       "      <td>34171</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>1</td>\n",
       "      <td>281760</td>\n",
       "      <td>282760</td>\n",
       "      <td>AT1G01780</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15642518,PMID:15659355,PMID:155317...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>316090</td>\n",
       "      <td>317090</td>\n",
       "      <td>AT1G01910</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18315867,PMID:18775970,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>1</td>\n",
       "      <td>344515</td>\n",
       "      <td>345515</td>\n",
       "      <td>AT1G02000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15247385,PMID:15911562,PMID:187759...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>1</td>\n",
       "      <td>728829</td>\n",
       "      <td>729829</td>\n",
       "      <td>AT1G03070</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=AtLFG4,LFG4,LIFEGUARD 4;Dbxref=PMID:1708...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>5</td>\n",
       "      <td>26120581</td>\n",
       "      <td>26121581</td>\n",
       "      <td>AT5G65360</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=H3.1,histone 3.1;Dbxref=PMID:12662315,PM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>5</td>\n",
       "      <td>26171008</td>\n",
       "      <td>26172008</td>\n",
       "      <td>AT5G65470</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18775970,locus:2168297;ID=AT5G6547...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>5</td>\n",
       "      <td>26467531</td>\n",
       "      <td>26468531</td>\n",
       "      <td>AT5G66250</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18650403,PMID:18775970,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>5</td>\n",
       "      <td>26817948</td>\n",
       "      <td>26818948</td>\n",
       "      <td>AT5G67210</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:21251108,PMID:21288268,locus:21555...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>5</td>\n",
       "      <td>26956914</td>\n",
       "      <td>26957914</td>\n",
       "      <td>AT5G67590</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12084824,PMID:12837548,PMID:146710...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     start      stop       gene dot strand     source  type dot2  \\\n",
       "229   1     33171     34171  AT1G01050   .      -  Araport11  gene    .   \n",
       "157   1    281760    282760  AT1G01780   .      +  Araport11  gene    .   \n",
       "45    1    316090    317090  AT1G01910   .      -  Araport11  gene    .   \n",
       "181   1    344515    345515  AT1G02000   .      +  Araport11  gene    .   \n",
       "149   1    728829    729829  AT1G03070   .      +  Araport11  gene    .   \n",
       "..   ..       ...       ...        ...  ..    ...        ...   ...  ...   \n",
       "346   5  26120581  26121581  AT5G65360   .      -  Araport11  gene    .   \n",
       "286   5  26171008  26172008  AT5G65470   .      +  Araport11  gene    .   \n",
       "360   5  26467531  26468531  AT5G66250   .      +  Araport11  gene    .   \n",
       "154   5  26817948  26818948  AT5G67210   .      +  Araport11  gene    .   \n",
       "210   5  26956914  26957914  AT5G67590   .      +  Araport11  gene    .   \n",
       "\n",
       "                                               details  \n",
       "229  Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...  \n",
       "157  Dbxref=PMID:15642518,PMID:15659355,PMID:155317...  \n",
       "45   Dbxref=PMID:18315867,PMID:18775970,PMID:207364...  \n",
       "181  Dbxref=PMID:15247385,PMID:15911562,PMID:187759...  \n",
       "149  Alias=AtLFG4,LFG4,LIFEGUARD 4;Dbxref=PMID:1708...  \n",
       "..                                                 ...  \n",
       "346  Alias=H3.1,histone 3.1;Dbxref=PMID:12662315,PM...  \n",
       "286  Dbxref=PMID:18775970,locus:2168297;ID=AT5G6547...  \n",
       "360  Dbxref=PMID:18650403,PMID:18775970,PMID:207364...  \n",
       "154  Dbxref=PMID:21251108,PMID:21288268,locus:21555...  \n",
       "210  Dbxref=PMID:12084824,PMID:12837548,PMID:146710...  \n",
       "\n",
       "[406 rows x 10 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_proms_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write out bed file of merged dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(../../data/FIMO/responsivepromoters.bed)>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.from_dataframe(filtered_proms_sorted).saveas('../../data/FIMO/responsivepromoters.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
