{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo_file = \"../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/FIMO/output/promoters_5UTR_FIMO/fimo.tsv\"\n",
    "promoter_bedfile = \"../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/promoters_5UTR.bed\"\n",
    "motifs_bed = \"../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/promoters_5UTR_motifs.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fimo_qfilter(fimo_file, q_value):\n",
    "    \"\"\"this uses a meme-suite version 5 fimo.tsv file, filters by a q-value, and returns a pandas df\"\"\"\n",
    "    fimo = pd.read_table(fimo_file, sep='\\t', engine='python')\n",
    "    #rename sequence column to just the AGI code\n",
    "    fimo.sequence_name = fimo.sequence_name.str.extract(r'(.*?)\\:')\n",
    "    #filter\n",
    "    fimo_qfilter = fimo[fimo['q-value'] <= q_value]    \n",
    "    return fimo_qfilter\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fimo2bed(filtered_fimo_df, promoters_bed, output_bed):\n",
    "    \"\"\"This function creates a bed file using fimo.tsv motif file, and the promoter.bed file (chromosome number is used from this). It sorts the bedfile by chromosome then by start\"\"\"\n",
    "    promoters = pd.read_table(promoters_bed, sep='\\t', header=None) #read in promoter bed file\n",
    "    #add column names\n",
    "    cols = ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details'] \n",
    "    promoters.columns = cols\n",
    "    #merge promoters.bed with the fimo motif file\n",
    "    merged = pd.merge(filtered_fimo_df, promoters, left_on='sequence_name', right_on='gene')\n",
    "    #add motif start position to the promoter start position, and minus 1 (bed file is 0-based), and motif end to promoter start pos (bed file end coord is non-inclusive). Then format with no decimal place\n",
    "    merged['correct_start'] = (merged['start_x'] + merged['start_y'] -1).astype(np.int64)\n",
    "    merged['correct_stop'] = (merged['stop_x'] + merged['start_y']).astype(np.int64)\n",
    "    #create motifs df in bed file column order\n",
    "    motifs_df = merged.loc[:, ['chr','correct_start','correct_stop', 'motif_id', 'score', 'strand_x', 'sequence_name', 'p-value', 'q-value', 'matched_sequence']]\n",
    "    #sort the df by chromosome then by motif start position\n",
    "    motifs_df_sorted = motifs_df.sort_values(['chr','correct_start'])\n",
    "    #create motifs bed file\n",
    "    motifs = BedTool.from_dataframe(motifs_df_sorted).saveas(output_bed)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo = pd.read_table(fimo_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT5G14390::5:4639575-4640835</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.8974</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530::4:15694958-15696283</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.8974</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT5G14390::5:4639575-4640835</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.8974</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530::4:15694958-15696283</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.8974</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530::4:15694958-15696283</td>\n",
       "      <td>6.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.8974</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   motif_id motif_alt_id                   sequence_name  \\\n",
       "0  BBRBPC_tnt.BPC5_col_a_m1         MEME    AT5G14390::5:4639575-4640835   \n",
       "1  BBRBPC_tnt.BPC5_col_a_m1         MEME  AT4G32530::4:15694958-15696283   \n",
       "2  BBRBPC_tnt.BPC5_col_a_m1         MEME    AT5G14390::5:4639575-4640835   \n",
       "3  BBRBPC_tnt.BPC5_col_a_m1         MEME  AT4G32530::4:15694958-15696283   \n",
       "4  BBRBPC_tnt.BPC5_col_a_m1         MEME  AT4G32530::4:15694958-15696283   \n",
       "\n",
       "   start  stop strand    score       p-value       q-value  \\\n",
       "0    1.0  30.0      -  61.8974  9.830000e-20  5.800000e-15   \n",
       "1    2.0  31.0      -  61.8974  9.830000e-20  5.800000e-15   \n",
       "2    3.0  32.0      -  61.8974  9.830000e-20  5.800000e-15   \n",
       "3    4.0  33.0      -  61.8974  9.830000e-20  5.800000e-15   \n",
       "4    6.0  35.0      -  61.8974  9.830000e-20  5.800000e-15   \n",
       "\n",
       "                 matched_sequence  \n",
       "0  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "1  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "2  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "3  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "4  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fimo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            AT5G14390::5:4639575-4640835\n",
       "1          AT4G32530::4:15694958-15696283\n",
       "2            AT5G14390::5:4639575-4640835\n",
       "3          AT4G32530::4:15694958-15696283\n",
       "4          AT4G32530::4:15694958-15696283\n",
       "                        ...              \n",
       "8661525    AT5G57565::5:23309283-23310871\n",
       "8661526      AT1G20690::1:7173437-7175085\n",
       "8661527                               NaN\n",
       "8661528                               NaN\n",
       "8661529                               NaN\n",
       "Name: sequence_name, Length: 8661530, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fimo.sequence_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo.sequence_name = fimo.sequence_name.str.extract(r'(.*?)\\:')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT5G14390</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.89740</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.89740</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT5G14390</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.89740</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.89740</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G32530</td>\n",
       "      <td>6.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>-</td>\n",
       "      <td>61.89740</td>\n",
       "      <td>9.830000e-20</td>\n",
       "      <td>5.800000e-15</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661525</th>\n",
       "      <td>AP2EREBP_tnt.ERF2_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT5G57565</td>\n",
       "      <td>1338.0</td>\n",
       "      <td>1358.0</td>\n",
       "      <td>+</td>\n",
       "      <td>4.77778</td>\n",
       "      <td>1.000000e-04</td>\n",
       "      <td>2.940000e-01</td>\n",
       "      <td>CTGTGGCCAAATCATCCTCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661526</th>\n",
       "      <td>AP2EREBP_tnt.ERF2_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT1G20690</td>\n",
       "      <td>1399.0</td>\n",
       "      <td>1419.0</td>\n",
       "      <td>+</td>\n",
       "      <td>4.77778</td>\n",
       "      <td>1.000000e-04</td>\n",
       "      <td>2.940000e-01</td>\n",
       "      <td>CGGCTCATGCTCCGGCCTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661527</th>\n",
       "      <td># FIMO (Find Individual Motif Occurrences): Ve...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661528</th>\n",
       "      <td># The format of this file is described at http...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661529</th>\n",
       "      <td># fimo --bfile ../../data/output/non-overlappi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8661530 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  motif_id motif_alt_id  \\\n",
       "0                                 BBRBPC_tnt.BPC5_col_a_m1         MEME   \n",
       "1                                 BBRBPC_tnt.BPC5_col_a_m1         MEME   \n",
       "2                                 BBRBPC_tnt.BPC5_col_a_m1         MEME   \n",
       "3                                 BBRBPC_tnt.BPC5_col_a_m1         MEME   \n",
       "4                                 BBRBPC_tnt.BPC5_col_a_m1         MEME   \n",
       "...                                                    ...          ...   \n",
       "8661525                      AP2EREBP_tnt.ERF2_colamp_a_m1         MEME   \n",
       "8661526                      AP2EREBP_tnt.ERF2_colamp_a_m1         MEME   \n",
       "8661527  # FIMO (Find Individual Motif Occurrences): Ve...          NaN   \n",
       "8661528  # The format of this file is described at http...          NaN   \n",
       "8661529  # fimo --bfile ../../data/output/non-overlappi...          NaN   \n",
       "\n",
       "        sequence_name   start    stop strand     score       p-value  \\\n",
       "0           AT5G14390     1.0    30.0      -  61.89740  9.830000e-20   \n",
       "1           AT4G32530     2.0    31.0      -  61.89740  9.830000e-20   \n",
       "2           AT5G14390     3.0    32.0      -  61.89740  9.830000e-20   \n",
       "3           AT4G32530     4.0    33.0      -  61.89740  9.830000e-20   \n",
       "4           AT4G32530     6.0    35.0      -  61.89740  9.830000e-20   \n",
       "...               ...     ...     ...    ...       ...           ...   \n",
       "8661525     AT5G57565  1338.0  1358.0      +   4.77778  1.000000e-04   \n",
       "8661526     AT1G20690  1399.0  1419.0      +   4.77778  1.000000e-04   \n",
       "8661527           NaN     NaN     NaN    NaN       NaN           NaN   \n",
       "8661528           NaN     NaN     NaN    NaN       NaN           NaN   \n",
       "8661529           NaN     NaN     NaN    NaN       NaN           NaN   \n",
       "\n",
       "              q-value                matched_sequence  \n",
       "0        5.800000e-15  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "1        5.800000e-15  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "2        5.800000e-15  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "3        5.800000e-15  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "4        5.800000e-15  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "...               ...                             ...  \n",
       "8661525  2.940000e-01           CTGTGGCCAAATCATCCTCCA  \n",
       "8661526  2.940000e-01           CGGCTCATGCTCCGGCCTTCT  \n",
       "8661527           NaN                             NaN  \n",
       "8661528           NaN                             NaN  \n",
       "8661529           NaN                             NaN  \n",
       "\n",
       "[8661530 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fimo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10803002"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fimo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT1G33350</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT3G30580</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT3G47600</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT3G55960</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   motif_id motif_alt_id sequence_name  start  stop strand  \\\n",
       "0  BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760    1.0  30.0      +   \n",
       "1  BBRBPC_tnt.BPC5_col_a_m1         MEME     AT1G33350    2.0  31.0      +   \n",
       "2  BBRBPC_tnt.BPC5_col_a_m1         MEME     AT3G30580    2.0  31.0      +   \n",
       "3  BBRBPC_tnt.BPC5_col_a_m1         MEME     AT3G47600    2.0  31.0      +   \n",
       "4  BBRBPC_tnt.BPC5_col_a_m1         MEME     AT3G55960    2.0  31.0      +   \n",
       "\n",
       "     score       p-value       q-value                matched_sequence  \n",
       "0  61.3544  1.460000e-19  1.350000e-14  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "1  61.3544  1.460000e-19  1.350000e-14  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "2  61.3544  1.460000e-19  1.350000e-14  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "3  61.3544  1.460000e-19  1.350000e-14  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  \n",
       "4  61.3544  1.460000e-19  1.350000e-14  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter qvalues below\n",
    "fimo_qfilter = fimo[fimo['q-value'] <= 0.05]\n",
    "fimo_qfilter.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1346100"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fimo_qfilter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2630</td>\n",
       "      <td>3630</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ANAC001,NAC domain containing protein 1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>9130</td>\n",
       "      <td>10130</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10100</td>\n",
       "      <td>11100</td>\n",
       "      <td>AT1G03987</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G03987;Name=AT1G03987;locus_type=long_no...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>13714</td>\n",
       "      <td>14714</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:11118137,PMID:15010618,PMID:166036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>22120</td>\n",
       "      <td>23120</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr  start   stop       gene dot strand     source  type dot2  \\\n",
       "0   1   2630   3630  AT1G01010   .      +  Araport11  gene    .   \n",
       "1   1   9130  10130  AT1G01020   .      -  Araport11  gene    .   \n",
       "2   1  10100  11100  AT1G03987   .      +  Araport11  gene    .   \n",
       "3   1  13714  14714  AT1G01030   .      -  Araport11  gene    .   \n",
       "4   1  22120  23120  AT1G01040   .      +  Araport11  gene    .   \n",
       "\n",
       "                                             details  \n",
       "0  Alias=ANAC001,NAC domain containing protein 1;...  \n",
       "1  Dbxref=PMID:15010618,locus:2200940;ID=AT1G0102...  \n",
       "2  ID=AT1G03987;Name=AT1G03987;locus_type=long_no...  \n",
       "3  Dbxref=PMID:11118137,PMID:15010618,PMID:166036...  \n",
       "4  Alias=ASU1,ABNORMAL SUSPENSOR 1,ATDCL1,DICER-L...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#need to read in the promoter gff file to find chromosome no. of each promoter\n",
    "promoters = pd.read_table(promoter_bedfile, sep='\\t', header=None)\n",
    "cols = ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']\n",
    "promoters.columns = cols\n",
    "promoters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start_x</th>\n",
       "      <th>stop_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_y</th>\n",
       "      <th>stop_y</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>18096218</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>18096218</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>2.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>+</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>4.370000e-19</td>\n",
       "      <td>3.860000e-14</td>\n",
       "      <td>GAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>18096218</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>4.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>+</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>4.370000e-19</td>\n",
       "      <td>3.860000e-14</td>\n",
       "      <td>GAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>18096218</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>5.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>+</td>\n",
       "      <td>55.4430</td>\n",
       "      <td>2.630000e-18</td>\n",
       "      <td>2.310000e-13</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAT</td>\n",
       "      <td>4</td>\n",
       "      <td>18096218</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      motif_id motif_alt_id sequence_name  start_x  stop_x  \\\n",
       "0     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      1.0    30.0   \n",
       "1     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      3.0    32.0   \n",
       "2  BBRBPC_tnt.BPC5_colamp_a_m1         MEME     AT4G38760      2.0    30.0   \n",
       "3  BBRBPC_tnt.BPC5_colamp_a_m1         MEME     AT4G38760      4.0    32.0   \n",
       "4     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      5.0    34.0   \n",
       "\n",
       "  strand_x    score       p-value       q-value  \\\n",
       "0        +  61.3544  1.460000e-19  1.350000e-14   \n",
       "1        +  61.3544  1.460000e-19  1.350000e-14   \n",
       "2        +  49.1860  4.370000e-19  3.860000e-14   \n",
       "3        +  49.1860  4.370000e-19  3.860000e-14   \n",
       "4        +  55.4430  2.630000e-18  2.310000e-13   \n",
       "\n",
       "                 matched_sequence chr   start_y    stop_y       gene dot  \\\n",
       "0  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG   4  18096218  18097218  AT4G38760   .   \n",
       "1  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG   4  18096218  18097218  AT4G38760   .   \n",
       "2   GAGAGAGAGAGAGAGAGAGAGAGAGAGAG   4  18096218  18097218  AT4G38760   .   \n",
       "3   GAGAGAGAGAGAGAGAGAGAGAGAGAGAG   4  18096218  18097218  AT4G38760   .   \n",
       "4  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAT   4  18096218  18097218  AT4G38760   .   \n",
       "\n",
       "  strand_y     source  type dot2  \\\n",
       "0        -  Araport11  gene    .   \n",
       "1        -  Araport11  gene    .   \n",
       "2        -  Araport11  gene    .   \n",
       "3        -  Araport11  gene    .   \n",
       "4        -  Araport11  gene    .   \n",
       "\n",
       "                                             details  \n",
       "0  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...  \n",
       "1  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...  \n",
       "2  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...  \n",
       "3  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...  \n",
       "4  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(fimo_qfilter, promoters, left_on='sequence_name', right_on='gene')\n",
    "merged_outer = pd.merge(fimo_qfilter, promoters, left_on='sequence_name', right_on='gene', how='outer')\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1346100"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1346710"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#these promoters had no motifs found in them - 610 promoters\n",
    "no_motifs = merged[merged['sequence_name'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start_x</th>\n",
       "      <th>stop_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_y</th>\n",
       "      <th>stop_y</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1346100</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>201102</td>\n",
       "      <td>202102</td>\n",
       "      <td>AT1G01560</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATMPK11,MAP kinase 11;Dbxref=PMID:115441...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346101</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1914895</td>\n",
       "      <td>1915895</td>\n",
       "      <td>AT1G06250</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15130548,PMID:18433503,PMID:235053...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346102</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1945745</td>\n",
       "      <td>1946745</td>\n",
       "      <td>AT1G06390</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATGSK1,ATSK2-3,A. THALIANA SHAGGY-LIKE K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346103</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2011109</td>\n",
       "      <td>2012109</td>\n",
       "      <td>AT1G04483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G04483;Name=AT1G04483;Note=Natural antis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346104</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2167560</td>\n",
       "      <td>2168560</td>\n",
       "      <td>AT1G07070</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15310832,PMID:16297073,PMID:179342...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        motif_id motif_alt_id sequence_name  start_x  stop_x strand_x  score  \\\n",
       "1346100      NaN          NaN           NaN      NaN     NaN      NaN    NaN   \n",
       "1346101      NaN          NaN           NaN      NaN     NaN      NaN    NaN   \n",
       "1346102      NaN          NaN           NaN      NaN     NaN      NaN    NaN   \n",
       "1346103      NaN          NaN           NaN      NaN     NaN      NaN    NaN   \n",
       "1346104      NaN          NaN           NaN      NaN     NaN      NaN    NaN   \n",
       "\n",
       "         p-value  q-value matched_sequence chr  start_y   stop_y       gene  \\\n",
       "1346100      NaN      NaN              NaN   1   201102   202102  AT1G01560   \n",
       "1346101      NaN      NaN              NaN   1  1914895  1915895  AT1G06250   \n",
       "1346102      NaN      NaN              NaN   1  1945745  1946745  AT1G06390   \n",
       "1346103      NaN      NaN              NaN   1  2011109  2012109  AT1G04483   \n",
       "1346104      NaN      NaN              NaN   1  2167560  2168560  AT1G07070   \n",
       "\n",
       "        dot strand_y     source  type dot2  \\\n",
       "1346100   .        +  Araport11  gene    .   \n",
       "1346101   .        -  Araport11  gene    .   \n",
       "1346102   .        +  Araport11  gene    .   \n",
       "1346103   .        -  Araport11  gene    .   \n",
       "1346104   .        +  Araport11  gene    .   \n",
       "\n",
       "                                                   details  \n",
       "1346100  Alias=ATMPK11,MAP kinase 11;Dbxref=PMID:115441...  \n",
       "1346101  Dbxref=PMID:15130548,PMID:18433503,PMID:235053...  \n",
       "1346102  Alias=ATGSK1,ATSK2-3,A. THALIANA SHAGGY-LIKE K...  \n",
       "1346103  ID=AT1G04483;Name=AT1G04483;Note=Natural antis...  \n",
       "1346104  Dbxref=PMID:15310832,PMID:16297073,PMID:179342...  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_motifs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#not all promoters had motifs found in them!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start_x</th>\n",
       "      <th>stop_x</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>stop_y</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "      <th>correct_start</th>\n",
       "      <th>correct_stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>...</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "      <td>18096218.0</td>\n",
       "      <td>18096247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>+</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>1.460000e-19</td>\n",
       "      <td>1.350000e-14</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>...</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "      <td>18096220.0</td>\n",
       "      <td>18096249.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>2.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>+</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>4.370000e-19</td>\n",
       "      <td>3.860000e-14</td>\n",
       "      <td>GAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>...</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "      <td>18096219.0</td>\n",
       "      <td>18096247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>4.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>+</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>4.370000e-19</td>\n",
       "      <td>3.860000e-14</td>\n",
       "      <td>GAGAGAGAGAGAGAGAGAGAGAGAGAGAG</td>\n",
       "      <td>...</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "      <td>18096221.0</td>\n",
       "      <td>18096249.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>MEME</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>5.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>+</td>\n",
       "      <td>55.4430</td>\n",
       "      <td>2.630000e-18</td>\n",
       "      <td>2.310000e-13</td>\n",
       "      <td>AGAGAGAGAGAGAGAGAGAGAGAGAGAGAT</td>\n",
       "      <td>...</td>\n",
       "      <td>18097218</td>\n",
       "      <td>AT4G38760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...</td>\n",
       "      <td>18096222.0</td>\n",
       "      <td>18096251.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      motif_id motif_alt_id sequence_name  start_x  stop_x  \\\n",
       "0     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      1.0    30.0   \n",
       "1     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      3.0    32.0   \n",
       "2  BBRBPC_tnt.BPC5_colamp_a_m1         MEME     AT4G38760      2.0    30.0   \n",
       "3  BBRBPC_tnt.BPC5_colamp_a_m1         MEME     AT4G38760      4.0    32.0   \n",
       "4     BBRBPC_tnt.BPC5_col_a_m1         MEME     AT4G38760      5.0    34.0   \n",
       "\n",
       "  strand_x    score       p-value       q-value  \\\n",
       "0        +  61.3544  1.460000e-19  1.350000e-14   \n",
       "1        +  61.3544  1.460000e-19  1.350000e-14   \n",
       "2        +  49.1860  4.370000e-19  3.860000e-14   \n",
       "3        +  49.1860  4.370000e-19  3.860000e-14   \n",
       "4        +  55.4430  2.630000e-18  2.310000e-13   \n",
       "\n",
       "                 matched_sequence  ...    stop_y       gene  dot strand_y  \\\n",
       "0  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  ...  18097218  AT4G38760    .        -   \n",
       "1  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAG  ...  18097218  AT4G38760    .        -   \n",
       "2   GAGAGAGAGAGAGAGAGAGAGAGAGAGAG  ...  18097218  AT4G38760    .        -   \n",
       "3   GAGAGAGAGAGAGAGAGAGAGAGAGAGAG  ...  18097218  AT4G38760    .        -   \n",
       "4  AGAGAGAGAGAGAGAGAGAGAGAGAGAGAT  ...  18097218  AT4G38760    .        -   \n",
       "\n",
       "      source  type dot2                                            details  \\\n",
       "0  Araport11  gene    .  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...   \n",
       "1  Araport11  gene    .  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...   \n",
       "2  Araport11  gene    .  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...   \n",
       "3  Araport11  gene    .  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...   \n",
       "4  Araport11  gene    .  Dbxref=locus:2141767;ID=AT4G38760;Name=AT4G387...   \n",
       "\n",
       "  correct_start correct_stop  \n",
       "0    18096218.0   18096247.0  \n",
       "1    18096220.0   18096249.0  \n",
       "2    18096219.0   18096247.0  \n",
       "3    18096221.0   18096249.0  \n",
       "4    18096222.0   18096251.0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#need to add motif start to the promoter start -1, and motif end to promoter start -1\n",
    "merged['correct_start'] = merged['start_x'] + merged['start_y'] -1\n",
    "merged['correct_stop'] = merged['stop_x'] + merged['start_y'] -1\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>correct_start</th>\n",
       "      <th>correct_stop</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand_x</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>18096218.0</td>\n",
       "      <td>18096247.0</td>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>18096220.0</td>\n",
       "      <td>18096249.0</td>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>61.3544</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>18096219.0</td>\n",
       "      <td>18096247.0</td>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>18096221.0</td>\n",
       "      <td>18096249.0</td>\n",
       "      <td>BBRBPC_tnt.BPC5_colamp_a_m1</td>\n",
       "      <td>49.1860</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>18096222.0</td>\n",
       "      <td>18096251.0</td>\n",
       "      <td>BBRBPC_tnt.BPC5_col_a_m1</td>\n",
       "      <td>55.4430</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr  correct_start  correct_stop                     motif_id    score  \\\n",
       "0   4     18096218.0    18096247.0     BBRBPC_tnt.BPC5_col_a_m1  61.3544   \n",
       "1   4     18096220.0    18096249.0     BBRBPC_tnt.BPC5_col_a_m1  61.3544   \n",
       "2   4     18096219.0    18096247.0  BBRBPC_tnt.BPC5_colamp_a_m1  49.1860   \n",
       "3   4     18096221.0    18096249.0  BBRBPC_tnt.BPC5_colamp_a_m1  49.1860   \n",
       "4   4     18096222.0    18096251.0     BBRBPC_tnt.BPC5_col_a_m1  55.4430   \n",
       "\n",
       "  strand_x  \n",
       "0        +  \n",
       "1        +  \n",
       "2        +  \n",
       "3        +  \n",
       "4        +  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create motifs df\n",
    "motifs_df = merged.loc[:, ['chr','correct_start','correct_stop', 'motif_id', 'score', 'strand_x',]]\n",
    "motifs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort by chromosome then by start position\n",
    "motifs_df_sorted = motifs_df.sort_values(['chr','correct_start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create motifs bed file\n",
    "motifs = BedTool.from_dataframe(motifs_df_sorted).saveas('../../data/FIMO/motifs.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "PermissionError",
     "evalue": "[Errno 13] Permission denied",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPermissionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-6e225a2f9c19>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#same again using the functions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mfiltered_fimo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfimo_qfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfimo_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0.05\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mfimo2bed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_fimo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpromoter_bedfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmotifs_bed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-fed0dfa3953f>\u001b[0m in \u001b[0;36mfimo_qfilter\u001b[0;34m(fimo_file, q_value)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mfimo_qfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfimo_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mq_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;34m\"\"\"this uses a meme-suite version 5 fimo.tsv file, filters by a q-value, and returns a pandas df\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mfimo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfimo_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'python'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mfimo_qfilter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfimo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mfimo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'q-value'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mq_value\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mfimo_qfilter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1145\u001b[0m                     \u001b[0;34m' \"python-fwf\")'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1146\u001b[0m                 )\n\u001b[0;32m-> 1147\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1148\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1149\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_failover_to_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, **kwds)\u001b[0m\n\u001b[1;32m   2308\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnum_original_columns\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2309\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2310\u001b[0;31m         ) = self._infer_columns()\n\u001b[0m\u001b[1;32m   2311\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2312\u001b[0m         \u001b[0;31m# Now self.columns has the set of columns that we will process.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_infer_columns\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2591\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhr\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mheader\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2592\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2593\u001b[0;31m                     \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_buffered_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2594\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2595\u001b[0m                     \u001b[0;32mwhile\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mline_pos\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mhr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_buffered_line\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2774\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2775\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2776\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_next_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2777\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2778\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_check_for_bom\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfirst_row\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_next_line\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2871\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2872\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2873\u001b[0;31m                 \u001b[0morig_line\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_next_iter_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow_num\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpos\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2874\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpos\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2875\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/PromoterArchitecturePipeline/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_next_iter_line\u001b[0;34m(self, row_num)\u001b[0m\n\u001b[1;32m   2931\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2932\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2933\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2934\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mcsv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2935\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarn_bad_lines\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_bad_lines\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mPermissionError\u001b[0m: [Errno 13] Permission denied"
     ]
    }
   ],
   "source": [
    "#same again using the functions\n",
    "filtered_fimo = fimo_qfilter(fimo_file, 0.05)\n",
    "fimo2bed(filtered_fimo, promoter_bedfile, motifs_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fimo_file = \"../../data/FIMO/output/responsivepromoters_FIMO/fimo.tsv\"\n",
    "promoter_bedfile = \"../../data/FIMO/responsivepromoters.bed\"\n",
    "motifs_bed = \"../../data/FIMO/responsivepromoters_motifs.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
