{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare files for TFBS gat enrichment test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool\n",
    "import os\n",
    "import io\n",
    "#import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parser = argparse.ArgumentParser(description='prepare_TFBS_enrichment')\n",
    "# parser.add_argument('file_names', type=str, help='Name of folder and filenames for the promoters extracted')\n",
    "# parser.add_argument('promoterpref', type=str, help='Promoter prefix name')\n",
    "# parser.add_argument('Czechowski_gene_categories', type=str, help='Input location of Czechowski gene categories text file')\n",
    "# parser.add_argument('promoter_bed_file', type=str, help='Input location of promoters bed file')\n",
    "# parser.add_argument('output_genecat_prefix', type=str, help='Gene category prefix (eg. Czechowski)')\n",
    "# parser.add_argument('mapped_motifs', type=str, help='input location of mapped_motifs bed file')#\n",
    "# parser.add_argument('mapped_motifs_out', type=str, help='output location of mapped_motifs bed file with TF family in column 4')\n",
    "# args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names = 'non-overlapping_includingbidirectional_all_genes_newannotation'\n",
    "promoterpref='promoters_5UTR'\n",
    "Czechowski_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_czechowski_constitutive_variable_random.txt'\n",
    "promoter_bed_file = f'../../data/output/{file_names}/FIMO/promoters_5UTR.bed'\n",
    "output_genecat_prefix = 'Czechowski'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/TFBS_enrichment  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the plots to be exported to\n",
    "dirName = f'../../data/output/{file_names}/TFBS_enrichment'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/TFBS_enrichment/plots  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the plots to be exported to\n",
    "dirName = f'../../data/output/{file_names}/TFBS_enrichment/plots'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/TFBS_enrichment/gat_analysis  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the plots to be exported to\n",
    "dirName = f'../../data/output/{file_names}/TFBS_enrichment/gat_analysis'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/TFBS_enrichment/other_analysis  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the plots to be exported to\n",
    "dirName = f'../../data/output/{file_names}/TFBS_enrichment/other_analysis'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_files(promoter_bed_file,mapped_motifs_file,gene_categories):\n",
    "    \"\"\"sort the promoters, motifs and genes types data ready for gat TATA enrichment\"\"\"\n",
    "    promoters = pd.read_table(promoter_bed_file, sep='\\t', header=None)\n",
    "    col = ['chr', 'start','stop','promoter_AGI','dot1','strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "    #read in gene categories\n",
    "    gene_cats = pd.read_table(gene_categories, sep='\\t', header=None)\n",
    "    cols = ['promoter_AGI','gene_type']\n",
    "    gene_cats.columns = cols\n",
    "    #merge with promoters\n",
    "    promoters = pd.merge(gene_cats,promoters, on='promoter_AGI', how='left')\n",
    "    #read in mapped_motifs file\n",
    "\n",
    "    #select the correct columns depending on the number of columns in the file\n",
    "    if len(motifs.columns) == 24:\n",
    "        cols = ['chr', 'start', 'stop', 'promoter_AGI','dot1','strand','source','type','dot2','attributes',\n",
    "            'motif_chr','motif_start','motif_stop','name_rep', 'score', 'motif_strand',\n",
    "            'promoter_AGI2', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI','bp_overlap']\n",
    "        motifs.columns = cols\n",
    "        #filter columns\n",
    "        motifs = motifs[['motif_chr','motif_start','motif_stop','name_rep', 'score', 'motif_strand',\n",
    "             'promoter_AGI2', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI']]\n",
    "        #rename columns\n",
    "        cols = ['chr', 'start', 'stop', 'name_rep', 'score', 'strand', 'promoter_AGI', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI']\n",
    "        motifs.columns = cols\n",
    "        \n",
    "    elif len(motifs.columns) == 13:\n",
    "        cols = ['chr', 'start', 'stop', 'name_rep', 'score', 'strand', 'promoter_AGI', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI']\n",
    "        motifs.columns = cols\n",
    "        \n",
    "    elif len(motifs.columns) == 17:\n",
    "        cols = ['chr', 'start', 'stop', 'name_rep', 'score', 'strand', 'promoter_AGI', 'p-value', \n",
    "             'q-value', 'matched_sequence', 'TF_name','TF_family','TF_AGI','chr_openchrom','start_openchrom','stop_openchrom','bp_overlap' ]\n",
    "        motifs.columns = cols\n",
    "    \n",
    "    #swap columns around so TF family is in column 4\n",
    "    motifs = motifs[['chr', 'start', 'stop', 'TF_family', 'score', 'strand', 'promoter_AGI', 'p-value', \n",
    "             'q-value', 'matched_sequence', 'TF_name','name_rep','TF_AGI']]\n",
    "    #write out the file\n",
    "    bed = BedTool.from_dataframe(motifs).saveas(mapped_motifs_out)\n",
    "    \n",
    "    return promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_gat(df):\n",
    "    \"\"\"prepare files for running gat analysis - outputs a workspace file containing all promoters, a variable promoter file and a constitutive promoter file\"\"\"\n",
    "    #make buffer to save promoters\n",
    "    buffer = io.StringIO()\n",
    "    df.to_csv(buffer,sep='\\t', header=None, index = False)\n",
    "    buffer.seek(0)\n",
    "    #select only constitutive and variable genes\n",
    "    #df = df[(df.gene_type == 'constitutive') | (df.gene_type == 'variable')]\n",
    "    #reorder columns\n",
    "    #with gene_type in fourth column for gat enrichment analysis\n",
    "    df_reordered_gene_type = df[['chr','start','stop','gene_type', 'strand', 'source', 'attributes','promoter_AGI']]\n",
    "    df_reordered_AGI = df[['chr','start','stop','promoter_AGI', 'strand', 'source', 'attributes','gene_type']]\n",
    "    #with promoter_AGI in fourth column for other enrichment analysis such as CiiiDER\n",
    "    #sort by chromosome and start\n",
    "    sorted_motifs_gene_type = df_reordered_gene_type.sort_values(['chr','start'])\n",
    "    sorted_motifs_AGI = df_reordered_AGI.sort_values(['chr','start'])\n",
    "    #save bed file with all gene types\n",
    "    bed = BedTool.from_dataframe(sorted_motifs_gene_type).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/{args.output_genecat_prefix}_{args.promoterpref}_allgenetypes_gat.bed')\n",
    "    bed = BedTool.from_dataframe(sorted_motifs_gene_type).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/{args.output_genecat_prefix}_{args.promoterpref}_allgenetypes_AGI.bed')\n",
    "    \n",
    "    #save bed file with only constitutive and variable genetypes\n",
    "    sorted_motifs_gene_type2 = sorted_motifs_gene_type[(sorted_motifs_gene_type.gene_type == 'constitutive') | (sorted_motifs_gene_type.gene_type == 'variable')]\n",
    "    sorted_motifs_AGI2 = sorted_motifs_AGI[(sorted_motifs_AGI.gene_type == 'constitutive') | (sorted_motifs_AGI.gene_type == 'variable')]\n",
    "    bed = BedTool.from_dataframe(sorted_motifs_gene_type2).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/{args.output_genecat_prefix}_{args.promoterpref}_constitutivevariable_gat.bed')\n",
    "    bed = BedTool.from_dataframe(sorted_motifs_AGI2).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/{args.output_genecat_prefix}_{args.promoterpref}_constitutivevariable_AGI.bed')\n",
    "    #make a new gat workspace file with all promoters (first 3 columns)\n",
    "    bed = BedTool.from_dataframe(sorted_motifs_gene_type[['chr','start','stop']]).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/gat_analysis/{args.output_genecat_prefix}_{args.promoterpref}_workspace.bed')\n",
    "    #select only variable promoters\n",
    "    variable_promoters_extended = sorted_motifs_gene_type[sorted_motifs_gene_type['gene_type'] == 'variable']\n",
    "    sorted_variable = variable_promoters_extended.sort_values(['chr','start'])\n",
    "    bed = BedTool.from_dataframe(sorted_variable).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/gat_analysis/{args.output_genecat_prefix}_{args.promoterpref}_variable_gat.bed')\n",
    "    #same again for other enrichment analyses\n",
    "    variable_promoters_extended_agi = sorted_motifs_AGI[sorted_motifs_AGI['gene_type'] == 'variable']\n",
    "    sorted_variable_AGI = variable_promoters_extended_agi.sort_values(['chr','start'])\n",
    "    bed = BedTool.from_dataframe(sorted_variable_AGI).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/other_analysis/{args.output_genecat_prefix}_{args.promoterpref}_variable_AGI.bed')\n",
    "    #make a constitutive only file\n",
    "    constitutive_promoters = sorted_motifs_gene_type[sorted_motifs_gene_type['gene_type'] == 'constitutive']\n",
    "    sorted_constitutive = constitutive_promoters.sort_values(['chr','start'])\n",
    "    bed = BedTool.from_dataframe(sorted_constitutive).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/gat_analysis/{args.output_genecat_prefix}_{args.promoterpref}_constitutive_gat.bed')\n",
    "    #same again for other enrichment analyses with AGI in fourth column\n",
    "    constitutive_promoters_AGI = sorted_motifs_AGI[sorted_motifs_AGI['gene_type'] == 'constitutive']\n",
    "    sorted_constitutive_AGI = constitutive_promoters_AGI.sort_values(['chr','start'])\n",
    "    bed = BedTool.from_dataframe(sorted_constitutive_AGI).saveas(f'../../data/output/{args.file_names}/TFBS_enrichment/other_analysis/{args.output_genecat_prefix}_{args.promoterpref}_constitutive_AGI.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "promoters = sort_files(promoter_bed_file,Czechowski_gene_categories)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>attributes</th>\n",
       "      <th>promoter_AGI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>1</td>\n",
       "      <td>344515</td>\n",
       "      <td>346051</td>\n",
       "      <td>variable</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT1G02000;Name=GAE2;biotype=protein_co...</td>\n",
       "      <td>AT1G02000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>792838</td>\n",
       "      <td>793314</td>\n",
       "      <td>variable</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT1G03240;biotype=protein_coding;descr...</td>\n",
       "      <td>AT1G03240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>1</td>\n",
       "      <td>864019</td>\n",
       "      <td>865372</td>\n",
       "      <td>variable</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT1G03457;biotype=protein_coding;descr...</td>\n",
       "      <td>AT1G03457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>1</td>\n",
       "      <td>1000629</td>\n",
       "      <td>1001472</td>\n",
       "      <td>variable</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT1G03920;biotype=protein_coding;descr...</td>\n",
       "      <td>AT1G03920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>1</td>\n",
       "      <td>1315823</td>\n",
       "      <td>1316918</td>\n",
       "      <td>variable</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT1G04700;biotype=protein_coding;descr...</td>\n",
       "      <td>AT1G04700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>5</td>\n",
       "      <td>22956628</td>\n",
       "      <td>22957985</td>\n",
       "      <td>variable</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT5G56750;Name=NDL1;biotype=protein_co...</td>\n",
       "      <td>AT5G56750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>5</td>\n",
       "      <td>23792382</td>\n",
       "      <td>23793238</td>\n",
       "      <td>constitutive</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT5G58920;biotype=protein_coding;descr...</td>\n",
       "      <td>AT5G58920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>5</td>\n",
       "      <td>24110303</td>\n",
       "      <td>24111442</td>\n",
       "      <td>variable</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT5G59845;Name=GASA10;biotype=protein_...</td>\n",
       "      <td>AT5G59845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>5</td>\n",
       "      <td>24326665</td>\n",
       "      <td>24327827</td>\n",
       "      <td>variable</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT5G60490;Name=FLA12;biotype=protein_c...</td>\n",
       "      <td>AT5G60490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>5</td>\n",
       "      <td>25909881</td>\n",
       "      <td>25910835</td>\n",
       "      <td>constitutive</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>ID=gene:AT5G64813;Name=LIP1;biotype=protein_co...</td>\n",
       "      <td>AT5G64813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr     start      stop     gene_type strand     source  \\\n",
       "186    1    344515    346051      variable      +  araport11   \n",
       "117    1    792838    793314      variable      -  araport11   \n",
       "143    1    864019    865372      variable      -  araport11   \n",
       "187    1   1000629   1001472      variable      +  araport11   \n",
       "193    1   1315823   1316918      variable      +  araport11   \n",
       "..   ...       ...       ...           ...    ...        ...   \n",
       "158    5  22956628  22957985      variable      +  araport11   \n",
       "80     5  23792382  23793238  constitutive      +  araport11   \n",
       "120    5  24110303  24111442      variable      +  araport11   \n",
       "182    5  24326665  24327827      variable      -  araport11   \n",
       "37     5  25909881  25910835  constitutive      +  araport11   \n",
       "\n",
       "                                            attributes promoter_AGI  \n",
       "186  ID=gene:AT1G02000;Name=GAE2;biotype=protein_co...    AT1G02000  \n",
       "117  ID=gene:AT1G03240;biotype=protein_coding;descr...    AT1G03240  \n",
       "143  ID=gene:AT1G03457;biotype=protein_coding;descr...    AT1G03457  \n",
       "187  ID=gene:AT1G03920;biotype=protein_coding;descr...    AT1G03920  \n",
       "193  ID=gene:AT1G04700;biotype=protein_coding;descr...    AT1G04700  \n",
       "..                                                 ...          ...  \n",
       "158  ID=gene:AT5G56750;Name=NDL1;biotype=protein_co...    AT5G56750  \n",
       "80   ID=gene:AT5G58920;biotype=protein_coding;descr...    AT5G58920  \n",
       "120  ID=gene:AT5G59845;Name=GASA10;biotype=protein_...    AT5G59845  \n",
       "182  ID=gene:AT5G60490;Name=FLA12;biotype=protein_c...    AT5G60490  \n",
       "37   ID=gene:AT5G64813;Name=LIP1;biotype=protein_co...    AT5G64813  \n",
       "\n",
       "[200 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create gat files\n",
    "prepare_gat(promoters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
