{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ast import literal_eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_ARF9_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_ARF9_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_ARF9_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_ARF18_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_filtered.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_flattened_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_filtered/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_DREB26_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_DREB26_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_DREB26_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_NLP7_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_NLP7_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_NLP7_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in files\n",
    "def read_files(genotyped_file,genotyped_only_mutated_file,genotyped_only_mutated_flattened_file):\n",
    "    \"\"\"read in the files to pandas\"\"\"\n",
    "    #read in files\n",
    "    genotyped_df = pd.read_table(genotyped_file,sep='\\t', header=0)\n",
    "    genotyped_only_mutated_df = pd.read_table(genotyped_only_mutated_file,sep='\\t', header=0)\n",
    "    genotyped_only_mutated_flattened_df = pd.read_table(genotyped_only_mutated_flattened_file,sep='\\t', header=0)\n",
    "    return genotyped_df,genotyped_only_mutated_df,genotyped_only_mutated_flattened_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_number_mutations_overlapping_at_least_one_TFBS(df, mutation_type):\n",
    "    \"\"\"function to count the number of mutations in the df which overlap at least one TFBS\"\"\"\n",
    "    #turn df into a dict\n",
    "    df_dict = df.to_dict(orient=\"records\")\n",
    "    mutations_overlapping_at_least_one_TFBS_count = 0\n",
    "    if mutation_type == 'insertion':\n",
    "        for row_dict in df_dict:\n",
    "                list_lists_of_overlapping_AGIs = []\n",
    "                genomic_positions = literal_eval(row_dict[f'{mutation_type}_genomic_positions_combined'])\n",
    "                #iterate over insertions. Some insertions are linger than 1 bp so the genomic position pairs need combining\n",
    "                #remaining pairs\n",
    "                remaining_pairs = []\n",
    "                for size in literal_eval(row_dict[f'{mutation_type}_sizes']):\n",
    "                                \n",
    "                    size = int(size)\n",
    "                    #select pairs up to the size of the insertion\n",
    "                    \n",
    "                    #pop out the first n pairs from the genomic positions list, where n is the size of the insertion - 1\n",
    "                    #if remaining pairs not empty, start with that\n",
    "                    if len(remaining_pairs)>0:\n",
    "                        #print('not []')\n",
    "                        current_pairs = remaining_pairs[:size]\n",
    "                    #else create the remaining pairs variable\n",
    "                    elif len(remaining_pairs)==0:\n",
    "                        #print('is []')\n",
    "                        current_pairs = genomic_positions[:size]\n",
    "                        remaining_pairs = genomic_positions[size:]          \n",
    "                    #print(f'currentpairs={current_pairs}remaining={remaining_pairs}')\n",
    "\n",
    "                    current_pairs_list_of_overlapping = []\n",
    "                    #iterate over pairs in the mutation_genomic_positions_combined list #evaluate the string representing the pair as if they are literals ie. a list\n",
    "                    for pair in current_pairs:\n",
    "                        #print(f'pair={pair}')\n",
    "                        list_of_AGIs = []\n",
    "                        list_of_overlapping_AGIs = []\n",
    "                        #for genomic position in pair \n",
    "                        #print(pair)\n",
    "                        for pos in pair:\n",
    "                            #make the genomic position string\n",
    "                            pos = str(pos)\n",
    "                            #if genomic position is in the mutation_overlapping_TFBS_AGI_combined dictionary\n",
    "                            if pos in literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined']):\n",
    "                                #print(pos)\n",
    "                                #append to list_of_AGIs\n",
    "                                mutation_overlapping_TFBS_AGI_combined = literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined'])\n",
    "                            \n",
    "                                #print(pos)\n",
    "                                AGI = mutation_overlapping_TFBS_AGI_combined[pos]\n",
    "                                list_of_AGIs += AGI\n",
    "                        #if empty list, pass\n",
    "                        \n",
    "                        for elem in list_of_AGIs:\n",
    "                            #print(elem)\n",
    "                            #if TFBS is only overlapping one of the genomic positions in the pair, pass - we only want to include TFBSs that overlapping both positions so that the insertion is totally overlapping\n",
    "                            if list_of_AGIs.count(elem) == 1:\n",
    "                                pass\n",
    "                            elif list_of_AGIs.count(elem) > 1:\n",
    "                                list_of_overlapping_AGIs.append(elem)\n",
    "                        #append to list of lists if lists_of_overlapping_AGIs is not empty\n",
    "                        #remove duplicates from list_of_overlapping_AGIs\n",
    "                        list_of_overlapping_AGIs_set = set(list_of_overlapping_AGIs)\n",
    "                        list_of_overlapping_AGIs = list(list_of_overlapping_AGIs_set)\n",
    "                        if len(list_of_overlapping_AGIs) > 0:\n",
    "                            #print(list_of_overlapping_AGIs)\n",
    "                            #get unique AGIs\n",
    "                            #list_of_overlapping_AGIs_unique\n",
    "                            current_pairs_list_of_overlapping+=list_of_overlapping_AGIs\n",
    "                    #get unique elements\n",
    "                    current_pairs_list_of_overlapping_set = set(current_pairs_list_of_overlapping)\n",
    "                    current_pairs_list_of_overlapping = list(current_pairs_list_of_overlapping_set)\n",
    "                    #if current_pairs_list_of_overlapping list isn't empty, append to list_lists_of_overlapping_AGIs \n",
    "                    if len(current_pairs_list_of_overlapping) > 0:\n",
    "                        #print(current_pairs_list_of_overlapping)\n",
    "                        #add one to the count of mutations which overlap at least one TFBS\n",
    "                        mutations_overlapping_at_least_one_TFBS_count+=1\n",
    "                        #append current_pairs_list_of_overlapping of TFBSs to list_lists_of_overlapping_AGIs\n",
    "                        list_lists_of_overlapping_AGIs.append(current_pairs_list_of_overlapping)\n",
    "    else:\n",
    "        for row_dict in df_dict:\n",
    "            list_lists_of_overlapping_AGIs = []\n",
    "            genomic_positions = literal_eval(row_dict[f'{mutation_type}_genomic_positions_combined'])\n",
    "            #iterate over insertions. Some insertions are linger than 1 bp so the genomic position pairs need combining\n",
    "            #remaining positions\n",
    "            remaining_positions = []\n",
    "            for size in literal_eval(row_dict[f'{mutation_type}_sizes']):\n",
    "                            \n",
    "                size = int(size)\n",
    "                #select positions up to the size of the insertion\n",
    "                \n",
    "                #pop out the first n positions from the genomic positions list, where n is the size of the insertion - 1\n",
    "                #if remaining positions not empty, start with that\n",
    "                if len(remaining_positions)>0:\n",
    "                    current_positions = remaining_positions[:size]\n",
    "                    #else create the remaining pairs variable\n",
    "                elif len(remaining_positions)==0:\n",
    "                    current_positions = genomic_positions[:size]\n",
    "                    remaining_positions = genomic_positions[size:]\n",
    "                list_of_overlapping_AGIs = []\n",
    "                #for genomic position in pair \n",
    "                #print(pair)\n",
    "                current_positions_list_of_overlapping = []\n",
    "                for pos in current_positions:\n",
    "                    #make the genomic position string\n",
    "                    pos = str(pos)\n",
    "                    #if genomic position is in the mutation_overlapping_TFBS_AGI_combined dictionary\n",
    "                    if pos in literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined']):\n",
    "                        #print(pos)\n",
    "                        #append to list_of_overlapping_AGIs\n",
    "                        mutation_overlapping_TFBS_AGI_combined = literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined'])\n",
    "                    \n",
    "                        #print(pos)\n",
    "                        AGI = mutation_overlapping_TFBS_AGI_combined[pos]\n",
    "                        list_of_overlapping_AGIs += AGI                     \n",
    "                        #append to list of lists if lists_of_overlapping_AGIs is not empty\n",
    "                #remove duplicates from list_of_overlapping_AGIs\n",
    "                list_of_overlapping_AGIs_set = set(list_of_overlapping_AGIs)\n",
    "                list_of_overlapping_AGIs = list(list_of_overlapping_AGIs_set)\n",
    "                if len(list_of_overlapping_AGIs) > 0:\n",
    "                    #print(list_of_overlapping_AGIs)\n",
    "                    #get unique AGIs\n",
    "                    #list_of_overlapping_AGIs_unique\n",
    "                    current_positions_list_of_overlapping+=list_of_overlapping_AGIs\n",
    "                #get unique elements\n",
    "                current_positions_list_of_overlapping_set = set(current_positions_list_of_overlapping)\n",
    "                current_positions_list_of_overlapping = list(current_positions_list_of_overlapping_set)\n",
    "                #if current_positions_list_of_overlapping list isn't empty, append to list_lists_of_overlapping_AGIs \n",
    "                if len(current_positions_list_of_overlapping) > 0:\n",
    "                    #add one to the count of mutations which overlap at least one TFBS\n",
    "                    mutations_overlapping_at_least_one_TFBS_count+=1\n",
    "                    #append current_positions_list_of_overlapping of TFBSs to list_lists_of_overlapping_AGIs\n",
    "                    list_lists_of_overlapping_AGIs.append(current_positions_list_of_overlapping)\n",
    "\n",
    "    return mutations_overlapping_at_least_one_TFBS_count\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_mutations(genotyped_df,genotyped_only_mutated_df,genotyped_only_mutated_flattened_df):\n",
    "    \"\"\"function to count number of genotyped plant lines and also number of plant lines containing mutations, and then number of mutations overall\"\"\"\n",
    "    #count number of unique plant IDs\n",
    "    genotyped_counts = genotyped_df.plant_ID.unique()\n",
    "    genotyped_only_mutated_counts = genotyped_only_mutated_df.plant_ID.unique()\n",
    "    #print counts\n",
    "    #total mutations\n",
    "    #total_mutations = genotyped_only_mutated_flattened_df\n",
    "    print(f'total plant lines sequenced={len(genotyped_counts)} \\ntotal mutated plant lines={len(genotyped_only_mutated_counts)}')#\\ntotal mutations={len(genotyped_only_mutated_df)}\n",
    "    #count number of chimeric plant lines\n",
    "    # unique_line_mutated = genotyped_only_mutated_df.drop_duplicates(subset='plant_ID',keep='first')\n",
    "    \n",
    "    unique_line_mutated = genotyped_only_mutated_flattened_df\n",
    "    #filter so only unique plantids are present - drop duplicates\n",
    "    #unique_unique_line_mutated = unique_line_mutated.drop_duplicates(subset='plant_ID',keep='first')\n",
    "\n",
    "    non_chimeric = unique_line_mutated[unique_line_mutated.genotype != 'chimeric']\n",
    "    non_chimeric_unique = non_chimeric.drop_duplicates(subset='plant_ID',keep='first')\n",
    "\n",
    "    homozygous = unique_line_mutated[unique_line_mutated.genotype == 'homozygous']\n",
    "    homozygous_unique = homozygous.drop_duplicates(subset='plant_ID',keep='first')\n",
    "\n",
    "    heterozygous = unique_line_mutated[unique_line_mutated.genotype == 'heterozygous']\n",
    "    heterozygous_unique = heterozygous.drop_duplicates(subset='plant_ID',keep='first')\n",
    "    biallelic = unique_line_mutated[unique_line_mutated.genotype == 'biallelic']\n",
    "    biallelic_unique = biallelic.drop_duplicates(subset='plant_ID',keep='first')\n",
    "\n",
    "    #print bold heading\n",
    "    print('\\033[1m'+'Genotypes'+'\\033[0m')\n",
    "    print(f'number of non-chimeric mutated plant lines = {len(non_chimeric_unique)} ({len(non_chimeric_unique)/len(genotyped_only_mutated_counts)*100}% of non-chimeric lines)')\n",
    "    print(f'number of homozygous mutated plant lines = {len(homozygous_unique)} ({len(homozygous_unique)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    print(f'number of heterozygous mutated plant lines = {len(heterozygous_unique)} ({len(heterozygous_unique)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    print(f'number of biallelic mutated plant lines = {len(biallelic_unique)} ({len(biallelic_unique)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    #count number of non-chimeric insertions, deletions, substitutions (if two insertions around same guide, counted as once)\n",
    "    non_chimeric_all = non_chimeric.copy()\n",
    "    insertions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('insertion')]\n",
    "    #print(insertions.insertion_sizes.apply(literal_eval).sum())\n",
    "    deletions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('deletion')]\n",
    "    substitutions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('substitution')]\n",
    "    #number of insertions. evaluate the strings as if they are literals (as if they are lists)\n",
    "    number_of_insertions = len(insertions.insertion_sizes.apply(literal_eval).sum())\n",
    "    #print(np.concatenate(insertions.insertion_genomic_positions_combined)\n",
    "    #number of deletions\n",
    "    all_deletion_lengths = deletions.deletion_sizes.apply(literal_eval).sum()\n",
    "    number_of_deletions = len(all_deletion_lengths)\n",
    "    #number of substitutions\n",
    "    number_of_substitutions = len(substitutions.substitution_sizes.apply(literal_eval).sum())\n",
    "    #number of 1bp insertions\n",
    "    insertions_1bp = list(filter(lambda x: x == 1, insertions.insertion_sizes.apply(literal_eval).sum()))\n",
    "    insertions_percent_1bp = len(insertions_1bp)/number_of_insertions*100\n",
    "    print(f'percentage insertions which are 1bp insertions = {insertions_percent_1bp}%')\n",
    "\n",
    "\n",
    "    median_reads_guide_sites = genotyped_df.read_number.median()\n",
    "    median_mutation_reads_guide_sites = genotyped_only_mutated_df.read_number.median()\n",
    "    #print bold heading\n",
    "    #filter out deletion values less than 10\n",
    "    deletions_over_10bp = list(filter(lambda x: x > 10, all_deletion_lengths))\n",
    "    total_mutations = number_of_insertions + number_of_deletions + number_of_substitutions\n",
    "    print('\\033[1m'+'Mutation types'+'\\033[0m')\n",
    "    print(f'total mutations = {total_mutations}')\n",
    "    print(f'number of insertions = {number_of_insertions} ({number_of_insertions/total_mutations*100}% of total mutations)')\n",
    "    print(f'number of deletions = {number_of_deletions} ({number_of_deletions/total_mutations*100}% of total mutations)')\n",
    "    #deletions over 10bp\n",
    "    #deletions.deletion_positions.str.len()\n",
    "    #deletions_over_10bp = deletions[deletions.deletion_position.str.len()>10]\n",
    "    \n",
    "    #print(all_deletion_lengths)\n",
    "    \n",
    "    #print(deletions_over_10bp)\n",
    "    print(f'deletions over 10bp = {len(deletions_over_10bp)} ({len(deletions_over_10bp)/number_of_deletions*100}% of deletions)')\n",
    "    print(f'number of substitutions = {number_of_substitutions} ({number_of_substitutions/total_mutations*100}% of total mutations)')\n",
    "    print(f'median reads at guide sites including wt = {median_reads_guide_sites}')\n",
    "    print(f'median reads at guide sites mutations only = {median_mutation_reads_guide_sites}')\n",
    "    \n",
    "    #find number of mutations overlapping FIMO predicted TFBS\n",
    "    insertions_overlapping_TFBS = insertions[insertions.insertion_overlapping_TFBS_AGI_combined.notna()]\n",
    "    deletions_overlapping_TFBS = deletions[deletions.deletion_overlapping_TFBS_AGI_combined.notna()]\n",
    "    substitutions_overlapping_TFBS = substitutions[substitutions.substitution_overlapping_TFBS_AGI_combined.notna()]\n",
    "    #count number of insertions overlapping TFBSs\n",
    "    insertions_count = count_number_mutations_overlapping_at_least_one_TFBS(insertions_overlapping_TFBS, 'insertion')\n",
    "    deletions_count = count_number_mutations_overlapping_at_least_one_TFBS(deletions_overlapping_TFBS, 'deletion')\n",
    "    substitutions_count = count_number_mutations_overlapping_at_least_one_TFBS(substitutions_overlapping_TFBS, 'substitution')\n",
    "          \n",
    "\n",
    "    #print bold heading\n",
    "    print('\\033[1m'+'Overlapping TFBSs'+'\\033[0m')\n",
    "    print(f'number of insertions overlapping TFBS = {insertions_count} ({insertions_count/number_of_insertions*100}% of insertions)')\n",
    "    print(f'number of deletions overlapping TFBS = {deletions_count} ({deletions_count/number_of_deletions*100}% of deletions)')\n",
    "    print(f'number of substitutions overlapping TFBS = {substitutions_count} ({substitutions_count/number_of_substitutions*100}% of substitutions)')\n",
    "\n",
    "    #how many within 7bp of cutsite\n",
    "    #insertions_7bp = insertions[[[i for i in x if int(i) <8] for x in insertions.insertion_cut_site_distance.str.strip('][').str.split(', ')]]\n",
    "\n",
    "   # print(insertions_7bp)\n",
    "\n",
    "\n",
    "    return genotyped_counts, genotyped_only_mutated_counts, genotyped_only_mutated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARF9_genotyped_df, ARF9_mutated_df,ARF9_flattened_df = read_files(mutations_ARF9_genotyped,mutations_ARF9_genotyped_only_mutated,mutations_ARF9_genotyped_only_mutated_flattened)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=468 \n",
      "total mutated plant lines=147\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 142 (96.5986394557823% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 92 (64.7887323943662% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 12 (8.450704225352112% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 38 (26.76056338028169% of non-chimeric lines)\n",
      "percentage insertions which are 1bp insertions = 98.35164835164835%\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 236\n",
      "number of insertions = 182 (77.11864406779661% of total mutations)\n",
      "number of deletions = 50 (21.1864406779661% of total mutations)\n",
      "deletions over 10bp = 9 (18.0% of deletions)\n",
      "number of substitutions = 4 (1.694915254237288% of total mutations)\n",
      "median reads at guide sites including wt = 361.0\n",
      "median reads at guide sites mutations only = 234.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 34 (18.681318681318682% of insertions)\n",
      "number of deletions overlapping TFBS = 15 (30.0% of deletions)\n",
      "number of substitutions overlapping TFBS = 0 (0.0% of substitutions)\n"
     ]
    }
   ],
   "source": [
    "ARF9_genotyped_plant_lines, ARF9_mutated_plant_lines,ARF9_mutations = count_mutations(ARF9_genotyped_df,ARF9_mutated_df,ARF9_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter ARF18 guide 14 since very repetitive and lots of sequencing errors\n",
    "ARF18_mutated_df = pd.read_table(mutations_ARF18_genotyped_only_mutated,sep='\\t', header=0)\n",
    "ARF18_mutated_plant_lines_filtered = ARF18_mutated_df[ARF18_mutated_df.guide != 'ARF18_guide14']\n",
    "#save df\n",
    "ARF18_mutated_plant_lines_filtered.to_csv(mutations_ARF18_genotyped_only_mutated_filtered, sep='\\t',header=True)\n",
    "#Then I ran pacbio_analyse_flattened_mutations.py on the filtered .tsv to generate the mutations_ARF18_genotyped_only_mutated_flattened_filtered file (f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_filtered/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARF18_genotyped_df, ARF18_mutated_df,ARF18_flattened_filtered_df = read_files(mutations_ARF18_genotyped,mutations_ARF18_genotyped_only_mutated,mutations_ARF18_genotyped_only_mutated_flattened_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#filter ARF18 guide 14 since very repetitive and lots of sequencing errors\n",
    "ARF18_genotyped_only_mutated_df = pd.read_table(mutations_ARF18_genotyped_only_mutated,sep='\\t', header=0)\n",
    "ARF18_mutated_plant_lines_filtered = ARF18_mutated_df[ARF18_mutated_df.guide != 'ARF18_guide14']\n",
    "#save as df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=373 \n",
      "total mutated plant lines=67\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 66 (98.50746268656717% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 27 (40.909090909090914% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 25 (37.878787878787875% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 14 (21.21212121212121% of non-chimeric lines)\n",
      "percentage insertions which are 1bp insertions = 100.0%\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 106\n",
      "number of insertions = 87 (82.0754716981132% of total mutations)\n",
      "number of deletions = 11 (10.377358490566039% of total mutations)\n",
      "deletions over 10bp = 3 (27.27272727272727% of deletions)\n",
      "number of substitutions = 8 (7.547169811320755% of total mutations)\n",
      "median reads at guide sites including wt = 623.0\n",
      "median reads at guide sites mutations only = 310.5\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 53 (60.91954022988506% of insertions)\n",
      "number of deletions overlapping TFBS = 7 (63.63636363636363% of deletions)\n",
      "number of substitutions overlapping TFBS = 0 (0.0% of substitutions)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#count mutations\n",
    "ARF18_genotyped_plant_lines, ARF18_mutated_plant_lines,ARF18_mutations = count_mutations(ARF18_genotyped_df,ARF18_mutated_plant_lines_filtered,ARF18_flattened_filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=366 \n",
      "total mutated plant lines=43\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 40 (93.02325581395348% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 15 (37.5% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 19 (47.5% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 6 (15.0% of non-chimeric lines)\n",
      "percentage insertions which are 1bp insertions = 100.0%\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 73\n",
      "number of insertions = 40 (54.794520547945204% of total mutations)\n",
      "number of deletions = 22 (30.136986301369863% of total mutations)\n",
      "deletions over 10bp = 2 (9.090909090909092% of deletions)\n",
      "number of substitutions = 11 (15.068493150684931% of total mutations)\n",
      "median reads at guide sites including wt = 344.0\n",
      "median reads at guide sites mutations only = 113.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 18 (45.0% of insertions)\n",
      "number of deletions overlapping TFBS = 16 (72.72727272727273% of deletions)\n",
      "number of substitutions overlapping TFBS = 2 (18.181818181818183% of substitutions)\n"
     ]
    }
   ],
   "source": [
    "DREB26_genotyped_df, DREB26_mutated_df,DREB26_flattened_df = read_files(mutations_DREB26_genotyped,mutations_DREB26_genotyped_only_mutated,mutations_DREB26_genotyped_only_mutated_flattened)\n",
    "DREB26_genotyped_plant_lines, DREB26_mutated_plant_lines,DREB26_mutations = count_mutations(DREB26_genotyped_df,DREB26_mutated_df,DREB26_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=478 \n",
      "total mutated plant lines=112\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 107 (95.53571428571429% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 53 (49.532710280373834% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 32 (29.906542056074763% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 22 (20.5607476635514% of non-chimeric lines)\n",
      "percentage insertions which are 1bp insertions = 100.0%\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 180\n",
      "number of insertions = 127 (70.55555555555556% of total mutations)\n",
      "number of deletions = 39 (21.666666666666668% of total mutations)\n",
      "deletions over 10bp = 10 (25.64102564102564% of deletions)\n",
      "number of substitutions = 14 (7.777777777777778% of total mutations)\n",
      "median reads at guide sites including wt = 646.0\n",
      "median reads at guide sites mutations only = 349.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 41 (32.28346456692913% of insertions)\n",
      "number of deletions overlapping TFBS = 7 (17.94871794871795% of deletions)\n",
      "number of substitutions overlapping TFBS = 1 (7.142857142857142% of substitutions)\n"
     ]
    }
   ],
   "source": [
    "NLP7_genotyped_df, NLP7_mutated_df,NLP7_flattened_df = read_files(mutations_NLP7_genotyped,mutations_NLP7_genotyped_only_mutated,mutations_NLP7_genotyped_only_mutated_flattened)\n",
    "NLP7_genotyped_plant_lines, NLP7_mutated_plant_lines,NLP7_mutations = count_mutations(NLP7_genotyped_df,NLP7_mutated_df,NLP7_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a combined dataframe of all the mutations\n",
    "genotyped_df_all = pd.concat([ARF9_genotyped_df,ARF18_genotyped_df,DREB26_genotyped_df,NLP7_genotyped_df])\n",
    "genotyped_only_mutated_df_all = pd.concat([ARF9_mutated_df,ARF18_mutated_plant_lines_filtered,DREB26_mutated_df,NLP7_mutated_df])\n",
    "genotyped_only_mutated_flattened_df_all = pd.concat([ARF9_flattened_df,ARF18_flattened_filtered_df,DREB26_flattened_df,NLP7_flattened_df])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plant_ID</th>\n",
       "      <th>chr</th>\n",
       "      <th>platename</th>\n",
       "      <th>library</th>\n",
       "      <th>first_reaction_primers</th>\n",
       "      <th>second_reaction_primers</th>\n",
       "      <th>genotype</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>guides</th>\n",
       "      <th>mutation_genotypes</th>\n",
       "      <th>...</th>\n",
       "      <th>deletion_genomic_positions_combined</th>\n",
       "      <th>deletion_positions_relative_to_TSS_combined</th>\n",
       "      <th>deletion_sizes</th>\n",
       "      <th>deletion_overlapping_TFBS_AGI_combined</th>\n",
       "      <th>deletion_overlapping_TFBS_family_combined</th>\n",
       "      <th>substitution_genomic_positions_combined</th>\n",
       "      <th>substitution_positions_relative_to_TSS_combined</th>\n",
       "      <th>substitution_sizes</th>\n",
       "      <th>substitution_overlapping_TFBS_AGI_combined</th>\n",
       "      <th>substitution_overlapping_TFBS_family_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plntEPSWT20056-1</td>\n",
       "      <td>4</td>\n",
       "      <td>p6ARF9</td>\n",
       "      <td>1022</td>\n",
       "      <td>SW348_SW444</td>\n",
       "      <td>ARF9_4F_6R</td>\n",
       "      <td>biallelic</td>\n",
       "      <td>['deletion']</td>\n",
       "      <td>['ARF9_guide1', 'ARF9_guide2', 'ARF9_guide2', ...</td>\n",
       "      <td>['biallelic', 'heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12451335, 12451336, 12451337, 12451338, 12451...</td>\n",
       "      <td>[193, 194, 195, 196, 197, 198, 199]</td>\n",
       "      <td>[7]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>plntEPSWT20056-2</td>\n",
       "      <td>4</td>\n",
       "      <td>p6ARF9</td>\n",
       "      <td>1022</td>\n",
       "      <td>SW360_SW456</td>\n",
       "      <td>ARF9_4F_6R</td>\n",
       "      <td>biallelic</td>\n",
       "      <td>['deletion']</td>\n",
       "      <td>['ARF9_guide2', 'ARF9_guide3', 'ARF9_guide3']</td>\n",
       "      <td>['biallelic', 'homozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12451334, 12451335, 12451336, 12451337]</td>\n",
       "      <td>[192, 193, 194, 195]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plntEPSWT20056-3</td>\n",
       "      <td>4</td>\n",
       "      <td>p6ARF9</td>\n",
       "      <td>1022</td>\n",
       "      <td>SW372_SW468</td>\n",
       "      <td>ARF9_4F_6R</td>\n",
       "      <td>biallelic</td>\n",
       "      <td>['deletion']</td>\n",
       "      <td>['ARF9_guide2', 'ARF9_guide2', 'ARF9_guide3', ...</td>\n",
       "      <td>['biallelic', 'heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12451332, 12451333, 12451334, 12451335, 12451...</td>\n",
       "      <td>[190, 191, 192, 193, 194, 195, 196, 197]</td>\n",
       "      <td>[8]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>plntEPSWT20056-4</td>\n",
       "      <td>4</td>\n",
       "      <td>p6ARF9</td>\n",
       "      <td>1022</td>\n",
       "      <td>SW384_SW480</td>\n",
       "      <td>ARF9_4F_6R</td>\n",
       "      <td>biallelic</td>\n",
       "      <td>['deletion', 'insertion']</td>\n",
       "      <td>['ARF9_guide1', 'ARF9_guide2', 'ARF9_guide2', ...</td>\n",
       "      <td>['biallelic', 'heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12451338, 12451339, 12451340, 12451341]</td>\n",
       "      <td>[196, 197, 198, 199]</td>\n",
       "      <td>[4]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plntEPSWT20056-5</td>\n",
       "      <td>4</td>\n",
       "      <td>p6ARF9</td>\n",
       "      <td>1022</td>\n",
       "      <td>SW396_SW492</td>\n",
       "      <td>ARF9_4F_6R</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>['deletion']</td>\n",
       "      <td>['ARF9_guide1', 'ARF9_guide1', 'ARF9_guide2', ...</td>\n",
       "      <td>['biallelic', 'chimeric']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12451333, 12451334, 12451335, 12451336, 12451...</td>\n",
       "      <td>[191, 192, 193, 194, 195, 196, 197, 198, 199]</td>\n",
       "      <td>[9]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>plntEPSWT20197-6</td>\n",
       "      <td>4</td>\n",
       "      <td>p5NLP7</td>\n",
       "      <td>1021</td>\n",
       "      <td>SW390_SW486</td>\n",
       "      <td>NLP7_4F_6R</td>\n",
       "      <td>homozygous</td>\n",
       "      <td>['insertion']</td>\n",
       "      <td>['NLP7_guide23']</td>\n",
       "      <td>['homozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>plntEPSWT20197-7</td>\n",
       "      <td>4</td>\n",
       "      <td>p5NLP7</td>\n",
       "      <td>1021</td>\n",
       "      <td>SW402_SW498</td>\n",
       "      <td>NLP7_4F_6R</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>['deletion']</td>\n",
       "      <td>['NLP7_guide23', 'NLP7_guide24']</td>\n",
       "      <td>['heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>[12479206, 12479207, 12479208, 12479209, 12479...</td>\n",
       "      <td>[-313, -314, -315, -316, -317, -318, -319, -32...</td>\n",
       "      <td>[9]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>plntEPSWT20197-8</td>\n",
       "      <td>4</td>\n",
       "      <td>p5NLP7</td>\n",
       "      <td>1021</td>\n",
       "      <td>SW414_SW510</td>\n",
       "      <td>NLP7_4F_6R</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>['insertion']</td>\n",
       "      <td>['NLP7_guide23']</td>\n",
       "      <td>['heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>plntEPSWT20197-10</td>\n",
       "      <td>4</td>\n",
       "      <td>p5NLP7</td>\n",
       "      <td>1021</td>\n",
       "      <td>SW438_SW534</td>\n",
       "      <td>NLP7_4F_6R</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>['insertion']</td>\n",
       "      <td>['NLP7_guide18', 'NLP7_guide19', 'NLP7_guide20...</td>\n",
       "      <td>['heterozygous', 'homozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>plntEPSWT20198-1</td>\n",
       "      <td>4</td>\n",
       "      <td>p5NLP7</td>\n",
       "      <td>1021</td>\n",
       "      <td>SW355_SW451</td>\n",
       "      <td>NLP7_4F_6R</td>\n",
       "      <td>heterozygous</td>\n",
       "      <td>['substitution']</td>\n",
       "      <td>['NLP7_guide15', 'NLP7_guide23']</td>\n",
       "      <td>['heterozygous']</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[12479224, 12479504]</td>\n",
       "      <td>[-23, -303]</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>369 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              plant_ID  chr platename  library first_reaction_primers  \\\n",
       "0     plntEPSWT20056-1    4    p6ARF9     1022            SW348_SW444   \n",
       "1     plntEPSWT20056-2    4    p6ARF9     1022            SW360_SW456   \n",
       "2     plntEPSWT20056-3    4    p6ARF9     1022            SW372_SW468   \n",
       "3     plntEPSWT20056-4    4    p6ARF9     1022            SW384_SW480   \n",
       "4     plntEPSWT20056-5    4    p6ARF9     1022            SW396_SW492   \n",
       "..                 ...  ...       ...      ...                    ...   \n",
       "107   plntEPSWT20197-6    4    p5NLP7     1021            SW390_SW486   \n",
       "108   plntEPSWT20197-7    4    p5NLP7     1021            SW402_SW498   \n",
       "109   plntEPSWT20197-8    4    p5NLP7     1021            SW414_SW510   \n",
       "110  plntEPSWT20197-10    4    p5NLP7     1021            SW438_SW534   \n",
       "111   plntEPSWT20198-1    4    p5NLP7     1021            SW355_SW451   \n",
       "\n",
       "    second_reaction_primers      genotype              mutation_type  \\\n",
       "0                ARF9_4F_6R     biallelic               ['deletion']   \n",
       "1                ARF9_4F_6R     biallelic               ['deletion']   \n",
       "2                ARF9_4F_6R     biallelic               ['deletion']   \n",
       "3                ARF9_4F_6R     biallelic  ['deletion', 'insertion']   \n",
       "4                ARF9_4F_6R      chimeric               ['deletion']   \n",
       "..                      ...           ...                        ...   \n",
       "107              NLP7_4F_6R    homozygous              ['insertion']   \n",
       "108              NLP7_4F_6R  heterozygous               ['deletion']   \n",
       "109              NLP7_4F_6R  heterozygous              ['insertion']   \n",
       "110              NLP7_4F_6R  heterozygous              ['insertion']   \n",
       "111              NLP7_4F_6R  heterozygous           ['substitution']   \n",
       "\n",
       "                                                guides  \\\n",
       "0    ['ARF9_guide1', 'ARF9_guide2', 'ARF9_guide2', ...   \n",
       "1        ['ARF9_guide2', 'ARF9_guide3', 'ARF9_guide3']   \n",
       "2    ['ARF9_guide2', 'ARF9_guide2', 'ARF9_guide3', ...   \n",
       "3    ['ARF9_guide1', 'ARF9_guide2', 'ARF9_guide2', ...   \n",
       "4    ['ARF9_guide1', 'ARF9_guide1', 'ARF9_guide2', ...   \n",
       "..                                                 ...   \n",
       "107                                   ['NLP7_guide23']   \n",
       "108                   ['NLP7_guide23', 'NLP7_guide24']   \n",
       "109                                   ['NLP7_guide23']   \n",
       "110  ['NLP7_guide18', 'NLP7_guide19', 'NLP7_guide20...   \n",
       "111                   ['NLP7_guide15', 'NLP7_guide23']   \n",
       "\n",
       "                 mutation_genotypes  ...  \\\n",
       "0     ['biallelic', 'heterozygous']  ...   \n",
       "1       ['biallelic', 'homozygous']  ...   \n",
       "2     ['biallelic', 'heterozygous']  ...   \n",
       "3     ['biallelic', 'heterozygous']  ...   \n",
       "4         ['biallelic', 'chimeric']  ...   \n",
       "..                              ...  ...   \n",
       "107                  ['homozygous']  ...   \n",
       "108                ['heterozygous']  ...   \n",
       "109                ['heterozygous']  ...   \n",
       "110  ['heterozygous', 'homozygous']  ...   \n",
       "111                ['heterozygous']  ...   \n",
       "\n",
       "                   deletion_genomic_positions_combined  \\\n",
       "0    [12451335, 12451336, 12451337, 12451338, 12451...   \n",
       "1             [12451334, 12451335, 12451336, 12451337]   \n",
       "2    [12451332, 12451333, 12451334, 12451335, 12451...   \n",
       "3             [12451338, 12451339, 12451340, 12451341]   \n",
       "4    [12451333, 12451334, 12451335, 12451336, 12451...   \n",
       "..                                                 ...   \n",
       "107                                                NaN   \n",
       "108  [12479206, 12479207, 12479208, 12479209, 12479...   \n",
       "109                                                NaN   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "\n",
       "           deletion_positions_relative_to_TSS_combined deletion_sizes  \\\n",
       "0                  [193, 194, 195, 196, 197, 198, 199]            [7]   \n",
       "1                                 [192, 193, 194, 195]            [4]   \n",
       "2             [190, 191, 192, 193, 194, 195, 196, 197]            [8]   \n",
       "3                                 [196, 197, 198, 199]            [4]   \n",
       "4        [191, 192, 193, 194, 195, 196, 197, 198, 199]            [9]   \n",
       "..                                                 ...            ...   \n",
       "107                                                NaN            NaN   \n",
       "108  [-313, -314, -315, -316, -317, -318, -319, -32...            [9]   \n",
       "109                                                NaN            NaN   \n",
       "110                                                NaN            NaN   \n",
       "111                                                NaN            NaN   \n",
       "\n",
       "    deletion_overlapping_TFBS_AGI_combined  \\\n",
       "0                                      NaN   \n",
       "1                                      NaN   \n",
       "2                                      NaN   \n",
       "3                                      NaN   \n",
       "4                                      NaN   \n",
       "..                                     ...   \n",
       "107                                    NaN   \n",
       "108                                    NaN   \n",
       "109                                    NaN   \n",
       "110                                    NaN   \n",
       "111                                    NaN   \n",
       "\n",
       "    deletion_overlapping_TFBS_family_combined  \\\n",
       "0                                         NaN   \n",
       "1                                         NaN   \n",
       "2                                         NaN   \n",
       "3                                         NaN   \n",
       "4                                         NaN   \n",
       "..                                        ...   \n",
       "107                                       NaN   \n",
       "108                                       NaN   \n",
       "109                                       NaN   \n",
       "110                                       NaN   \n",
       "111                                       NaN   \n",
       "\n",
       "    substitution_genomic_positions_combined  \\\n",
       "0                                       NaN   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "..                                      ...   \n",
       "107                                     NaN   \n",
       "108                                     NaN   \n",
       "109                                     NaN   \n",
       "110                                     NaN   \n",
       "111                    [12479224, 12479504]   \n",
       "\n",
       "    substitution_positions_relative_to_TSS_combined substitution_sizes  \\\n",
       "0                                               NaN                NaN   \n",
       "1                                               NaN                NaN   \n",
       "2                                               NaN                NaN   \n",
       "3                                               NaN                NaN   \n",
       "4                                               NaN                NaN   \n",
       "..                                              ...                ...   \n",
       "107                                             NaN                NaN   \n",
       "108                                             NaN                NaN   \n",
       "109                                             NaN                NaN   \n",
       "110                                             NaN                NaN   \n",
       "111                                     [-23, -303]             [1, 1]   \n",
       "\n",
       "    substitution_overlapping_TFBS_AGI_combined  \\\n",
       "0                                          NaN   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "..                                         ...   \n",
       "107                                        NaN   \n",
       "108                                        NaN   \n",
       "109                                        NaN   \n",
       "110                                        NaN   \n",
       "111                                        NaN   \n",
       "\n",
       "    substitution_overlapping_TFBS_family_combined  \n",
       "0                                             NaN  \n",
       "1                                             NaN  \n",
       "2                                             NaN  \n",
       "3                                             NaN  \n",
       "4                                             NaN  \n",
       "..                                            ...  \n",
       "107                                           NaN  \n",
       "108                                           NaN  \n",
       "109                                           NaN  \n",
       "110                                           NaN  \n",
       "111                                           NaN  \n",
       "\n",
       "[369 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotyped_only_mutated_flattened_df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARF9' 'ARF18' 'DREB26' 'NLP7']\n",
      "unique mutated plant lines = 327\n",
      "number of plant lines with mutations in more than one gene = 41\n",
      "41\n",
      "(ARF9, NLP7)            27\n",
      "(ARF18, ARF9)            9\n",
      "(ARF18, NLP7)            2\n",
      "(ARF9, DREB26)           2\n",
      "(ARF9, DREB26, NLP7)     1\n",
      "Name: genes_with_mutations_sorted, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#find plant lines with mutations in more than one gene\n",
    "genotyped_only_mutated_flattened_df_all['gene'] = genotyped_only_mutated_flattened_df_all['second_reaction_primers'].str.split('_').str[0]\n",
    "print(genotyped_only_mutated_flattened_df_all.gene.unique())\n",
    "unique_mutated_lines = genotyped_only_mutated_flattened_df_all['plant_ID'].unique()\n",
    "print(f'unique mutated plant lines = {len(unique_mutated_lines)}')\n",
    "# make df containing only plant lines with mutations in more than one gene\n",
    "#make df containing all plant lines with duplicates based on the plant_ID column\n",
    "duplicated_plant_lines_all = genotyped_only_mutated_flattened_df_all[genotyped_only_mutated_flattened_df_all.duplicated(subset=['plant_ID'], keep=False)]\n",
    "#get unique plant lines\n",
    "duplicated_plant_lines_unique = duplicated_plant_lines_all['plant_ID'].unique()\n",
    "# duplicated_plant_lines_unique = duplicated_plant_lines_all[]\n",
    "#sort by plant_id\n",
    "duplicated_plant_lines_all = duplicated_plant_lines_all.sort_values(by=['plant_ID'])\n",
    "# duplicated_plant_lines_unique = duplicated_plant_lines_unique.sort_values(by=['plant_ID'])\n",
    "#reset index\n",
    "duplicated_plant_lines_all = duplicated_plant_lines_all.reset_index(drop=True)\n",
    "# duplicated_plant_lines_unique = duplicated_plant_lines_unique.reset_index(drop=True)\n",
    "\n",
    "#remove nan values\n",
    "# duplicated_plant_lines_unique = duplicated_plant_lines_unique[duplicated_plant_lines_unique['plant_ID'].notna()]\n",
    "print(f'number of plant lines with mutations in more than one gene = {len(duplicated_plant_lines_unique)}')\n",
    "#duplicated_plant_lines_all = genotyped_only_mutated_flattened_df_all[genotyped_only_mutated_flattened_df_all.duplicated(subset=['plant_ID'], keep=True)]\n",
    "#make new flattened df containing list of unique plants with a column showing which genes they have mutations in\n",
    "# flattened_multiple_mutations = duplicated_plant_lines_all.pivot_table(index='plant_ID', columns='gene', values='gene', aggfunc='count')\n",
    "# # print(len(flattened_multiple_mutations)) \n",
    "# # print(flattened_multiple_mutations)\n",
    "# print(len(flattened_multiple_mutations))\n",
    "# print(flattened_multiple_mutations)\n",
    "# create flattened_multiple_mutations containing only unique plant lines, with a gene column containing a list of genes with mutations\n",
    "#flattened_multiple_mutations = duplicated_plant_lines_unique['plant_ID'].to_frame()\n",
    "\n",
    "flattened_multiple_mutations = duplicated_plant_lines_all.groupby('plant_ID')['gene'].apply(list).to_frame()\n",
    "flattened_multiple_mutations = flattened_multiple_mutations.reset_index()\n",
    "flattened_multiple_mutations = flattened_multiple_mutations.rename(columns={'gene':'genes_with_mutations'})\n",
    "#make into df\n",
    "flattened_multiple_mutations = pd.DataFrame(flattened_multiple_mutations)\n",
    "#sort by plant_id\n",
    "flattened_multiple_mutations = flattened_multiple_mutations.sort_values(by=['plant_ID'])\n",
    "print(len(flattened_multiple_mutations))\n",
    "flattened_multiple_mutations\n",
    "# #check differences in plant_ID column between flattened_multiple_mutations and duplicated_plant_lines_unique\n",
    "# print(len(flattened_multiple_mutations['plant_ID'].unique()))\n",
    "# print(len(duplicated_plant_lines_unique['plant_ID'].unique()))\n",
    "# print(flattened_multiple_mutations['plant_ID'].unique() == duplicated_plant_lines_unique['plant_ID'].unique())\n",
    "# print(flattened_multiple_mutations['plant_ID'].unique() != duplicated_plant_lines_unique['plant_ID'].unique())\n",
    "#count how many genes_with_mutations lists are unique\n",
    "# out = flattened_multiple_mutations.genes_with_mutations.apply(tuple).drop_duplicates().apply(list).tolist()\n",
    "flattened_multiple_mutations['genes_with_mutations_sorted'] = flattened_multiple_mutations.genes_with_mutations.sort_values().apply(lambda x: sorted(x))\n",
    "print(flattened_multiple_mutations['genes_with_mutations_sorted'].apply(tuple).value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['plntEPSWT20056-1', 'plntEPSWT20056-2', 'plntEPSWT20056-3',\n",
       "       'plntEPSWT20056-5', 'plntEPSWT20056-7', 'plntEPSWT20056-8',\n",
       "       'plntEPSWT20056-9', 'plntEPSWT20056-10', 'plntEPSWT20125-1',\n",
       "       'plntEPSWT20125-2', 'plntEPSWT20125-3', 'plntEPSWT20125-4',\n",
       "       'plntEPSWT20125-5', 'plntEPSWT20125-6', 'plntEPSWT20125-8',\n",
       "       'plntEPSWT20125-9', 'plntEPSWT20125-10', 'plntEPSWT20126-1',\n",
       "       'plntEPSWT20126-2', 'plntEPSWT20126-3', 'plntEPSWT20126-4',\n",
       "       'plntEPSWT20126-5', 'plntEPSWT20126-6', 'plntEPSWT20126-8',\n",
       "       'plntEPSWT20126-9', 'plntEPSWT20126-10', 'plntEPSWT20128-2',\n",
       "       'plntEPSWT20138-3', 'plntEPSWT20138-6', 'plntEPSWT20138-10',\n",
       "       'plntEPSWT20154-4', 'plntEPSWT20154-10', 'plntEPSWT20198-1',\n",
       "       'plntEPSWT20198-9', 'plntEPSWT20199-2', 'plntEPSWT20199-5',\n",
       "       'plntEPSWT20199-7', 'plntEPSWT20199-8', 'plntEPSWT20199-9',\n",
       "       'plntEPSWT20072-9', 'plntEPSWT20075-1'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_plant_lines_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=499 \n",
      "total mutated plant lines=327\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 315 (96.3302752293578% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 176 (49.57746478873239% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 84 (23.661971830985916% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 73 (20.56338028169014% of non-chimeric lines)\n",
      "percentage insertions which are 1bp insertions = 99.31192660550458%\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 595\n",
      "number of insertions = 436 (73.27731092436974% of total mutations)\n",
      "number of deletions = 122 (20.504201680672267% of total mutations)\n",
      "deletions over 10bp = 24 (19.672131147540984% of deletions)\n",
      "number of substitutions = 37 (6.218487394957983% of total mutations)\n",
      "median reads at guide sites including wt = 475.0\n",
      "median reads at guide sites mutations only = 253.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 146 (33.48623853211009% of insertions)\n",
      "number of deletions overlapping TFBS = 45 (36.885245901639344% of deletions)\n",
      "number of substitutions overlapping TFBS = 3 (8.108108108108109% of substitutions)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array(['plntEPSWT20057-1', 'plntEPSWT20057-10', 'plntEPSWT20057-2',\n",
       "        'plntEPSWT20057-3', 'plntEPSWT20057-4', 'plntEPSWT20057-5',\n",
       "        'plntEPSWT20057-6', 'plntEPSWT20057-7', 'plntEPSWT20057-8',\n",
       "        'plntEPSWT20057-9', 'plntEPSWT20058-1', 'plntEPSWT20058-10',\n",
       "        'plntEPSWT20058-2', 'plntEPSWT20058-3', 'plntEPSWT20058-5',\n",
       "        'plntEPSWT20058-6', 'plntEPSWT20058-7', 'plntEPSWT20058-8',\n",
       "        'plntEPSWT20058-9', 'plntEPSWT20059-1', 'plntEPSWT20059-10',\n",
       "        'plntEPSWT20059-2', 'plntEPSWT20059-3', 'plntEPSWT20059-4',\n",
       "        'plntEPSWT20059-5', 'plntEPSWT20059-6', 'plntEPSWT20059-7',\n",
       "        'plntEPSWT20059-8', 'plntEPSWT20059-9', 'plntEPSWT20061-1',\n",
       "        'plntEPSWT20061-10', 'plntEPSWT20061-2', 'plntEPSWT20061-3',\n",
       "        'plntEPSWT20061-4', 'plntEPSWT20061-5', 'plntEPSWT20061-6',\n",
       "        'plntEPSWT20061-7', 'plntEPSWT20061-8', 'plntEPSWT20061-9',\n",
       "        'plntEPSWT20062-1', 'plntEPSWT20062-10', 'plntEPSWT20062-2',\n",
       "        'plntEPSWT20062-3', 'plntEPSWT20062-4', 'plntEPSWT20062-5',\n",
       "        'plntEPSWT20062-6', 'plntEPSWT20062-7', 'plntEPSWT20062-8',\n",
       "        'plntEPSWT20062-9', 'plntEPSWT20063-10', 'plntEPSWT20063-2',\n",
       "        'plntEPSWT20063-3', 'plntEPSWT20063-4', 'plntEPSWT20063-5',\n",
       "        'plntEPSWT20063-6', 'plntEPSWT20063-7', 'plntEPSWT20063-8',\n",
       "        'plntEPSWT20063-9', 'plntEPSWT20064-1', 'plntEPSWT20064-10',\n",
       "        'plntEPSWT20064-2', 'plntEPSWT20064-3', 'plntEPSWT20064-4',\n",
       "        'plntEPSWT20064-5', 'plntEPSWT20064-6', 'plntEPSWT20064-7',\n",
       "        'plntEPSWT20064-8', 'plntEPSWT20064-9', 'plntEPSWT20066-1',\n",
       "        'plntEPSWT20066-10', 'plntEPSWT20066-2', 'plntEPSWT20066-3',\n",
       "        'plntEPSWT20066-4', 'plntEPSWT20066-5', 'plntEPSWT20066-6',\n",
       "        'plntEPSWT20066-7', 'plntEPSWT20066-8', 'plntEPSWT20066-9',\n",
       "        'plntEPSWT20067-1', 'plntEPSWT20067-10', 'plntEPSWT20067-2',\n",
       "        'plntEPSWT20067-3', 'plntEPSWT20067-4', 'plntEPSWT20067-5',\n",
       "        'plntEPSWT20067-6', 'plntEPSWT20067-7', 'plntEPSWT20067-8',\n",
       "        'plntEPSWT20067-9', 'plntEPSWT20068-1', 'plntEPSWT20068-2',\n",
       "        'plntEPSWT20068-3', 'plntEPSWT20068-4', 'plntEPSWT20068-5',\n",
       "        'plntEPSWT20068-10', 'plntEPSWT20068-6', 'plntEPSWT20068-7',\n",
       "        'plntEPSWT20068-9', 'plntEPSWT20069-1', 'plntEPSWT20069-2',\n",
       "        'plntEPSWT20069-3', 'plntEPSWT20069-4', 'plntEPSWT20069-5',\n",
       "        'plntEPSWT20069-6', 'plntEPSWT20069-7', 'plntEPSWT20069-8',\n",
       "        'plntEPSWT20069-9', 'plntEPSWT20070-10', 'plntEPSWT20070-3',\n",
       "        'plntEPSWT20070-7', 'plntEPSWT20070-8', 'plntEPSWT20070-9',\n",
       "        'plntEPSWT20071-1', 'plntEPSWT20071-10', 'plntEPSWT20071-2',\n",
       "        'plntEPSWT20071-3', 'plntEPSWT20071-4', 'plntEPSWT20071-5',\n",
       "        'plntEPSWT20071-6', 'plntEPSWT20071-7', 'plntEPSWT20071-8',\n",
       "        'plntEPSWT20071-9', 'plntEPSWT20072-1', 'plntEPSWT20072-10',\n",
       "        'plntEPSWT20072-2', 'plntEPSWT20072-3', 'plntEPSWT20072-4',\n",
       "        'plntEPSWT20072-5', 'plntEPSWT20072-6', 'plntEPSWT20072-7',\n",
       "        'plntEPSWT20072-8', 'plntEPSWT20072-9', 'plntEPSWT20075-1',\n",
       "        'plntEPSWT20075-10', 'plntEPSWT20075-2', 'plntEPSWT20075-3',\n",
       "        'plntEPSWT20075-4', 'plntEPSWT20075-5', 'plntEPSWT20075-6',\n",
       "        'plntEPSWT20075-8', 'plntEPSWT20075-9', 'plntEPSWT20076-1',\n",
       "        'plntEPSWT20076-10', 'plntEPSWT20076-2', 'plntEPSWT20076-3',\n",
       "        'plntEPSWT20076-4', 'plntEPSWT20076-5', 'plntEPSWT20076-6',\n",
       "        'plntEPSWT20076-7', 'plntEPSWT20076-8', 'plntEPSWT20076-9',\n",
       "        'plntEPSWT20077-1', 'plntEPSWT20077-10', 'plntEPSWT20077-2',\n",
       "        'plntEPSWT20077-3', 'plntEPSWT20077-4', 'plntEPSWT20077-5',\n",
       "        'plntEPSWT20077-6', 'plntEPSWT20077-7', 'plntEPSWT20077-8',\n",
       "        'plntEPSWT20077-9', 'plntEPSWT20125-1', 'plntEPSWT20125-10',\n",
       "        'plntEPSWT20125-2', 'plntEPSWT20125-3', 'plntEPSWT20125-4',\n",
       "        'plntEPSWT20125-5', 'plntEPSWT20125-6', 'plntEPSWT20125-8',\n",
       "        'plntEPSWT20125-9', 'plntEPSWT20126-1', 'plntEPSWT20126-10',\n",
       "        'plntEPSWT20126-2', 'plntEPSWT20126-3', 'plntEPSWT20126-4',\n",
       "        'plntEPSWT20126-5', 'plntEPSWT20126-6', 'plntEPSWT20126-8',\n",
       "        'plntEPSWT20126-9', 'plntEPSWT20127-1', 'plntEPSWT20127-10',\n",
       "        'plntEPSWT20127-2', 'plntEPSWT20127-3', 'plntEPSWT20127-4',\n",
       "        'plntEPSWT20127-5', 'plntEPSWT20127-6', 'plntEPSWT20127-7',\n",
       "        'plntEPSWT20127-8', 'plntEPSWT20127-9', 'plntEPSWT20128-1',\n",
       "        'plntEPSWT20128-2', 'plntEPSWT20128-7', 'plntEPSWT20128-9',\n",
       "        'plntEPSWT20130-1', 'plntEPSWT20130-10', 'plntEPSWT20130-11',\n",
       "        'plntEPSWT20130-12', 'plntEPSWT20130-13', 'plntEPSWT20130-14',\n",
       "        'plntEPSWT20130-2', 'plntEPSWT20130-3', 'plntEPSWT20130-4',\n",
       "        'plntEPSWT20130-5', 'plntEPSWT20130-6', 'plntEPSWT20130-7',\n",
       "        'plntEPSWT20130-8', 'plntEPSWT20130-9', 'plntEPSWT20131-1',\n",
       "        'plntEPSWT20131-10', 'plntEPSWT20131-2', 'plntEPSWT20131-3',\n",
       "        'plntEPSWT20131-4', 'plntEPSWT20131-5', 'plntEPSWT20131-7',\n",
       "        'plntEPSWT20131-8', 'plntEPSWT20131-9', 'plntEPSWT20134-1',\n",
       "        'plntEPSWT20134-10', 'plntEPSWT20134-2', 'plntEPSWT20134-3',\n",
       "        'plntEPSWT20134-4', 'plntEPSWT20134-5', 'plntEPSWT20134-6',\n",
       "        'plntEPSWT20134-7', 'plntEPSWT20134-8', 'plntEPSWT20134-9',\n",
       "        'plntEPSWT20135-1', 'plntEPSWT20135-10', 'plntEPSWT20135-2',\n",
       "        'plntEPSWT20135-3', 'plntEPSWT20135-4', 'plntEPSWT20135-5',\n",
       "        'plntEPSWT20135-6', 'plntEPSWT20135-7', 'plntEPSWT20135-8',\n",
       "        'plntEPSWT20135-9', 'plntEPSWT20136-1', 'plntEPSWT20136-10',\n",
       "        'plntEPSWT20136-2', 'plntEPSWT20136-3', 'plntEPSWT20136-4',\n",
       "        'plntEPSWT20136-5', 'plntEPSWT20136-6', 'plntEPSWT20136-7',\n",
       "        'plntEPSWT20136-8', 'plntEPSWT20136-9', 'plntEPSWT20137-1',\n",
       "        'plntEPSWT20137-10', 'plntEPSWT20137-2', 'plntEPSWT20137-3',\n",
       "        'plntEPSWT20137-4', 'plntEPSWT20137-5', 'plntEPSWT20137-6',\n",
       "        'plntEPSWT20137-7', 'plntEPSWT20137-8', 'plntEPSWT20137-9',\n",
       "        'plntEPSWT20138-10', 'plntEPSWT20138-2', 'plntEPSWT20138-3',\n",
       "        'plntEPSWT20138-6', 'plntEPSWT20138-7', 'plntEPSWT20139-1',\n",
       "        'plntEPSWT20139-2', 'plntEPSWT20139-10', 'plntEPSWT20139-3',\n",
       "        'plntEPSWT20139-4', 'plntEPSWT20139-5', 'plntEPSWT20139-6',\n",
       "        'plntEPSWT20139-7', 'plntEPSWT20139-8', 'plntEPSWT20139-9',\n",
       "        'plntEPSWT20141-1', 'plntEPSWT20141-10', 'plntEPSWT20141-2',\n",
       "        'plntEPSWT20141-3', 'plntEPSWT20141-4', 'plntEPSWT20141-5',\n",
       "        'plntEPSWT20141-6', 'plntEPSWT20141-8', 'plntEPSWT20141-9',\n",
       "        'plntEPSWT20142-1', 'plntEPSWT20142-10', 'plntEPSWT20142-2',\n",
       "        'plntEPSWT20142-3', 'plntEPSWT20142-4', 'plntEPSWT20142-5',\n",
       "        'plntEPSWT20142-6', 'plntEPSWT20142-7', 'plntEPSWT20142-8',\n",
       "        'plntEPSWT20142-9', 'plntEPSWT20143-1', 'plntEPSWT20143-10',\n",
       "        'plntEPSWT20143-2', 'plntEPSWT20143-3', 'plntEPSWT20143-4',\n",
       "        'plntEPSWT20143-5', 'plntEPSWT20143-6', 'plntEPSWT20143-7',\n",
       "        'plntEPSWT20143-8', 'plntEPSWT20143-9', 'plntEPSWT20144-1',\n",
       "        'plntEPSWT20144-10', 'plntEPSWT20144-2', 'plntEPSWT20144-3',\n",
       "        'plntEPSWT20144-4', 'plntEPSWT20144-5', 'plntEPSWT20144-6',\n",
       "        'plntEPSWT20144-7', 'plntEPSWT20144-8', 'plntEPSWT20144-9',\n",
       "        'plntEPSWT20145-1', 'plntEPSWT20145-10', 'plntEPSWT20145-2',\n",
       "        'plntEPSWT20145-3', 'plntEPSWT20145-4', 'plntEPSWT20145-5',\n",
       "        'plntEPSWT20145-6', 'plntEPSWT20145-7', 'plntEPSWT20145-8',\n",
       "        'plntEPSWT20145-9', 'plntEPSWT20146-1', 'plntEPSWT20146-10',\n",
       "        'plntEPSWT20146-2', 'plntEPSWT20146-3', 'plntEPSWT20146-4',\n",
       "        'plntEPSWT20146-5', 'plntEPSWT20146-6', 'plntEPSWT20146-7',\n",
       "        'plntEPSWT20146-8', 'plntEPSWT20146-9', 'plntEPSWT20147-1',\n",
       "        'plntEPSWT20147-10', 'plntEPSWT20147-2', 'plntEPSWT20147-3',\n",
       "        'plntEPSWT20147-4', 'plntEPSWT20147-5', 'plntEPSWT20147-6',\n",
       "        'plntEPSWT20147-7', 'plntEPSWT20147-8', 'plntEPSWT20147-9',\n",
       "        'plntEPSWT20153-1', 'plntEPSWT20153-10', 'plntEPSWT20153-2',\n",
       "        'plntEPSWT20153-3', 'plntEPSWT20153-4', 'plntEPSWT20153-5',\n",
       "        'plntEPSWT20153-6', 'plntEPSWT20153-7', 'plntEPSWT20153-8',\n",
       "        'plntEPSWT20153-9', 'plntEPSWT20154-1', 'plntEPSWT20154-2',\n",
       "        'plntEPSWT20154-3', 'plntEPSWT20154-4', 'plntEPSWT20154-10',\n",
       "        'plntEPSWT20154-9', 'plntEPSWT20156-1', 'plntEPSWT20156-10',\n",
       "        'plntEPSWT20156-2', 'plntEPSWT20156-3', 'plntEPSWT20156-4',\n",
       "        'plntEPSWT20156-5', 'plntEPSWT20156-6', 'plntEPSWT20156-7',\n",
       "        'plntEPSWT20156-8', 'plntEPSWT20156-9', 'plntEPSWT20158-1',\n",
       "        'plntEPSWT20158-10', 'plntEPSWT20158-2', 'plntEPSWT20158-4',\n",
       "        'plntEPSWT20158-5', 'plntEPSWT20158-6', 'plntEPSWT20158-7',\n",
       "        'plntEPSWT20158-8', 'plntEPSWT20158-9', 'plntEPSWT20165-1',\n",
       "        'plntEPSWT20165-10', 'plntEPSWT20165-2', 'plntEPSWT20165-3',\n",
       "        'plntEPSWT20165-4', 'plntEPSWT20165-5', 'plntEPSWT20165-6',\n",
       "        'plntEPSWT20165-7', 'plntEPSWT20165-8', 'plntEPSWT20165-9',\n",
       "        'plntEPSWT20167-1', 'plntEPSWT20167-10', 'plntEPSWT20167-2',\n",
       "        'plntEPSWT20167-3', 'plntEPSWT20167-4', 'plntEPSWT20167-5',\n",
       "        'plntEPSWT20167-6', 'plntEPSWT20167-7', 'plntEPSWT20167-8',\n",
       "        'plntEPSWT20167-9', 'plntEPSWT20176-1', 'plntEPSWT20176-10',\n",
       "        'plntEPSWT20176-2', 'plntEPSWT20176-3', 'plntEPSWT20176-4',\n",
       "        'plntEPSWT20176-5', 'plntEPSWT20176-6', 'plntEPSWT20176-7',\n",
       "        'plntEPSWT20176-8', 'plntEPSWT20176-9', 'plntEPSWT20187-1',\n",
       "        'plntEPSWT20187-10', 'plntEPSWT20187-2', 'plntEPSWT20187-3',\n",
       "        'plntEPSWT20187-4', 'plntEPSWT20187-5', 'plntEPSWT20187-6',\n",
       "        'plntEPSWT20187-7', 'plntEPSWT20187-9', 'plntEPSWT20197-1',\n",
       "        'plntEPSWT20197-10', 'plntEPSWT20197-2', 'plntEPSWT20197-3',\n",
       "        'plntEPSWT20197-4', 'plntEPSWT20197-5', 'plntEPSWT20197-6',\n",
       "        'plntEPSWT20197-7', 'plntEPSWT20197-8', 'plntEPSWT20197-9',\n",
       "        'plntEPSWT20198-1', 'plntEPSWT20198-10', 'plntEPSWT20198-2',\n",
       "        'plntEPSWT20198-3', 'plntEPSWT20198-4', 'plntEPSWT20198-6',\n",
       "        'plntEPSWT20198-7', 'plntEPSWT20198-8', 'plntEPSWT20198-9',\n",
       "        'plntEPSWT20199-1', 'plntEPSWT20199-2', 'plntEPSWT20199-3',\n",
       "        'plntEPSWT20199-5', 'plntEPSWT20199-7', 'plntEPSWT20199-8',\n",
       "        'plntEPSWT20199-9', 'plntEPSWT20201-1', 'plntEPSWT20201-3',\n",
       "        'plntEPSWT20201-4', 'plntEPSWT20056-1', 'plntEPSWT20056-10',\n",
       "        'plntEPSWT20056-2', 'plntEPSWT20056-3', 'plntEPSWT20056-4',\n",
       "        'plntEPSWT20056-5', 'plntEPSWT20056-6', 'plntEPSWT20056-7',\n",
       "        'plntEPSWT20056-8', 'plntEPSWT20056-9', 'plntEPSWT20201-10',\n",
       "        'plntEPSWT20201-5', 'plntEPSWT20201-6', 'plntEPSWT20201-7',\n",
       "        'plntEPSWT20201-8', 'plntEPSWT20201-9', 'plntEPSWT20204-1',\n",
       "        'plntEPSWT20204-10', 'plntEPSWT20204-2', 'plntEPSWT20204-3',\n",
       "        'plntEPSWT20204-4', 'plntEPSWT20204-5', 'plntEPSWT20204-6',\n",
       "        'plntEPSWT20204-7', 'plntEPSWT20204-8', 'plntEPSWT20204-9',\n",
       "        'plntEPSWT20068-8', 'plntEPSWT20070-1', 'plntEPSWT20070-2',\n",
       "        'plntEPSWT20070-4', 'plntEPSWT20070-5', 'plntEPSWT20070-6',\n",
       "        'plntEPSWT20075-7', 'plntEPSWT20125-7', 'plntEPSWT20126-7',\n",
       "        'plntEPSWT20128-10', 'plntEPSWT20128-3', 'plntEPSWT20128-4',\n",
       "        'plntEPSWT20128-6', 'plntEPSWT20128-8', 'plntEPSWT20131-6',\n",
       "        'plntEPSWT20138-1', 'plntEPSWT20138-4', 'plntEPSWT20138-5',\n",
       "        'plntEPSWT20138-8', 'plntEPSWT20138-9', 'plntEPSWT20154-5',\n",
       "        'plntEPSWT20154-6', 'plntEPSWT20154-7', 'plntEPSWT20154-8',\n",
       "        'plntEPSWT20187-8', 'plntEPSWT20198-5', 'plntEPSWT20199-10',\n",
       "        'plntEPSWT20199-4', 'plntEPSWT20199-6', 'plntEPSWT20201-2',\n",
       "        'plntEPSWT20063-1'], dtype=object),\n",
       " array(['plntEPSWT20057-1', 'plntEPSWT20057-10', 'plntEPSWT20057-2',\n",
       "        'plntEPSWT20057-3', 'plntEPSWT20057-4', 'plntEPSWT20057-5',\n",
       "        'plntEPSWT20057-6', 'plntEPSWT20057-7', 'plntEPSWT20057-8',\n",
       "        'plntEPSWT20057-9', 'plntEPSWT20059-1', 'plntEPSWT20063-10',\n",
       "        'plntEPSWT20063-2', 'plntEPSWT20063-3', 'plntEPSWT20063-4',\n",
       "        'plntEPSWT20063-5', 'plntEPSWT20063-6', 'plntEPSWT20063-7',\n",
       "        'plntEPSWT20063-8', 'plntEPSWT20063-9', 'plntEPSWT20068-1',\n",
       "        'plntEPSWT20068-2', 'plntEPSWT20068-3', 'plntEPSWT20068-4',\n",
       "        'plntEPSWT20068-5', 'plntEPSWT20068-10', 'plntEPSWT20068-6',\n",
       "        'plntEPSWT20068-7', 'plntEPSWT20068-9', 'plntEPSWT20070-10',\n",
       "        'plntEPSWT20070-3', 'plntEPSWT20070-7', 'plntEPSWT20070-8',\n",
       "        'plntEPSWT20070-9', 'plntEPSWT20125-1', 'plntEPSWT20125-10',\n",
       "        'plntEPSWT20125-2', 'plntEPSWT20125-3', 'plntEPSWT20125-4',\n",
       "        'plntEPSWT20125-5', 'plntEPSWT20125-6', 'plntEPSWT20125-8',\n",
       "        'plntEPSWT20125-9', 'plntEPSWT20126-1', 'plntEPSWT20126-10',\n",
       "        'plntEPSWT20126-2', 'plntEPSWT20126-3', 'plntEPSWT20126-4',\n",
       "        'plntEPSWT20126-5', 'plntEPSWT20126-6', 'plntEPSWT20126-8',\n",
       "        'plntEPSWT20126-9', 'plntEPSWT20128-2', 'plntEPSWT20137-1',\n",
       "        'plntEPSWT20137-10', 'plntEPSWT20137-2', 'plntEPSWT20137-3',\n",
       "        'plntEPSWT20137-4', 'plntEPSWT20137-5', 'plntEPSWT20137-6',\n",
       "        'plntEPSWT20137-7', 'plntEPSWT20137-8', 'plntEPSWT20137-9',\n",
       "        'plntEPSWT20138-10', 'plntEPSWT20138-3', 'plntEPSWT20138-6',\n",
       "        'plntEPSWT20139-1', 'plntEPSWT20139-2', 'plntEPSWT20139-10',\n",
       "        'plntEPSWT20139-3', 'plntEPSWT20139-4', 'plntEPSWT20139-5',\n",
       "        'plntEPSWT20139-6', 'plntEPSWT20139-7', 'plntEPSWT20139-8',\n",
       "        'plntEPSWT20139-9', 'plntEPSWT20143-1', 'plntEPSWT20143-10',\n",
       "        'plntEPSWT20143-2', 'plntEPSWT20143-3', 'plntEPSWT20143-4',\n",
       "        'plntEPSWT20143-5', 'plntEPSWT20143-6', 'plntEPSWT20143-7',\n",
       "        'plntEPSWT20143-8', 'plntEPSWT20143-9', 'plntEPSWT20154-1',\n",
       "        'plntEPSWT20154-2', 'plntEPSWT20154-4', 'plntEPSWT20154-10',\n",
       "        'plntEPSWT20154-9', 'plntEPSWT20156-1', 'plntEPSWT20156-10',\n",
       "        'plntEPSWT20156-2', 'plntEPSWT20156-3', 'plntEPSWT20156-4',\n",
       "        'plntEPSWT20156-5', 'plntEPSWT20156-6', 'plntEPSWT20156-7',\n",
       "        'plntEPSWT20156-8', 'plntEPSWT20156-9', 'plntEPSWT20165-1',\n",
       "        'plntEPSWT20165-10', 'plntEPSWT20165-2', 'plntEPSWT20165-3',\n",
       "        'plntEPSWT20165-4', 'plntEPSWT20165-5', 'plntEPSWT20165-6',\n",
       "        'plntEPSWT20165-7', 'plntEPSWT20165-8', 'plntEPSWT20165-9',\n",
       "        'plntEPSWT20187-1', 'plntEPSWT20187-10', 'plntEPSWT20187-2',\n",
       "        'plntEPSWT20187-3', 'plntEPSWT20187-4', 'plntEPSWT20187-5',\n",
       "        'plntEPSWT20187-6', 'plntEPSWT20187-7', 'plntEPSWT20187-9',\n",
       "        'plntEPSWT20198-1', 'plntEPSWT20198-10', 'plntEPSWT20198-2',\n",
       "        'plntEPSWT20198-3', 'plntEPSWT20198-4', 'plntEPSWT20198-6',\n",
       "        'plntEPSWT20198-7', 'plntEPSWT20198-8', 'plntEPSWT20198-9',\n",
       "        'plntEPSWT20199-1', 'plntEPSWT20199-2', 'plntEPSWT20199-3',\n",
       "        'plntEPSWT20199-5', 'plntEPSWT20199-7', 'plntEPSWT20199-8',\n",
       "        'plntEPSWT20199-9', 'plntEPSWT20056-1', 'plntEPSWT20056-10',\n",
       "        'plntEPSWT20056-2', 'plntEPSWT20056-3', 'plntEPSWT20056-4',\n",
       "        'plntEPSWT20056-5', 'plntEPSWT20056-6', 'plntEPSWT20056-7',\n",
       "        'plntEPSWT20056-8', 'plntEPSWT20056-9', 'plntEPSWT20201-9',\n",
       "        'plntEPSWT20071-2', 'plntEPSWT20071-3', 'plntEPSWT20071-4',\n",
       "        'plntEPSWT20071-6', 'plntEPSWT20071-8', 'plntEPSWT20071-9',\n",
       "        'plntEPSWT20072-9', 'plntEPSWT20075-1', 'plntEPSWT20128-1',\n",
       "        'plntEPSWT20128-10', 'plntEPSWT20128-3', 'plntEPSWT20128-4',\n",
       "        'plntEPSWT20128-6', 'plntEPSWT20128-7', 'plntEPSWT20128-8',\n",
       "        'plntEPSWT20128-9', 'plntEPSWT20130-1', 'plntEPSWT20130-10',\n",
       "        'plntEPSWT20130-11', 'plntEPSWT20130-12', 'plntEPSWT20130-13',\n",
       "        'plntEPSWT20130-14', 'plntEPSWT20130-2', 'plntEPSWT20130-3',\n",
       "        'plntEPSWT20130-4', 'plntEPSWT20130-5', 'plntEPSWT20130-6',\n",
       "        'plntEPSWT20130-7', 'plntEPSWT20130-8', 'plntEPSWT20130-9',\n",
       "        'plntEPSWT20138-1', 'plntEPSWT20138-2', 'plntEPSWT20138-4',\n",
       "        'plntEPSWT20138-5', 'plntEPSWT20138-8', 'plntEPSWT20138-9',\n",
       "        'plntEPSWT20146-1', 'plntEPSWT20146-10', 'plntEPSWT20146-2',\n",
       "        'plntEPSWT20146-3', 'plntEPSWT20146-4', 'plntEPSWT20146-5',\n",
       "        'plntEPSWT20146-6', 'plntEPSWT20146-7', 'plntEPSWT20146-8',\n",
       "        'plntEPSWT20146-9', 'plntEPSWT20147-2', 'plntEPSWT20147-3',\n",
       "        'plntEPSWT20147-6', 'plntEPSWT20147-7', 'plntEPSWT20147-8',\n",
       "        'plntEPSWT20147-9', 'plntEPSWT20176-1', 'plntEPSWT20176-3',\n",
       "        'plntEPSWT20176-4', 'plntEPSWT20199-10', 'plntEPSWT20199-4',\n",
       "        'plntEPSWT20199-6', 'plntEPSWT20127-10', 'plntEPSWT20127-3',\n",
       "        'plntEPSWT20127-4', 'plntEPSWT20127-5', 'plntEPSWT20127-6',\n",
       "        'plntEPSWT20127-7', 'plntEPSWT20127-8', 'plntEPSWT20127-9',\n",
       "        'plntEPSWT20131-1', 'plntEPSWT20131-10', 'plntEPSWT20131-3',\n",
       "        'plntEPSWT20131-4', 'plntEPSWT20131-7', 'plntEPSWT20134-1',\n",
       "        'plntEPSWT20134-10', 'plntEPSWT20134-2', 'plntEPSWT20134-3',\n",
       "        'plntEPSWT20134-4', 'plntEPSWT20134-5', 'plntEPSWT20134-6',\n",
       "        'plntEPSWT20134-7', 'plntEPSWT20134-8', 'plntEPSWT20134-9',\n",
       "        'plntEPSWT20144-1', 'plntEPSWT20144-10', 'plntEPSWT20144-2',\n",
       "        'plntEPSWT20144-3', 'plntEPSWT20144-5', 'plntEPSWT20144-6',\n",
       "        'plntEPSWT20144-7', 'plntEPSWT20144-8', 'plntEPSWT20154-3',\n",
       "        'plntEPSWT20154-5', 'plntEPSWT20154-8', 'plntEPSWT20158-10',\n",
       "        'plntEPSWT20158-2', 'plntEPSWT20158-5', 'plntEPSWT20158-6',\n",
       "        'plntEPSWT20158-7', 'plntEPSWT20158-8', 'plntEPSWT20061-1',\n",
       "        'plntEPSWT20061-10', 'plntEPSWT20061-2', 'plntEPSWT20061-3',\n",
       "        'plntEPSWT20061-6', 'plntEPSWT20061-7', 'plntEPSWT20061-8',\n",
       "        'plntEPSWT20061-9', 'plntEPSWT20064-1', 'plntEPSWT20064-4',\n",
       "        'plntEPSWT20067-7', 'plntEPSWT20069-1', 'plntEPSWT20069-2',\n",
       "        'plntEPSWT20069-3', 'plntEPSWT20069-5', 'plntEPSWT20069-7',\n",
       "        'plntEPSWT20069-8', 'plntEPSWT20069-9', 'plntEPSWT20072-1',\n",
       "        'plntEPSWT20072-2', 'plntEPSWT20072-5', 'plntEPSWT20072-8',\n",
       "        'plntEPSWT20075-10', 'plntEPSWT20075-3', 'plntEPSWT20075-4',\n",
       "        'plntEPSWT20075-9', 'plntEPSWT20076-10', 'plntEPSWT20076-7',\n",
       "        'plntEPSWT20077-1', 'plntEPSWT20077-10', 'plntEPSWT20077-2',\n",
       "        'plntEPSWT20077-4', 'plntEPSWT20077-5', 'plntEPSWT20077-7',\n",
       "        'plntEPSWT20077-9', 'plntEPSWT20125-7', 'plntEPSWT20126-7',\n",
       "        'plntEPSWT20136-1', 'plntEPSWT20136-10', 'plntEPSWT20136-2',\n",
       "        'plntEPSWT20136-3', 'plntEPSWT20136-4', 'plntEPSWT20136-5',\n",
       "        'plntEPSWT20136-6', 'plntEPSWT20136-7', 'plntEPSWT20136-8',\n",
       "        'plntEPSWT20136-9', 'plntEPSWT20138-7', 'plntEPSWT20141-1',\n",
       "        'plntEPSWT20141-2', 'plntEPSWT20141-4', 'plntEPSWT20141-6',\n",
       "        'plntEPSWT20141-9', 'plntEPSWT20142-1', 'plntEPSWT20142-10',\n",
       "        'plntEPSWT20142-2', 'plntEPSWT20142-3', 'plntEPSWT20142-4',\n",
       "        'plntEPSWT20142-5', 'plntEPSWT20142-6', 'plntEPSWT20142-7',\n",
       "        'plntEPSWT20142-8', 'plntEPSWT20145-10', 'plntEPSWT20145-3',\n",
       "        'plntEPSWT20145-5', 'plntEPSWT20147-10', 'plntEPSWT20167-1',\n",
       "        'plntEPSWT20167-10', 'plntEPSWT20167-2', 'plntEPSWT20167-3',\n",
       "        'plntEPSWT20167-4', 'plntEPSWT20167-5', 'plntEPSWT20167-6',\n",
       "        'plntEPSWT20167-7', 'plntEPSWT20167-8', 'plntEPSWT20167-9',\n",
       "        'plntEPSWT20197-10', 'plntEPSWT20197-2', 'plntEPSWT20197-4',\n",
       "        'plntEPSWT20197-6', 'plntEPSWT20197-7', 'plntEPSWT20197-8'],\n",
       "       dtype=object),\n",
       "      chr          plant_ID platename  library first_reaction_primers  \\\n",
       " 0      4  plntEPSWT20057-1    p1ARF9     1017            SW346_SW442   \n",
       " 1      4  plntEPSWT20057-1    p1ARF9     1017            SW346_SW442   \n",
       " 2      4  plntEPSWT20057-1    p1ARF9     1017            SW346_SW442   \n",
       " 3      4  plntEPSWT20057-1    p1ARF9     1017            SW346_SW442   \n",
       " 4      4  plntEPSWT20057-1    p1ARF9     1017            SW346_SW442   \n",
       " ..   ...               ...       ...      ...                    ...   \n",
       " 370    4  plntEPSWT20056-9    p6NLP7     1022            SW349_SW445   \n",
       " 371    4  plntEPSWT20056-9    p6NLP7     1022            SW349_SW445   \n",
       " 372    4  plntEPSWT20056-9    p6NLP7     1022            SW349_SW445   \n",
       " 373    4  plntEPSWT20056-9    p6NLP7     1022            SW349_SW445   \n",
       " 374    4  plntEPSWT20056-9    p6NLP7     1022            SW349_SW445   \n",
       " \n",
       "     second_reaction_primers         guide  guide_number  \\\n",
       " 0                ARF9_1F_8R   ARF9_guide6             6   \n",
       " 1                ARF9_1F_8R   ARF9_guide6             6   \n",
       " 2                ARF9_1F_8R   ARF9_guide7             7   \n",
       " 3                ARF9_1F_8R   ARF9_guide7             7   \n",
       " 4                ARF9_1F_8R   ARF9_guide8             8   \n",
       " ..                      ...           ...           ...   \n",
       " 370              NLP7_4F_6R   NLP7_guide5             5   \n",
       " 371              NLP7_4F_6R   NLP7_guide6             6   \n",
       " 372              NLP7_4F_6R  NLP7_guide10            10   \n",
       " 373              NLP7_4F_6R  NLP7_guide24            24   \n",
       " 374              NLP7_4F_6R  NLP7_guide25            25   \n",
       " \n",
       "                              aligned_sequence  \\\n",
       " 0    ATCGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTGA   \n",
       " 1    AACGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTGA   \n",
       " 2    AATCGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTG   \n",
       " 3    AAACGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTG   \n",
       " 4    GAAACGGAGATAAAAGCAAATCGGAGGTGTTAACTTAAAG   \n",
       " ..                                        ...   \n",
       " 370  GGAAAGAGATCAACCAACGATTAGAGGAGAAGAAGAACTT   \n",
       " 371  CGGAAAGAGATCAACCAACGATTAGAGGAGAAGAAGAACT   \n",
       " 372  AGTTGTATTGTGAAGCAGAGTAATCGGTTCTCAATATATT   \n",
       " 373  GATTTGTCTTTTCTCCACTTACTCAATTTCCCATTAATAA   \n",
       " 374  AGAATCAGGTAGGTCCAGGATTTGTCTTTTCTCCACTTAC   \n",
       " \n",
       "                            reference_sequence  ...  \\\n",
       " 0    A-CGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTGA  ...   \n",
       " 1    A-CGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTGA  ...   \n",
       " 2    AA-CGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTG  ...   \n",
       " 3    AA-CGGAGGTGTTAACTTAAAGACGGGAGAGAAGCGAGTG  ...   \n",
       " 4    GAAACGGAGATAAAAGCAAA-CGGAGGTGTTAACTTAAAG  ...   \n",
       " ..                                        ...  ...   \n",
       " 370  GGAAAGAGATCAACCAACGA-TAGAGGAGAAGAAGAACTT  ...   \n",
       " 371  CGGAAAGAGATCAACCAACGA-TAGAGGAGAAGAAGAACT  ...   \n",
       " 372  AGTTGTATTGTGAAGCAGAG-AATCGGTTCTCAATATATT  ...   \n",
       " 373  GATTTGTCTTTTCTCCACTT-CTCAATTTCCCATTAATAA  ...   \n",
       " 374  AGAATCAGGTAGGTCCAGGATTTGTCTTTTCTCCACTT-C  ...   \n",
       " \n",
       "     deletion_overlapping_TFBS_family deletion_overlapping_TFBS_AGI  \\\n",
       " 0                                NaN                           NaN   \n",
       " 1                                NaN                           NaN   \n",
       " 2                                NaN                           NaN   \n",
       " 3                                NaN                           NaN   \n",
       " 4                                NaN                           NaN   \n",
       " ..                               ...                           ...   \n",
       " 370                              NaN                           NaN   \n",
       " 371                              NaN                           NaN   \n",
       " 372                              NaN                           NaN   \n",
       " 373                              NaN                           NaN   \n",
       " 374                              NaN                           NaN   \n",
       " \n",
       "      substitution_overlapping_TFBS_family  substitution_overlapping_TFBS_AGI  \\\n",
       " 0                                     NaN                                NaN   \n",
       " 1                                     NaN                                NaN   \n",
       " 2                                     NaN                                NaN   \n",
       " 3                                     NaN                                NaN   \n",
       " 4                                     NaN                                NaN   \n",
       " ..                                    ...                                ...   \n",
       " 370                                   NaN                                NaN   \n",
       " 371                                   NaN                                NaN   \n",
       " 372                                   NaN                                NaN   \n",
       " 373                                   NaN                                NaN   \n",
       " 374                                   NaN                                NaN   \n",
       " \n",
       "     number_of_different_alleles number_of_different_non_wt_alleles  \\\n",
       " 0                             2                                  2   \n",
       " 1                             2                                  2   \n",
       " 2                             2                                  2   \n",
       " 3                             2                                  2   \n",
       " 4                             2                                  2   \n",
       " ..                          ...                                ...   \n",
       " 370                           2                                  1   \n",
       " 371                           2                                  1   \n",
       " 372                           2                                  1   \n",
       " 373                           1                                  1   \n",
       " 374                           1                                  1   \n",
       " \n",
       "                                    mutated_guide_sites  \\\n",
       " 0    ['ARF9_guide10', 'ARF9_guide11', 'ARF9_guide6'...   \n",
       " 1    ['ARF9_guide10', 'ARF9_guide11', 'ARF9_guide6'...   \n",
       " 2    ['ARF9_guide10', 'ARF9_guide11', 'ARF9_guide6'...   \n",
       " 3    ['ARF9_guide10', 'ARF9_guide11', 'ARF9_guide6'...   \n",
       " 4    ['ARF9_guide10', 'ARF9_guide11', 'ARF9_guide6'...   \n",
       " ..                                                 ...   \n",
       " 370  ['NLP7_guide10', 'NLP7_guide24', 'NLP7_guide25...   \n",
       " 371  ['NLP7_guide10', 'NLP7_guide24', 'NLP7_guide25...   \n",
       " 372  ['NLP7_guide10', 'NLP7_guide24', 'NLP7_guide25...   \n",
       " 373  ['NLP7_guide10', 'NLP7_guide24', 'NLP7_guide25...   \n",
       " 374  ['NLP7_guide10', 'NLP7_guide24', 'NLP7_guide25...   \n",
       " \n",
       "                                   approved_guide_pairs  \\\n",
       " 0    [('ARF9_guide11', 'ARF9_guide7'), ('ARF9_guide...   \n",
       " 1    [('ARF9_guide11', 'ARF9_guide7'), ('ARF9_guide...   \n",
       " 2    [('ARF9_guide11', 'ARF9_guide7'), ('ARF9_guide...   \n",
       " 3    [('ARF9_guide11', 'ARF9_guide7'), ('ARF9_guide...   \n",
       " 4    [('ARF9_guide11', 'ARF9_guide7'), ('ARF9_guide...   \n",
       " ..                                                 ...   \n",
       " 370  [('NLP7_guide10', 'NLP7_guide6'), ('NLP7_guide...   \n",
       " 371  [('NLP7_guide10', 'NLP7_guide6'), ('NLP7_guide...   \n",
       " 372  [('NLP7_guide10', 'NLP7_guide6'), ('NLP7_guide...   \n",
       " 373  [('NLP7_guide10', 'NLP7_guide6'), ('NLP7_guide...   \n",
       " 374  [('NLP7_guide10', 'NLP7_guide6'), ('NLP7_guide...   \n",
       " \n",
       "                   guides_not_in_pairs estimated_agro_strain_number  \n",
       " 0     ['ARF9_guide10', 'ARF9_guide9']                            4  \n",
       " 1     ['ARF9_guide10', 'ARF9_guide9']                            4  \n",
       " 2     ['ARF9_guide10', 'ARF9_guide9']                            4  \n",
       " 3     ['ARF9_guide10', 'ARF9_guide9']                            4  \n",
       " 4     ['ARF9_guide10', 'ARF9_guide9']                            4  \n",
       " ..                                ...                          ...  \n",
       " 370  ['NLP7_guide24', 'NLP7_guide25']                            3  \n",
       " 371  ['NLP7_guide24', 'NLP7_guide25']                            3  \n",
       " 372  ['NLP7_guide24', 'NLP7_guide25']                            3  \n",
       " 373  ['NLP7_guide24', 'NLP7_guide25']                            3  \n",
       " 374  ['NLP7_guide24', 'NLP7_guide25']                            3  \n",
       " \n",
       " [1723 rows x 39 columns])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_mutations(genotyped_df_all,genotyped_only_mutated_df_all,genotyped_only_mutated_flattened_df_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('pybedtools')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2879a2396fb26a2192d4c546fae66e67014459dff7902256699e5c7228156d44"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
