{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ast import literal_eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_ARF9_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_ARF9_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_ARF9_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_ARF18_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_filtered.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "mutations_ARF18_genotyped_only_mutated_flattened_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_filtered/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_DREB26_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_DREB26_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_DREB26_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "mutations_NLP7_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped.tsv'\n",
    "mutations_NLP7_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "mutations_NLP7_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in files\n",
    "def read_files(genotyped_file,genotyped_only_mutated_file,genotyped_only_mutated_flattened_file):\n",
    "    \"\"\"read in the files to pandas\"\"\"\n",
    "    #read in files\n",
    "    genotyped_df = pd.read_table(genotyped_file,sep='\\t', header=0)\n",
    "    genotyped_only_mutated_df = pd.read_table(genotyped_only_mutated_file,sep='\\t', header=0)\n",
    "    genotyped_only_mutated_flattened_df = pd.read_table(genotyped_only_mutated_flattened_file,sep='\\t', header=0)\n",
    "    return genotyped_df,genotyped_only_mutated_df,genotyped_only_mutated_flattened_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_number_mutations_overlapping_at_least_one_TFBS(df, mutation_type):\n",
    "    \"\"\"function to count the number of mutations in the df which overlap at least one TFBS\"\"\"\n",
    "    #turn df into a dict\n",
    "    df_dict = df.to_dict(orient=\"records\")\n",
    "    mutations_overlapping_at_least_one_TFBS_count = 0\n",
    "    if mutation_type == 'insertion':\n",
    "        for row_dict in df_dict:\n",
    "                list_lists_of_overlapping_AGIs = []\n",
    "                genomic_positions = literal_eval(row_dict[f'{mutation_type}_genomic_positions_combined'])\n",
    "                #iterate over insertions. Some insertions are linger than 1 bp so the genomic position pairs need combining\n",
    "                #remaining pairs\n",
    "                remaining_pairs = []\n",
    "                for size in literal_eval(row_dict[f'{mutation_type}_sizes']):\n",
    "                                \n",
    "                    size = int(size)\n",
    "                    #select pairs up to the size of the insertion\n",
    "                    \n",
    "                    #pop out the first n pairs from the genomic positions list, where n is the size of the insertion - 1\n",
    "                    #if remaining pairs not empty, start with that\n",
    "                    if len(remaining_pairs)>0:\n",
    "                        #print('not []')\n",
    "                        current_pairs = remaining_pairs[:size]\n",
    "                    #else create the remaining pairs variable\n",
    "                    elif len(remaining_pairs)==0:\n",
    "                        #print('is []')\n",
    "                        current_pairs = genomic_positions[:size]\n",
    "                        remaining_pairs = genomic_positions[size:]          \n",
    "                    #print(f'currentpairs={current_pairs}remaining={remaining_pairs}')\n",
    "\n",
    "                    current_pairs_list_of_overlapping = []\n",
    "                    #iterate over pairs in the mutation_genomic_positions_combined list #evaluate the string representing the pair as if they are literals ie. a list\n",
    "                    for pair in current_pairs:\n",
    "                        #print(f'pair={pair}')\n",
    "                        list_of_AGIs = []\n",
    "                        list_of_overlapping_AGIs = []\n",
    "                        #for genomic position in pair \n",
    "                        #print(pair)\n",
    "                        for pos in pair:\n",
    "                            #make the genomic position string\n",
    "                            pos = str(pos)\n",
    "                            #if genomic position is in the mutation_overlapping_TFBS_AGI_combined dictionary\n",
    "                            if pos in literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined']):\n",
    "                                #print(pos)\n",
    "                                #append to list_of_AGIs\n",
    "                                mutation_overlapping_TFBS_AGI_combined = literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined'])\n",
    "                            \n",
    "                                #print(pos)\n",
    "                                AGI = mutation_overlapping_TFBS_AGI_combined[pos]\n",
    "                                list_of_AGIs += AGI\n",
    "                        #if empty list, pass\n",
    "                        \n",
    "                        for elem in list_of_AGIs:\n",
    "                            #print(elem)\n",
    "                            #if TFBS is only overlapping one of the genomic positions in the pair, pass - we only want to include TFBSs that overlapping both positions so that the insertion is totally overlapping\n",
    "                            if list_of_AGIs.count(elem) == 1:\n",
    "                                pass\n",
    "                            elif list_of_AGIs.count(elem) > 1:\n",
    "                                list_of_overlapping_AGIs.append(elem)\n",
    "                        #append to list of lists if lists_of_overlapping_AGIs is not empty\n",
    "                        #remove duplicates from list_of_overlapping_AGIs\n",
    "                        list_of_overlapping_AGIs_set = set(list_of_overlapping_AGIs)\n",
    "                        list_of_overlapping_AGIs = list(list_of_overlapping_AGIs_set)\n",
    "                        if len(list_of_overlapping_AGIs) > 0:\n",
    "                            #print(list_of_overlapping_AGIs)\n",
    "                            #get unique AGIs\n",
    "                            #list_of_overlapping_AGIs_unique\n",
    "                            current_pairs_list_of_overlapping+=list_of_overlapping_AGIs\n",
    "                    #get unique elements\n",
    "                    current_pairs_list_of_overlapping_set = set(current_pairs_list_of_overlapping)\n",
    "                    current_pairs_list_of_overlapping = list(current_pairs_list_of_overlapping_set)\n",
    "                    #if current_pairs_list_of_overlapping list isn't empty, append to list_lists_of_overlapping_AGIs \n",
    "                    if len(current_pairs_list_of_overlapping) > 0:\n",
    "                        #print(current_pairs_list_of_overlapping)\n",
    "                        #add one to the count of mutations which overlap at least one TFBS\n",
    "                        mutations_overlapping_at_least_one_TFBS_count+=1\n",
    "                        #append current_pairs_list_of_overlapping of TFBSs to list_lists_of_overlapping_AGIs\n",
    "                        list_lists_of_overlapping_AGIs.append(current_pairs_list_of_overlapping)\n",
    "    else:\n",
    "        for row_dict in df_dict:\n",
    "            list_lists_of_overlapping_AGIs = []\n",
    "            genomic_positions = literal_eval(row_dict[f'{mutation_type}_genomic_positions_combined'])\n",
    "            #iterate over insertions. Some insertions are linger than 1 bp so the genomic position pairs need combining\n",
    "            #remaining positions\n",
    "            remaining_positions = []\n",
    "            for size in literal_eval(row_dict[f'{mutation_type}_sizes']):\n",
    "                            \n",
    "                size = int(size)\n",
    "                #select positions up to the size of the insertion\n",
    "                \n",
    "                #pop out the first n positions from the genomic positions list, where n is the size of the insertion - 1\n",
    "                #if remaining positions not empty, start with that\n",
    "                if len(remaining_positions)>0:\n",
    "                    current_positions = remaining_positions[:size]\n",
    "                    #else create the remaining pairs variable\n",
    "                elif len(remaining_positions)==0:\n",
    "                    current_positions = genomic_positions[:size]\n",
    "                    remaining_positions = genomic_positions[size:]\n",
    "                list_of_overlapping_AGIs = []\n",
    "                #for genomic position in pair \n",
    "                #print(pair)\n",
    "                current_positions_list_of_overlapping = []\n",
    "                for pos in current_positions:\n",
    "                    #make the genomic position string\n",
    "                    pos = str(pos)\n",
    "                    #if genomic position is in the mutation_overlapping_TFBS_AGI_combined dictionary\n",
    "                    if pos in literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined']):\n",
    "                        #print(pos)\n",
    "                        #append to list_of_overlapping_AGIs\n",
    "                        mutation_overlapping_TFBS_AGI_combined = literal_eval(row_dict[f'{mutation_type}_overlapping_TFBS_AGI_combined'])\n",
    "                    \n",
    "                        #print(pos)\n",
    "                        AGI = mutation_overlapping_TFBS_AGI_combined[pos]\n",
    "                        list_of_overlapping_AGIs += AGI                     \n",
    "                        #append to list of lists if lists_of_overlapping_AGIs is not empty\n",
    "                #remove duplicates from list_of_overlapping_AGIs\n",
    "                list_of_overlapping_AGIs_set = set(list_of_overlapping_AGIs)\n",
    "                list_of_overlapping_AGIs = list(list_of_overlapping_AGIs_set)\n",
    "                if len(list_of_overlapping_AGIs) > 0:\n",
    "                    #print(list_of_overlapping_AGIs)\n",
    "                    #get unique AGIs\n",
    "                    #list_of_overlapping_AGIs_unique\n",
    "                    current_positions_list_of_overlapping+=list_of_overlapping_AGIs\n",
    "                #get unique elements\n",
    "                current_positions_list_of_overlapping_set = set(current_positions_list_of_overlapping)\n",
    "                current_positions_list_of_overlapping = list(current_positions_list_of_overlapping_set)\n",
    "                #if current_positions_list_of_overlapping list isn't empty, append to list_lists_of_overlapping_AGIs \n",
    "                if len(current_positions_list_of_overlapping) > 0:\n",
    "                    #add one to the count of mutations which overlap at least one TFBS\n",
    "                    mutations_overlapping_at_least_one_TFBS_count+=1\n",
    "                    #append current_positions_list_of_overlapping of TFBSs to list_lists_of_overlapping_AGIs\n",
    "                    list_lists_of_overlapping_AGIs.append(current_positions_list_of_overlapping)\n",
    "\n",
    "    return mutations_overlapping_at_least_one_TFBS_count\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_mutations(genotyped_df,genotyped_only_mutated_df,genotyped_only_mutated_flattened_df):\n",
    "    \"\"\"function to count number of genotyped plant lines and also number of plant lines containing mutations, and then number of mutations overall\"\"\"\n",
    "    #count number of unique plant IDs\n",
    "    genotyped_counts = genotyped_df.plant_ID.unique()\n",
    "    genotyped_only_mutated_counts = genotyped_only_mutated_df.plant_ID.unique()\n",
    "    #print counts\n",
    "    #total mutations\n",
    "    #total_mutations = genotyped_only_mutated_flattened_df\n",
    "    print(f'total plant lines sequenced={len(genotyped_counts)} \\ntotal mutated plant lines={len(genotyped_only_mutated_counts)}')#\\ntotal mutations={len(genotyped_only_mutated_df)}\n",
    "    #count number of chimeric plant lines\n",
    "    # unique_line_mutated = genotyped_only_mutated_df.drop_duplicates(subset='plant_ID',keep='first')\n",
    "    \n",
    "    unique_line_mutated = genotyped_only_mutated_flattened_df\n",
    "    non_chimeric = unique_line_mutated[unique_line_mutated.genotype != 'chimeric']\n",
    "    homozygous = unique_line_mutated[unique_line_mutated.genotype == 'homozygous']\n",
    "    heterozygous = unique_line_mutated[unique_line_mutated.genotype == 'heterozygous']\n",
    "    biallelic = unique_line_mutated[unique_line_mutated.genotype == 'biallelic']\n",
    "\n",
    "    #print bold heading\n",
    "    print('\\033[1m'+'Genotypes'+'\\033[0m')\n",
    "    print(f'number of non-chimeric mutated plant lines = {len(non_chimeric)} ({len(non_chimeric)/len(genotyped_only_mutated_counts)*100}% of non-chimeric lines)')\n",
    "    print(f'number of homozygous mutated plant lines = {len(homozygous)} ({len(homozygous)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    print(f'number of heterozygous mutated plant lines = {len(heterozygous)} ({len(heterozygous)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    print(f'number of biallelic mutated plant lines = {len(biallelic)} ({len(biallelic)/len(non_chimeric)*100}% of non-chimeric lines)')\n",
    "    #count number of non-chimeric insertions, deletions, substitutions (if two insertions around same guide, counted as once)\n",
    "    non_chimeric_all = non_chimeric.copy()\n",
    "    insertions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('insertion')]\n",
    "    deletions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('deletion')]\n",
    "    substitutions = non_chimeric_all[non_chimeric_all.mutation_type.str.contains('substitution')]\n",
    "    #number of insertions. evaluate the strings as if they are literals (as if they are lists)\n",
    "    number_of_insertions = len(insertions.insertion_sizes.apply(literal_eval).sum())\n",
    "    #print(np.concatenate(insertions.insertion_genomic_positions_combined)\n",
    "    #number of deletions\n",
    "    all_deletion_lengths = deletions.deletion_sizes.apply(literal_eval).sum()\n",
    "    number_of_deletions = len(all_deletion_lengths)\n",
    "    #number of substitutions\n",
    "    number_of_substitutions = len(substitutions.substitution_sizes.apply(literal_eval).sum())\n",
    "\n",
    "\n",
    "    median_reads_guide_sites = genotyped_df.read_number.median()\n",
    "    median_mutation_reads_guide_sites = genotyped_only_mutated_df.read_number.median()\n",
    "    #print bold heading\n",
    "    #filter out deletion values less than 10\n",
    "    deletions_over_10bp = list(filter(lambda x: x > 10, all_deletion_lengths))\n",
    "    total_mutations = number_of_insertions + len(deletions_over_10bp) + number_of_substitutions\n",
    "    print('\\033[1m'+'Mutation types'+'\\033[0m')\n",
    "    print(f'total mutations = {total_mutations}')\n",
    "    print(f'number of insertions = {number_of_insertions} ({number_of_insertions/total_mutations*100}% of total mutations)')\n",
    "    print(f'number of deletions = {number_of_deletions} ({number_of_deletions/total_mutations*100}% of total mutations)')\n",
    "    #deletions over 10bp\n",
    "    #deletions.deletion_positions.str.len()\n",
    "    #deletions_over_10bp = deletions[deletions.deletion_position.str.len()>10]\n",
    "    \n",
    "    #print(all_deletion_lengths)\n",
    "    \n",
    "    #print(deletions_over_10bp)\n",
    "    print(f'deletions over 10bp = {len(deletions_over_10bp)} ({len(deletions_over_10bp)/number_of_deletions*100}% of deletions)')\n",
    "    print(f'number of substitutions = {number_of_substitutions} ({number_of_substitutions/total_mutations*100}% of total mutations)')\n",
    "    print(f'median reads at guide sites including wt = {median_reads_guide_sites}')\n",
    "    print(f'median reads at guide sites mutations only = {median_mutation_reads_guide_sites}')\n",
    "    \n",
    "    #find number of mutations overlapping FIMO predicted TFBS\n",
    "    insertions_overlapping_TFBS = insertions[insertions.insertion_overlapping_TFBS_AGI_combined.notna()]\n",
    "    deletions_overlapping_TFBS = deletions[deletions.deletion_overlapping_TFBS_AGI_combined.notna()]\n",
    "    substitutions_overlapping_TFBS = substitutions[substitutions.substitution_overlapping_TFBS_AGI_combined.notna()]\n",
    "    #count number of insertions overlapping TFBSs\n",
    "    insertions_count = count_number_mutations_overlapping_at_least_one_TFBS(insertions_overlapping_TFBS, 'insertion')\n",
    "    deletions_count = count_number_mutations_overlapping_at_least_one_TFBS(deletions_overlapping_TFBS, 'deletion')\n",
    "    substitutions_count = count_number_mutations_overlapping_at_least_one_TFBS(substitutions_overlapping_TFBS, 'substitution')\n",
    "          \n",
    "\n",
    "    #print bold heading\n",
    "    print('\\033[1m'+'Overlapping TFBSs'+'\\033[0m')\n",
    "    print(f'number of insertions overlapping TFBS = {insertions_count} ({insertions_count/number_of_insertions*100}% of insertions)')\n",
    "    print(f'number of deletions overlapping TFBS = {deletions_count} ({deletions_count/number_of_deletions*100}% of insertions)')\n",
    "    print(f'number of substitutions overlapping TFBS = {substitutions_count} ({substitutions_count/number_of_substitutions*100}% of insertions)')\n",
    "\n",
    "    #how many within 7bp of cutsite\n",
    "    #insertions_7bp = insertions[[[i for i in x if int(i) <8] for x in insertions.insertion_cut_site_distance.str.strip('][').str.split(', ')]]\n",
    "\n",
    "   # print(insertions_7bp)\n",
    "\n",
    "\n",
    "    return genotyped_counts, genotyped_only_mutated_counts, genotyped_only_mutated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARF9_genotyped_df, ARF9_mutated_df,ARF9_flattened_df = read_files(mutations_ARF9_genotyped,mutations_ARF9_genotyped_only_mutated,mutations_ARF9_genotyped_only_mutated_flattened)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=468 \n",
      "total mutated plant lines=147\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 142 (96.5986394557823% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 92 (64.7887323943662% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 12 (8.450704225352112% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 38 (26.76056338028169% of non-chimeric lines)\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 195\n",
      "number of insertions = 182 (93.33333333333333% of total mutations)\n",
      "number of deletions = 50 (25.64102564102564% of total mutations)\n",
      "deletions over 10bp = 9 (18.0% of deletions)\n",
      "number of substitutions = 4 (2.051282051282051% of total mutations)\n",
      "median reads at guide sites including wt = 361.0\n",
      "median reads at guide sites mutations only = 234.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 34 (18.681318681318682% of insertions)\n",
      "number of deletions overlapping TFBS = 15 (30.0% of insertions)\n",
      "number of substitutions overlapping TFBS = 0 (0.0% of insertions)\n"
     ]
    }
   ],
   "source": [
    "ARF9_genotyped_plant_lines, ARF9_mutated_plant_lines,ARF9_mutations = count_mutations(ARF9_genotyped_df,ARF9_mutated_df,ARF9_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter ARF18 guide 14 since very repetitive and lots of sequencing errors\n",
    "ARF18_mutated_df = pd.read_table(mutations_ARF18_genotyped_only_mutated,sep='\\t', header=0)\n",
    "ARF18_mutated_plant_lines_filtered = ARF18_mutated_df[ARF18_mutated_df.guide != 'ARF18_guide14']\n",
    "#save df\n",
    "ARF18_mutated_plant_lines_filtered.to_csv(mutations_ARF18_genotyped_only_mutated_filtered, sep='\\t',header=True)\n",
    "#Then I ran pacbio_analyse_flattened_mutations.py on the filtered .tsv to generate the mutations_ARF18_genotyped_only_mutated_flattened_filtered file (f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_filtered/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARF18_genotyped_df, ARF18_mutated_df,ARF18_flattened_filtered_df = read_files(mutations_ARF18_genotyped,mutations_ARF18_genotyped_only_mutated,mutations_ARF18_genotyped_only_mutated_flattened_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#filter ARF18 guide 14 since very repetitive and lots of sequencing errors\n",
    "ARF18_genotyped_only_mutated_df = pd.read_table(mutations_ARF18_genotyped_only_mutated,sep='\\t', header=0)\n",
    "ARF18_mutated_plant_lines_filtered = ARF18_mutated_df[ARF18_mutated_df.guide != 'ARF18_guide14']\n",
    "#save as df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=373 \n",
      "total mutated plant lines=67\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 66 (98.50746268656717% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 27 (40.909090909090914% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 25 (37.878787878787875% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 14 (21.21212121212121% of non-chimeric lines)\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 98\n",
      "number of insertions = 87 (88.77551020408163% of total mutations)\n",
      "number of deletions = 11 (11.224489795918368% of total mutations)\n",
      "deletions over 10bp = 3 (27.27272727272727% of deletions)\n",
      "number of substitutions = 8 (8.16326530612245% of total mutations)\n",
      "median reads at guide sites including wt = 623.0\n",
      "median reads at guide sites mutations only = 310.5\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 53 (60.91954022988506% of insertions)\n",
      "number of deletions overlapping TFBS = 7 (63.63636363636363% of insertions)\n",
      "number of substitutions overlapping TFBS = 0 (0.0% of insertions)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#count mutations\n",
    "ARF18_genotyped_plant_lines, ARF18_mutated_plant_lines,ARF18_mutations = count_mutations(ARF18_genotyped_df,ARF18_mutated_plant_lines_filtered,ARF18_flattened_filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=366 \n",
      "total mutated plant lines=43\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 40 (93.02325581395348% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 15 (37.5% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 19 (47.5% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 6 (15.0% of non-chimeric lines)\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 53\n",
      "number of insertions = 40 (75.47169811320755% of total mutations)\n",
      "number of deletions = 22 (41.509433962264154% of total mutations)\n",
      "deletions over 10bp = 2 (9.090909090909092% of deletions)\n",
      "number of substitutions = 11 (20.754716981132077% of total mutations)\n",
      "median reads at guide sites including wt = 344.0\n",
      "median reads at guide sites mutations only = 113.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 18 (45.0% of insertions)\n",
      "number of deletions overlapping TFBS = 16 (72.72727272727273% of insertions)\n",
      "number of substitutions overlapping TFBS = 2 (18.181818181818183% of insertions)\n"
     ]
    }
   ],
   "source": [
    "DREB26_genotyped_df, DREB26_mutated_df,DREB26_flattened_df = read_files(mutations_DREB26_genotyped,mutations_DREB26_genotyped_only_mutated,mutations_DREB26_genotyped_only_mutated_flattened)\n",
    "DREB26_genotyped_plant_lines, DREB26_mutated_plant_lines,DREB26_mutations = count_mutations(DREB26_genotyped_df,DREB26_mutated_df,DREB26_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total plant lines sequenced=478 \n",
      "total mutated plant lines=112\n",
      "\u001b[1mGenotypes\u001b[0m\n",
      "number of non-chimeric mutated plant lines = 107 (95.53571428571429% of non-chimeric lines)\n",
      "number of homozygous mutated plant lines = 53 (49.532710280373834% of non-chimeric lines)\n",
      "number of heterozygous mutated plant lines = 32 (29.906542056074763% of non-chimeric lines)\n",
      "number of biallelic mutated plant lines = 22 (20.5607476635514% of non-chimeric lines)\n",
      "\u001b[1mMutation types\u001b[0m\n",
      "total mutations = 151\n",
      "number of insertions = 127 (84.10596026490066% of total mutations)\n",
      "number of deletions = 39 (25.82781456953642% of total mutations)\n",
      "deletions over 10bp = 10 (25.64102564102564% of deletions)\n",
      "number of substitutions = 14 (9.271523178807946% of total mutations)\n",
      "median reads at guide sites including wt = 646.0\n",
      "median reads at guide sites mutations only = 349.0\n",
      "\u001b[1mOverlapping TFBSs\u001b[0m\n",
      "number of insertions overlapping TFBS = 41 (32.28346456692913% of insertions)\n",
      "number of deletions overlapping TFBS = 7 (17.94871794871795% of insertions)\n",
      "number of substitutions overlapping TFBS = 1 (7.142857142857142% of insertions)\n"
     ]
    }
   ],
   "source": [
    "NLP7_genotyped_df, NLP7_mutated_df,NLP7_flattened_df = read_files(mutations_NLP7_genotyped,mutations_NLP7_genotyped_only_mutated,mutations_NLP7_genotyped_only_mutated_flattened)\n",
    "NLP7_genotyped_plant_lines, NLP7_mutated_plant_lines,NLP7_mutations = count_mutations(NLP7_genotyped_df,NLP7_mutated_df,NLP7_flattened_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('pybedtools')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2879a2396fb26a2192d4c546fae66e67014459dff7902256699e5c7228156d44"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
