{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alleles_frequency_table.zip can be unzipped to a tab-separated text file that shows all reads and alignments to references. The first column shows the aligned sequence of the sequenced read. The second column shows the aligned sequence of the reference sequence. Gaps in each of these columns represent insertions and deletions. The next column 'Reference_Name' shows the name of the reference that the read aligned to. The fourth column, 'Read_Status' shows whether the read was modified or unmodified. The fifth through seventh columns ('n_deleted', 'n_inserted', 'n_substituted') show the number of bases deleted, inserted, and substituted as compared to the reference sequence. The eighth column shows the number of reads having that sequence, and the ninth column shows the percentage of all reads having that sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #features to add:\n",
    "# Distance from TSS - get relative position of mutation in the guide site - done. Add distance from cut site metric. - done Then calculate distance from Araport TSS - done\n",
    "# for this: first create a bed file for all of the mutations (relative to whole Arabidopsis genome). Then do bedtools merge or intersect (or bedtools coverage (../data_sorting/./TFBS_coverage.sh)) with the mapped motif bed file (all TFBSs for all genes). Record each TFBS that overlaps the mutation\n",
    "# Overlapping TFBSs - subnetwork and all TFs\n",
    "\n",
    "# Include secondary mutations in case both deletion and substitution for example - done\n",
    "# Plant ID\n",
    "# How many biallelic or homozygous? How many wildtype?\n",
    "# More than 2 alleles for a gene - record alleles until 80% of reads accounted for\n",
    "# Prioiritse homozygous or biallelic\n",
    "# How many plants had mutations? How many guides produced mutations in each gene?\n",
    "#check window around cut site - at the moment I am including mutations 20bp either side, maybe cut the alignments down to 7bp either side before comparing them with find_indels_substitutions()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use env pybedtools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import io\n",
    "from pybedtools import BedTool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_in_files(mutations_file,mapped_motifs):\n",
    "    \"\"\"read in the files\"\"\"\n",
    "    #read in mapped motifs bed file\n",
    "    mapped_motifs = pd.read_table(mapped_motifs_bed, sep=\"\\t\", header=None)\n",
    "    if len(mapped_motifs.columns) == 24:\n",
    "        cols = [\n",
    "            \"chr\",\n",
    "            \"start\",\n",
    "            \"stop\",\n",
    "            \"promoter_AGI\",\n",
    "            \"dot1\",\n",
    "            \"strand\",\n",
    "            \"source\",\n",
    "            \"type\",\n",
    "            \"dot2\",\n",
    "            \"attributes\",\n",
    "            \"motif_chr\",\n",
    "            \"motif_start\",\n",
    "            \"motif_stop\",\n",
    "            \"name_rep\",\n",
    "            \"score\",\n",
    "            \"motif_strand\",\n",
    "            \"promoter_AGI2\",\n",
    "            \"p-value\",\n",
    "            \"q-value\",\n",
    "            \"matched_sequence\",\n",
    "            \"TF_name\",\n",
    "            \"TF_family\",\n",
    "            \"TF_AGI\",\n",
    "            \"bp_overlap\",\n",
    "        ]\n",
    "        mapped_motifs.columns = cols\n",
    "        # filter columns\n",
    "        mapped_motifs = mapped_motifs[\n",
    "            [\n",
    "                \"motif_chr\",\n",
    "                \"motif_start\",\n",
    "                \"motif_stop\",\n",
    "                \"name_rep\",\n",
    "                \"score\",\n",
    "                \"motif_strand\",\n",
    "                \"promoter_AGI2\",\n",
    "                \"p-value\",\n",
    "                \"q-value\",\n",
    "                \"matched_sequence\",\n",
    "                \"TF_name\",\n",
    "                \"TF_family\",\n",
    "                \"TF_AGI\",\n",
    "            ]\n",
    "        ]\n",
    "        #rename columns\n",
    "        cols = [\n",
    "            \"motif_chr\",\n",
    "            \"motif_start\",\n",
    "            \"motif_stop\",\n",
    "            \"name_rep\",\n",
    "            \"score\",\n",
    "            \"motif_strand\",\n",
    "            \"promoter_AGI\",\n",
    "            \"p-value\",\n",
    "            \"q-value\",\n",
    "            \"matched_sequence\",\n",
    "            \"TF_name\",\n",
    "            \"TF_family\",\n",
    "            \"TF_AGI\",\n",
    "        ]\n",
    "        mapped_motifs.columns = cols\n",
    "\n",
    "    elif len(mapped_motifs.columns) == 13:\n",
    "        cols = [\n",
    "            \"motif_chr\",\n",
    "            \"motif_start\",\n",
    "            \"motif_stop\",\n",
    "            \"name_rep\",\n",
    "            \"score\",\n",
    "            \"motif_strand\",\n",
    "            \"promoter_AGI\",\n",
    "            \"p-value\",\n",
    "            \"q-value\",\n",
    "            \"matched_sequence\",\n",
    "            \"TF_name\",\n",
    "            \"TF_family\",\n",
    "            \"TF_AGI\",\n",
    "        ]\n",
    "        mapped_motifs.columns = cols\n",
    "\n",
    "    elif len(mapped_motifs.columns) == 17:\n",
    "        cols = [\n",
    "            \"motif_chr\",\n",
    "            \"motif_start\",\n",
    "            \"motif_stop\",\n",
    "            \"name_rep\",\n",
    "            \"score\",\n",
    "            \"motif_strand\",\n",
    "            \"promoter_AGI\",\n",
    "            \"p-value\",\n",
    "            \"q-value\",\n",
    "            \"matched_sequence\",\n",
    "            \"TF_name\",\n",
    "            \"TF_family\",\n",
    "            \"TF_AGI\",\n",
    "            \"chr_openchrom\",\n",
    "            \"start_openchrom\",\n",
    "            \"stop_openchrom\",\n",
    "            \"bp_overlap\",\n",
    "        ]\n",
    "        mapped_motifs.columns = cols\n",
    "\n",
    "    mutations_df = pd.read_table(mutations_file,sep='\\t',header=0)\n",
    "    return mutations_df,mapped_motifs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_bedfiles(bedfile, mapped_motifs_bed, output_buffer):\n",
    "    \"\"\"perform bedtools intersect on the two dfs\"\"\"\n",
    "    df = BedTool(bedfile)\n",
    "    motifs = BedTool(mapped_motifs_bed)\n",
    "    # -wao =Write the original A and B entries plus the number of base pairs of overlap between the two features.\n",
    "    # However, A features w/o overlap are also reported with a NULL B feature and overlap = 0\n",
    "    intersect = df.intersect(motifs, wao=True)\n",
    "    # Write to output_file\n",
    "    # Each line in the file contains bed entry a and bed entry b that it overlaps plus the number of bp in the overlap so 19 columns\n",
    "    output_buffer.write(str(intersect))\n",
    "    #go back to beginning of buffer\n",
    "    output_buffer.seek(0)\n",
    "    mapped_motifs_bed.seek(0)\n",
    "    return output_buffer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlapping_TFBSs(mutations_df,mapped_motifs_df):\n",
    "    \"\"\"function to find any overlapping TFBSs from FIMO mapped motif file\"\"\"\n",
    "   \n",
    "\n",
    "    #for each guide containing mutations, create a temporary bed file containing each mutation and then do bedtools intersect to find which overlap TFBs\n",
    "    #then add the TFBS names into a new column for that row\n",
    "\n",
    "    #get column names from mapped_motifs_df\n",
    "    mapped_motifs_bed_columns = list(mapped_motifs_df.columns)\n",
    "\n",
    "\n",
    "    #turn mapped_motifs_df into a buffer\n",
    "    mapped_motifs_bed = io.StringIO()\n",
    "    mapped_motifs_df.to_csv(mapped_motifs_bed, sep=\"\\t\", index=False, header=None)\n",
    "    #go back to start of buffer\n",
    "    mapped_motifs_bed.seek(0)\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "    #first make certain columns string\n",
    "    #make columns containing lists string for now so can use groupby\n",
    "    to_string = ['insertion_positions','deletion_positions','substitution_positions','insertion_cut_site_distance','deletion_cut_site_distance','substitution_cut_site_distance','insertion_positions_relative_to_TSS','insertion_genomic_positions','deletion_positions_relative_to_TSS','deletion_genomic_positions','substitution_positions_relative_to_TSS','substitution_genomic_positions']\n",
    "    mutations_df[to_string] = mutations_df[to_string].astype(str)\n",
    "    for index,row in mutations_df.iterrows():\n",
    "            \n",
    "        if row.mutation_type == 'None':\n",
    "            pass\n",
    "        else:\n",
    "            #create temporary df in bed format\n",
    "            \n",
    "            cols = [\"chr\",\n",
    "            \"start\",\n",
    "            \"stop\",\n",
    "            \"mutation_type\",'mutation_count']\n",
    "            temp_df = pd.DataFrame(columns=cols)\n",
    "            chr = row.chr\n",
    "            #if not NaN\n",
    "            if row.insertion_genomic_positions != 'nan':\n",
    "                #print(\"Index:\", index)\n",
    "                #print(row.insertion_genomic_positions)\n",
    "                #convert genomic positions from string to list\n",
    "                # Convert string to list if more than one\n",
    "                insertion_genomic_positions = row.insertion_genomic_positions.strip('][').split(', ')\n",
    "                #count which mutation number currently on to be added to the temporary bed file\n",
    "                count = 0\n",
    "                for gen_pos in insertion_genomic_positions:\n",
    "                    count += 1\n",
    "                    #get index\n",
    "                    # if len(insertion_genomic_positions) > 1:\n",
    "                    #     index = insertion_genomic_positions.index(gen_pos)\n",
    "                    # else:\n",
    "                    #     index = 'nan'\n",
    "                    start = int(gen_pos)\n",
    "                    stop = start + 1\n",
    "                    mutation_type = \"insertion\"\n",
    "                    #add to temp_df\n",
    "                    temp_list = [chr,start,stop,mutation_type,count]\n",
    "                    temp_df.loc[len(temp_df)] = temp_list\n",
    "            if row.deletion_genomic_positions != \"nan\":\n",
    "                # Convert string to list\n",
    "                deletion_genomic_positions = row.deletion_genomic_positions.strip('][').split(', ')\n",
    "                #count which mutation number currently on to be added to the temporary bed file\n",
    "                count = 0\n",
    "                for gen_pos in deletion_genomic_positions:\n",
    "                    count += 1\n",
    "                    #get index\n",
    "                    # if len(deletion_genomic_positions) > 1:\n",
    "                    #     index = deletion_genomic_positions.index(gen_pos)\n",
    "                    # else:\n",
    "                    #     index = 'nan'\n",
    "                    start = int(gen_pos)\n",
    "                    stop = start + 1\n",
    "                    mutation_type = \"deletion\"\n",
    "                    #add to temp_df\n",
    "                    temp_list = [chr,start,stop,mutation_type,count]\n",
    "                    temp_df.loc[len(temp_df)] = temp_list\n",
    "            if row.substitution_genomic_positions != \"nan\":\n",
    "                # Convert string to list\n",
    "                substitution_genomic_positions = row.substitution_genomic_positions.strip('][').split(', ')\n",
    "                #count which mutation number currently on to be added to the temporary bed file\n",
    "                count = 0\n",
    "                for gen_pos in substitution_genomic_positions:\n",
    "                    count += 1\n",
    "                    #get index\n",
    "                    # if len(substitution_genomic_positions) > 1:\n",
    "                    #     index = substitution_genomic_positions.index(gen_pos)\n",
    "                    # else:\n",
    "                    #     index = 'nan'\n",
    "                    start = int(gen_pos)\n",
    "                    stop = start + 1\n",
    "                    mutation_type = \"substitution\"\n",
    "                    #add to temp_df\n",
    "                    temp_list = [chr,start,stop,mutation_type,count]\n",
    "                    temp_df.loc[len(temp_df)] = temp_list\n",
    "            #now do bedtools intersect to find which TFBSs overlap with which mutations\n",
    "            #sort by chr then start\n",
    "            temp_df = temp_df.sort_values([\"chr\", \"start\"]).reset_index(drop=True)\n",
    "            # write to buffer\n",
    "            temp_df_buffer = io.StringIO()\n",
    "            temp_df.to_csv(temp_df_buffer, sep=\"\\t\", index=False, header=None)\n",
    "            temp_df_buffer.seek(0)\n",
    "            \n",
    "            output_buffer = io.StringIO()\n",
    "        \n",
    "            output_buffer = merge_bedfiles(temp_df_buffer, mapped_motifs_bed, output_buffer)\n",
    "            #read in output buffer as df\n",
    "            output_df = pd.read_table(output_buffer, sep='\\t')\n",
    "            #get column names and rename columns\n",
    "            output_df_cols = cols+mapped_motifs_bed_columns+['bp_overlap']\n",
    "            output_df.columns = output_df_cols\n",
    "            #for each mutation type get list of overlapping TFBSs. Add theses to a dictionary column in the mutations_df            \n",
    "            print(output_df)      \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# folder = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call'\n",
    "# ARF9_root = f'{folder}/ARF9_sgRNAs/test'\n",
    "# ARF18_root = f'{folder}/ARF18_sgRNAs/7bp_window_noplots'\n",
    "# DREB26_root = f'{folder}/DREB26_sgRNAs/7bp_window_noplots'\n",
    "# NLP7_root = f'{folder}/NLP7_sgRNAs/7bp_window_noplots'\n",
    "# output = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/'\n",
    "# #dictionary of reference fasta file locations\n",
    "# reference_folder = '../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/References'\n",
    "# reference_gene_dict = {'ARF9':f'{reference_folder}/ARF9_promoter.fa','ARF18':f'{reference_folder}/ARF18_promoter.fa','DREB26':f'{reference_folder}/DREB26_promoter.fa','NLP7':f'{reference_folder}/NLP7_promoter.fa'}\n",
    "# reference_fasta = f'{reference_folder}/genes_longest_region.fa'\n",
    "# reference_promoter_bed = f'{reference_folder}/genes_longest_region.bed'\n",
    "# #promoters bed file when 3' end is the TSS (used the bed file from promoter architecture non-overlapping_includingbidirectional_all_genes_newannotation project FIMO folder)\n",
    "# all_promoters_bed = f'{reference_folder}/promoters.bed'\n",
    "# #mapped motif bed file of TFBSs scanned with FIMO\n",
    "# mapped_motifs_bed = '../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation_3KB/FIMO/promoters_5UTR_motifs_mapped_q0_05.bed'\n",
    "# plant_IDs = f'{reference_folder}/plant_IDs.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_motifs_bed = '../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation_3KB/FIMO/promoters_5UTR_motifs_mapped_q0_05.bed'\n",
    "mutations_ARF9 = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9_merged.tsv'\n",
    "mutations_ARF18 = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_merged.tsv'\n",
    "mutations_DREB26 = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26_merged.tsv'\n",
    "mutations_NLP7 = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7_merged.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in files\n",
    "mutations_ARF9_df,mapped_motifs_df =  read_in_files(mutations_ARF9, mapped_motifs_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0     4  12451173  12451174     insertion               1          4   \n",
      "1     4  12451173  12451174     insertion               1          4   \n",
      "2     4  12451173  12451174     insertion               1          4   \n",
      "3     4  12451174  12451175     insertion               2          4   \n",
      "4     4  12451174  12451175     insertion               2          4   \n",
      "5     4  12451174  12451175     insertion               2          4   \n",
      "6     4  12451174  12451175     insertion               2          4   \n",
      "7     4  12451187  12451188      deletion               1          4   \n",
      "8     4  12451188  12451189      deletion               2          4   \n",
      "9     4  12451189  12451190      deletion               3          4   \n",
      "10    4  12451196  12451197  substitution               1          4   \n",
      "\n",
      "    motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0      12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1      12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2      12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3      12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4      12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5      12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6      12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "7      12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "8      12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "9      12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "10     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "   promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0     AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1     AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2     AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3     AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4     AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5     AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6     AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "7     AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "8     AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "9     AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "10    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "   TF_family     TF_AGI  bp_overlap  \n",
      "0     BBRBPC  AT4G38910           1  \n",
      "1     BBRBPC  AT4G38910           1  \n",
      "2     BBRBPC  AT4G38910           1  \n",
      "3     BBRBPC  AT4G38910           1  \n",
      "4     BBRBPC  AT4G38910           1  \n",
      "5     BBRBPC  AT4G38910           1  \n",
      "6     BBRBPC  AT4G38910           1  \n",
      "7     BBRBPC  AT4G38910           1  \n",
      "8     BBRBPC  AT4G38910           1  \n",
      "9     BBRBPC  AT4G38910           1  \n",
      "10    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count motif_chr  \\\n",
      "0    4  12451081  12451082     insertion               2         .   \n",
      "\n",
      "   motif_start  motif_stop name_rep score motif_strand promoter_AGI p-value  \\\n",
      "0           -1          -1        .     .            .            .       .   \n",
      "\n",
      "  q-value matched_sequence TF_name TF_family TF_AGI  bp_overlap  \n",
      "0       .                .       .         .      .           0  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12451173  12451174     insertion               1          4   \n",
      "1    4  12451173  12451174     insertion               1          4   \n",
      "2    4  12451173  12451174     insertion               1          4   \n",
      "3    4  12451174  12451175     insertion               2          4   \n",
      "4    4  12451174  12451175     insertion               2          4   \n",
      "5    4  12451174  12451175     insertion               2          4   \n",
      "6    4  12451174  12451175     insertion               2          4   \n",
      "\n",
      "   motif_start  motif_stop                  name_rep     score motif_strand  \\\n",
      "0     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "1     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "2     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "3     12451145    12451175     BBRBPC_tnt.BPC5_col_a -31.39740            +   \n",
      "4     12451149    12451179     BBRBPC_tnt.BPC5_col_a -40.41030            +   \n",
      "5     12451151    12451181     BBRBPC_tnt.BPC5_col_a -43.23080            +   \n",
      "6     12451173    12451202  BBRBPC_tnt.BPC5_colamp_a  -5.23256            +   \n",
      "\n",
      "  promoter_AGI   p-value  q-value                matched_sequence TF_name  \\\n",
      "0    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "1    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "2    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "3    AT4G23980  0.000003  0.00705  AGAGAGAGTGAAACGGAGATAAAAGCAAAC    BPC5   \n",
      "4    AT4G23980  0.000018  0.02230  AGAGTGAAACGGAGATAAAAGCAAACGGAG    BPC5   \n",
      "5    AT4G23980  0.000030  0.03040  AGTGAAACGGAGATAAAAGCAAACGGAGGT    BPC5   \n",
      "6    AT4G23980  0.000087  0.04210   ACGGAGGTGTTAACTTAAAGACGGGAGAG    BPC5   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0    BBRBPC  AT4G38910           1  \n",
      "1    BBRBPC  AT4G38910           1  \n",
      "2    BBRBPC  AT4G38910           1  \n",
      "3    BBRBPC  AT4G38910           1  \n",
      "4    BBRBPC  AT4G38910           1  \n",
      "5    BBRBPC  AT4G38910           1  \n",
      "6    BBRBPC  AT4G38910           1  \n",
      "   chr     start      stop mutation_type  mutation_count motif_chr  \\\n",
      "0    4  12451081  12451082     insertion               2         .   \n",
      "\n",
      "   motif_start  motif_stop name_rep score motif_strand promoter_AGI p-value  \\\n",
      "0           -1          -1        .     .            .            .       .   \n",
      "\n",
      "  q-value matched_sequence TF_name TF_family TF_AGI  bp_overlap  \n",
      "0       .                .       .         .      .           0  \n",
      "   chr     start      stop mutation_type  mutation_count  motif_chr  \\\n",
      "0    4  12450378  12450379  substitution               2          4   \n",
      "1    4  12450378  12450379  substitution               2          4   \n",
      "\n",
      "   motif_start  motif_stop                     name_rep    score motif_strand  \\\n",
      "0     12450364    12450391  C2C2dof_tnt.AT1G69570_col_a  13.5873            -   \n",
      "1     12450368    12450389       C2C2dof_tnt.OBP3_col_a  13.2500            -   \n",
      "\n",
      "  promoter_AGI   p-value  q-value             matched_sequence    TF_name  \\\n",
      "0    AT4G23980  0.000011   0.0463  CCGTCACTTTTACCGTTTGCTCTTATG  AT1G69570   \n",
      "1    AT4G23980  0.000014   0.0407        GTCACTTTTACCGTTTGCTCT       OBP3   \n",
      "\n",
      "  TF_family     TF_AGI  bp_overlap  \n",
      "0   C2C2DOF  AT1G69570           1  \n",
      "1   C2C2DOF  AT3G55370           1  \n"
     ]
    }
   ],
   "source": [
    "find_overlapping_TFBSs(mutations_ARF9_df,mapped_motifs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('pybedtools')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2879a2396fb26a2192d4c546fae66e67014459dff7902256699e5c7228156d44"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
