{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\"\"\"script to generate subplots showing read depth across each of the four amplicons\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib import rcParams\n",
    "# use pacbio conda environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set matplotlib rc parameters\n",
    "def set_rc_params():\n",
    "    #set matplotlib default parameters\n",
    "    rcParams['xtick.major.width'] = 2\n",
    "    rcParams['ytick.major.width'] = 2\n",
    "    rcParams['axes.linewidth'] = 2\n",
    "    #rcParams['lines.linewidth'] = 2\n",
    "    #remove top and right lines\n",
    "    rcParams['axes.spines.top'] = False\n",
    "    rcParams['axes.spines.right'] = False\n",
    "    #font size\n",
    "    fontsize = 14\n",
    "    rcParams['font.size'] = fontsize\n",
    "    #for getting the microsoft font Arial working, please follow this guide: https://alexanderlabwhoi.github.io/post/2021-03-missingfont/\n",
    "    rcParams['font.family'] = 'sans-serif'\n",
    "    rcParams['font.sans-serif'] = ['Arial']\n",
    "    #allow font to be edited later in pdf editor\n",
    "    #make svg text editable\n",
    "    rcParams['svg.fonttype'] = 'none'\n",
    "    rcParams ['pdf.fonttype'] = 42 \n",
    "    #align y-axis top most tick with end of axis\n",
    "    rcParams['axes.autolimit_mode'] = 'round_numbers'\n",
    "    #set margins to ensure any error bars fit\n",
    "    rcParams['axes.xmargin'] = 0.2\n",
    "    rcParams['axes.ymargin'] = 0.2\n",
    "    #define bar width\n",
    "    #bar_width = 0.65\n",
    "    return fontsize\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process data function\n",
    "def process_data(file):\n",
    "    df = pd.read_csv(file, header=0)\n",
    "    # create gene column\n",
    "    df['gene'] = df['barcode_pair_gene'].str.split('_').str[1]\n",
    "    # remove nan values\n",
    "    df = df.dropna()    \n",
    "    # \n",
    "    # bc1017_ARF18_SW346_SW442.trim.fasta\n",
    "   # print(df)\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make plot showing read depth for each gene\n",
    "def plot_read_depth(df, output_location):\n",
    "    \"\"\"function to plot read counts (count column) for each gene\"\"\"\n",
    "\n",
    "\n",
    "    #df_grouped = df.groupby(['gene'])\n",
    "    # sort by read depth\n",
    "    #df_grouped = df_grouped.sort_values(by=['count'], ascending=False)\n",
    "    # plot\n",
    "\n",
    "    order=['ARF9','ARF18', 'DREB26','NLP7']\n",
    "    # sort df by order of list\n",
    "    df['gene'] = pd.Categorical(df['gene'], order)\n",
    "\n",
    "    # df.sort_values(by=['gene'], key=lambda x: x.map(order), inplace=True)\n",
    "    # print(df)\n",
    "\n",
    "    plt.figure(figsize=(10,10))\n",
    "    fig = sns.histplot(x='gene',y='count', data=df, )\n",
    "    #fig = sns.stripplot(x='gene',y='count', data=df)\n",
    "    #plt.xticks(rotation=90)\n",
    "    plt.xlabel('Amplicon')\n",
    "    plt.ylabel('Number of mapped reads')\n",
    "    # set y lim to start at 0\n",
    "    plt.ylim(0)\n",
    "    #plt.title('Read Depth for Each Gene')\n",
    "    # change x axis order to\n",
    "    #fig.set_xticklabels(fig.get_xticklabels(), rotation=90)\n",
    "    #save svg and pdf\n",
    "    plt.savefig(f'{output_location}/reads_mapped.svg', format='svg', bbox_inches='tight')\n",
    "    plt.savefig(f'{output_location}/reads_mapped.pdf', format='pdf', bbox_inches='tight')\n",
    "    #remove plot\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "    \n",
    "    #return df_grouped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make plot showing reads at each mutated site\n",
    "def plot_mutated_sites(files, output_location):\n",
    "    \"\"\"function to plot reads containing each mutation\"\"\"\n",
    "    #read in files\n",
    "    arf9 = pd.read_csv(files[0], header=0, sep='\\t')\n",
    "    #make column for gene\n",
    "    arf9['gene'] = 'ARF9'\n",
    "    arf18 = pd.read_csv(files[1], header=0, sep='\\t')\n",
    "    #make column for gene\n",
    "    arf18['gene'] = 'ARF18'\n",
    "    dreb26 = pd.read_csv(files[2], header=0, sep='\\t')\n",
    "    #make column for gene\n",
    "    dreb26['gene'] = 'DREB26'\n",
    "\n",
    "    nlp7 = pd.read_csv(files[3], header=0, sep='\\t')\n",
    "    #make column for gene\n",
    "    nlp7['gene'] = 'NLP7'\n",
    "    \n",
    "    #concatenate dfs\n",
    "    df = pd.concat([arf9, arf18, dreb26, nlp7], axis=0)\n",
    "    print(df.columns)\n",
    "    #reset index\n",
    "    df = df.reset_index(drop=True)\n",
    "    \n",
    "    # #sort df by gene\n",
    "\n",
    "    # order=['ARF9','ARF18', 'DREB26','NLP7']\n",
    "    # # sort df by order of list\n",
    "    # df['gene'] = pd.Categorical(df['gene'], order)\n",
    "\n",
    "    # df.sort_values(by=['gene'], key=lambda x: x.map(order), inplace=True)\n",
    "    # print(df)\n",
    "\n",
    "    plt.figure(figsize=(10,10))\n",
    "    fig = sns.histplot(x='gene',y='read_number', data=df )\n",
    "    #fig = sns.stripplot(x='gene',y='count', data=df)\n",
    "    #plt.xticks(rotation=90)\n",
    "    plt.xlabel('Gene')\n",
    "    plt.ylabel('Mutation read count')\n",
    "    # set y lim to start at 0\n",
    "    plt.ylim(0)\n",
    "    #plt.title('Read Depth for Each Gene')\n",
    "    # change x axis order to\n",
    "    #fig.set_xticklabels(fig.get_xticklabels(), rotation=90)\n",
    "    #save svg and pdf\n",
    "    plt.savefig(f'{output_location}/mutation_read_counts.svg', format='svg', bbox_inches='tight')\n",
    "    plt.savefig(f'{output_location}/mutation_read_counts.pdf', format='pdf', bbox_inches='tight')\n",
    "    #remove plot\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "#main function\n",
    "def main():\n",
    "    # file location\n",
    "    file_loc = '../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/CB-PPBFX-994_Sam_Witham_EI_SW_ENQ-5142_A_01_demux_by_gene.csv'\n",
    "    #genotyped_only_mutated_df_file = \n",
    "   # mutations_ARF9_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped.tsv'\n",
    "    mutations_ARF9_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "    #mutations_ARF9_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF9/ARF9_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "    #mutations_ARF18_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped.tsv'\n",
    "    #mutations_ARF18_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "    mutations_ARF18_genotyped_only_mutated_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_filtered.tsv'\n",
    "    #mutations_ARF18_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "    #mutations_ARF18_genotyped_only_mutated_flattened_filtered = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/ARF18_filtered/ARF18_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "    #mutations_DREB26_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped.tsv'\n",
    "    mutations_DREB26_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "    #mutations_DREB26_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/DREB26/DREB26_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "    #mutations_NLP7_genotyped = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped.tsv'\n",
    "    mutations_NLP7_genotyped_only_mutated = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated.tsv'\n",
    "    #mutations_NLP7_genotyped_only_mutated_flattened = f'../../data/CRISPR_library/pacbio/demultiplexed/Data_Package_Batch_04_04_2022/Sam_Witham_EI_SW_ENQ-5142_A_01_Additional_Barcode_Analysis/Variant_call/NLP7/NLP7_TFBSoverlapping_genotyped_only_mutated_flattened.tsv'\n",
    "\n",
    "    output_location = '../../data/CRISPR_library'\n",
    "    # read in and process data \n",
    "    df = process_data(file_loc)\n",
    "    # set matplotlib rc parameters\n",
    "    fontsize = set_rc_params()\n",
    "\n",
    "    #m make new folder\n",
    "    dirName = f'{output_location}/sequencing_plots'\n",
    "    try:\n",
    "        # Create target Directory\n",
    "        os.mkdir(dirName)\n",
    "        print(\"Directory \" , dirName ,  \" created\") \n",
    "    except FileExistsError:\n",
    "        print(\"Directory \" , dirName , \" already exists\")\n",
    "    # plot data\n",
    "    plot_read_depth(df,dirName)\n",
    "    #median read depth for arf9\n",
    "    arf9_median = df.loc[df['gene'] == 'ARF9']['count'].median()\n",
    "    print(f'median reads mapped for ARF9 is {arf9_median}')\n",
    "    #median read depth for arf18\n",
    "    arf18_median = df.loc[df['gene'] == 'ARF18']['count'].median()\n",
    "    print(f'median reads mapped for ARF18 is {arf18_median}')\n",
    "    #median read depth for dreb26\n",
    "    dreb26_median = df.loc[df['gene'] == 'DREB26']['count'].median()\n",
    "    print(f'median reads mapped for DREB26 is {dreb26_median}')\n",
    "    #median read depth for nlp7\n",
    "    nlp7_median = df.loc[df['gene'] == 'NLP7']['count'].median()\n",
    "    print(f'median reads mapped for NLP7 is {nlp7_median}')\n",
    "\n",
    "    # plot reads at mutation sites\n",
    "    #make four more subplots\n",
    "    # fig2, axes2 = plt.subplots(2, 2, figsize=(10, 8), sharex=False)\n",
    "    # #flatten axis array\n",
    "    # axes2 = axes2.flatten()\n",
    "    # make deletion size plots\n",
    "    plot_mutated_sites([mutations_ARF9_genotyped_only_mutated,mutations_ARF18_genotyped_only_mutated_filtered,mutations_DREB26_genotyped_only_mutated,mutations_NLP7_genotyped_only_mutated],dirName)\n",
    "    # plot_mutated_sites(mutations_ARF18_genotyped_only_mutated_filtered, axes2[1], 'ARF18')\n",
    "    # plot_mutated_sites(mutations_DREB26_genotyped_only_mutated, axes2[2], 'DREB26')\n",
    "    # plot_mutated_sites(mutations_NLP7_genotyped_only_mutated, axes2[3], 'NLP7')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # mean read depth for arf9\n",
    "    # arf9_mean = df.loc[df['gene'] == 'ARF9']['count'].mean()\n",
    "    # print(f'ARF9 mean read depth: {arf9_mean}')\n",
    "    # # mean read depth for arf18\n",
    "    # arf18_mean = df.loc[df['gene'] == 'ARF18']['count'].mean()\n",
    "    # print(f'ARF18 mean read depth: {arf18_mean}')\n",
    "    # # mean read depth for dreb26\n",
    "    # dreb26_mean = df.loc[df['gene'] == 'DREB26']['count'].mean()\n",
    "    # print(f'DREB26 mean read depth: {dreb26_mean}')\n",
    "    # # mean read depth for nlp7\n",
    "    # nlp7_mean = df.loc[df['gene'] == 'NLP7']['count'].mean()\n",
    "    # print(f'NLP7 mean read depth: {nlp7_mean}')\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/CRISPR_library/sequencing_plots  already exists\n",
      "median reads mapped for ARF9 is 314.5\n",
      "median reads mapped for ARF18 is 680.0\n",
      "median reads mapped for DREB26 is 342.5\n",
      "median reads mapped for NLP7 is 660.0\n",
      "Index(['chr', 'plant_ID', 'platename', 'library', 'first_reaction_primers',\n",
      "       'second_reaction_primers', 'guide', 'guide_number', 'aligned_sequence',\n",
      "       'reference_sequence', 'mutation_type', 'genotype', 'read_number',\n",
      "       'read_percentage', 'insertion_positions', 'deletion_positions',\n",
      "       'substitution_positions', 'insertion_cut_site_distance',\n",
      "       'deletion_cut_site_distance', 'substitution_cut_site_distance',\n",
      "       'cut_site_promoter_position', 'insertion_positions_relative_to_TSS',\n",
      "       'insertion_genomic_positions', 'deletion_positions_relative_to_TSS',\n",
      "       'deletion_genomic_positions', 'substitution_positions_relative_to_TSS',\n",
      "       'substitution_genomic_positions', 'insertion_overlapping_TFBS_family',\n",
      "       'insertion_overlapping_TFBS_AGI', 'deletion_overlapping_TFBS_family',\n",
      "       'deletion_overlapping_TFBS_AGI', 'substitution_overlapping_TFBS_family',\n",
      "       'substitution_overlapping_TFBS_AGI', 'number_of_different_alleles',\n",
      "       'number_of_different_non_wt_alleles', 'mutated_guide_sites',\n",
      "       'approved_guide_pairs', 'guides_not_in_pairs',\n",
      "       'estimated_agro_strain_number', 'gene', 'Unnamed: 0'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "#run main\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('pacbio')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "6d7f1c559912387f2a5cf637f9ec9c1bbd03351b317a3bab0d49d1cb8d0c8d65"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
