{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 521,
   "metadata": {},
   "outputs": [],
   "source": [
    "#consider using svist4get api for adding open chromatin peak tracks: https://bitbucket.org/artegorov/svist4get/src/eeb5151f49c31fa887dbfc168320c63d66a17334/docs/API.md\n",
    "#actually use https://github.com/deeptools/pyGenomeTracks as can install with conda\n",
    "#then import each track as vector image like here: https://stackoverflow.com/questions/31452451/importing-an-svg-file-into-a-matplotlib-figure using svgutils\n",
    "#integrate it with matplotlib like mentioned here: https://github.com/deeptools/pyGenomeTracks/issues/20\n",
    "#the bigwig files I used were from potter et al 2018 NaOH treatment https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116287 called GSE116287_Roots_NaOH_Merged.bw and GSE116287_Shoots_NaOH_Merged.bw\n",
    "#had to install \"conda install -c conda-forge tqdm\" too\n",
    "\n",
    "##NOTE - if using genbank files from Benchling, first add custom fields on the metadata page with the start and end index (eg. start_index 6015558)\n",
    "##also ensure the name of the genebank file (and sequence id) include the AGI locus name enclosed in parentheses(eg. \"(AT4G24020)\")\n",
    "##This is so the chromosome number can be extracted\n",
    "##This is to ensure the start and end chromosome positions open chromatin data\n",
    "#make sure when exporting the .gb file from Benchling that you do not tick \"Convert Non-Standard Annotation Types\"\n",
    "#I merged bigwig files with DNaseseq data from Sullivan et al 2014 using WiggleTools (not shown in this notebook)\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import matplotlib.text as mtext\n",
    "\n",
    "#latex style rendering so can make parts of text bold\n",
    "from matplotlib import rc\n",
    "#import matplotlib as mpl\n",
    "from Bio import SeqIO\n",
    "\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, ExactPosition, BeforePosition, AfterPosition, BetweenPosition, WithinPosition, CompoundLocation\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "#allows custom colours\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "import pandas as pd\n",
    "from itertools import cycle\n",
    "import re\n",
    "import pandas as pd\n",
    "#allow converting from RGB to CMYK\n",
    "#import cv2\n",
    "from PIL import Image\n",
    "# from PIL import Image\n",
    "# from io import BytesIO\n",
    "import pygenometracks.tracks as pygtk\n",
    "\n",
    "###note  - use conda env gene_feature_plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a class specifying feature colours\n",
    "#make feature_list and colour_dict so that each feature name and colour is only added once to legend if more than one share the same name\n",
    "feature_list = []\n",
    "colour_dict = {}\n",
    "#set probel_labels on or off here\n",
    "probe_labels = False\n",
    "class MyCustomTranslator(BiopythonTranslator):\n",
    "    \"\"\"Custom translator iplementing the following theme:\n",
    "    -Colour promoter in pale green\n",
    "    -colour exons in dark grey\n",
    "    -colour introns in light grey\n",
    "    -colour TFs from colour palette\"\"\"\n",
    "\n",
    "    #import colour blind palette\n",
    "    #colour palette from mkweb.bcgsc.ca/colorblind/palettes.mhtml\n",
    "    # df = pd.read_csv(\"colourblind_palette.csv\", header=0)\n",
    "    # #convert to floats\n",
    "    # floats=df.divide(255)\n",
    "    # #make sets of each row to get the red, green blue colours\n",
    "    # CB_colour_palette = floats.apply(tuple, axis = 1)\n",
    "    # #make into df\n",
    "    # df = CB_colour_palette.to_frame()\n",
    "    # #save file\n",
    "    # df.to_csv('../../data/TRAMP/colour_list')\n",
    "    \n",
    "    # #turn into a list of colour sets\n",
    "    # list_colours = list(CB_colour_palette)\n",
    "    # colour_list = list_colours\n",
    "    #colour_list = ['#88CCEE', '#44AA99', '#117733', '#332288', '#DDCC77', '#999933','#CC6677', '#882255', '#AA4499', '#DDDDDD']\n",
    "    colour_list = ['#2271B2',\n",
    "    '#3DB7E9',\n",
    "    '#F748A5',    \n",
    "    '#d55e00',\n",
    "    '#e69f00',\n",
    "    #'#f0e442',\n",
    "    '#228833',\n",
    "    '#000000',]\n",
    "    #make colour iterator\n",
    "    colour_iterator=cycle(colour_list)\n",
    "    #change colour cycle\n",
    "    #mpl.rcParams['axes.prop_cycle'] = mpl.cycler(color=list_colours)\n",
    "    #count = -1\n",
    "    #set colour count\n",
    "\n",
    "\n",
    "    def compute_feature_color(self, feature):\n",
    "        \"\"\"set colour of each feature\"\"\"      \n",
    "\n",
    "        if feature.type == \"promoter\":\n",
    "            return \"#F5F5F5\"\n",
    "        elif feature.type == \"gene_upstream\":\n",
    "            return \"#DDCC77\"#2f4f4f\"#dark slate grey\n",
    "        elif feature.type == \"mRNA_upstream\":\n",
    "            return \"#DDCC77\"#dark slate grey\n",
    "        elif feature.type == \"exon_upstream\":\n",
    "            return \"#DDCC77\"#dark slate grey\n",
    "        elif feature.type == \"exon\":\n",
    "            return \"#635147\"#umber\n",
    "        elif feature.type == \"gene\":\n",
    "            return \"#F5F5F5\"\n",
    "            #return (169,169,169)\n",
    "        elif feature.type == \"intron\":\n",
    "            return \"lightgrey\"  \n",
    "            #return (211,211,211)\n",
    "        elif feature.type == \"5'UTR\":\n",
    "            return \"c4aead\"#silver pink\n",
    "        elif feature.type == \"start_codon\":\n",
    "            return \"black\"\n",
    "        # elif feature.type == \"TRAMP_probe_tested\":\n",
    "        #     if feature.qualifiers.get(\"label\")[0] == \"NLP7#7\":\n",
    "        #         col = \"#e69f00\"\n",
    "        #     elif feature.qualifiers.get(\"label\")[0] == \"NLP7#9\":\n",
    "        #         col = \"#d55e00\"\n",
    "        #     elif feature.qualifiers.get(\"label\")[0] == \"NLP7#10\":\n",
    "        #         col = \"#2271B2\"\n",
    "        #     elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#4\":\n",
    "        #         col = \"#2271B2\"\n",
    "        #     elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#8\":\n",
    "        #         col = \"#F748A5\"\n",
    "        #     elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#3\":\n",
    "        #         col = \"#000000\"\n",
    "\n",
    "        #     else:        \n",
    "        #         pass\n",
    "                        \n",
    "        #     return col\n",
    "            \n",
    "        elif feature.type == \"TFBS\":\n",
    "            if feature.qualifiers.get(\"label\")[0] in colour_dict.keys():\n",
    "                col = colour_dict[feature.qualifiers.get(\"label\")[0]]\n",
    "            else:\n",
    "                col = next(self.colour_iterator)\n",
    "                colour_dict[feature.qualifiers.get(\"label\")[0]] = col\n",
    "                        \n",
    "            return col\n",
    "        else:\n",
    "            return \"white\"\n",
    "\n",
    "    def compute_feature_box_linewidth(self, feature):\n",
    "        \"\"\"change shape of features\"\"\"\n",
    "        if feature.type ==\"TRAMP_probe_tested\":\n",
    "            if probe_labels is True:\n",
    "                print(f'probe labels = True')\n",
    "                #if prob labels = True, add probe labels\n",
    "                return 1       \n",
    "        else:\n",
    "            return 0\n",
    "\n",
    "    def compute_feature_linewidth(self, feature):\n",
    "        \"\"\"change linewidth of feature's arrow/rectangle\"\"\"\n",
    "        #remove border from certain features\n",
    "        if feature.type == \"gene_upstream\":\n",
    "            return 0\n",
    "        elif feature.type == \"mRNA_upstream\":\n",
    "            return 0\n",
    "        elif feature.type == \"exon_upstream\":\n",
    "            return 0\n",
    "        elif feature.type == \"misc_RNA_upstream\":\n",
    "            return 0  \n",
    "        elif feature.type == \"exon\":\n",
    "            return 0\n",
    "        elif feature.type == \"intron\":\n",
    "            return 0\n",
    "        elif feature.type == \"5'UTR\":\n",
    "            return 0\n",
    "        elif feature.type == \"TFBS\":\n",
    "            return 0\n",
    "        elif feature.type ==\"TRAMP_probe_tested\":\n",
    "            if probe_labels is True:\n",
    "                #if prob labels = True, add probe labels\n",
    "                return 2\n",
    "            elif probe_labels is False:\n",
    "                return 0 \n",
    "        return 0\n",
    "    \n",
    "    # def compute_feature_linecolor(self, feature):\n",
    "    #     if feature.type == \"TRAMP_probe_tested\":\n",
    "    #         if feature.qualifiers.get(\"label\")[0] == \"NLP7#7\":\n",
    "    #             col = \"#e69f00\"\n",
    "    #         elif feature.qualifiers.get(\"label\")[0] == \"NLP7#9\":\n",
    "    #             col = \"#d55e00\"\n",
    "    #         elif feature.qualifiers.get(\"label\")[0] == \"NLP7#10\":\n",
    "    #             col = \"#2271B2\"\n",
    "    #         elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#4\":\n",
    "    #             col = \"#2271B2\"\n",
    "    #         elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#8\":\n",
    "    #             col = \"#F748A5\"\n",
    "    #         elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#3\":\n",
    "    #             col = \"#000000\"\n",
    "\n",
    "    #         else:        \n",
    "    #             pass\n",
    "                        \n",
    "    #         return col\n",
    "    \n",
    "    \n",
    "    \n",
    "    def compute_feature_box_color(self, feature):\n",
    "        \"\"\"change colour of feature box border\"\"\"\n",
    "        if feature.type == \"TRAMP_probe_tested\":\n",
    "            if feature.qualifiers.get(\"label\")[0] == \"NLP7#7\":                \n",
    "                #col = colour_dict[\"NLP6/7\"]\n",
    "                col = \"#228833\"\n",
    "            elif feature.qualifiers.get(\"label\")[0] == \"NLP7#9\":\n",
    "                #col = colour_dict[\"DREB26\"]\n",
    "                col = \"#e69f00\"\n",
    "            elif feature.qualifiers.get(\"label\")[0] == \"NLP7#10\":\n",
    "                #col = colour_dict[\"ANAC032*\"]\n",
    "                col = \"#d55e00\"\n",
    "            elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#4\":\n",
    "                #col = colour_dict[\"ANAC032*\"]\n",
    "                col = \"#d55e00\"\n",
    "            elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#8\":\n",
    "                #col = colour_dict[\"ARF9/18*\"]\n",
    "                col = \"#2271B2\"\n",
    "            elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#3\":\n",
    "                #col = colour_dict[\"TGA1\"]\n",
    "                col = \"#F748A5\"\n",
    "            else:        \n",
    "                return \"black\"\n",
    "                        \n",
    "            return col\n",
    "            #'ARF9/18*': '#2271B2', 'ANR1*': '#3DB7E9', 'TGA1': '#F748A5', 'ANAC032*': '#d55e00', 'DREB26': '#e69f00', 'NLP6/7': '#228833'\n",
    "\n",
    "    def compute_feature_open_left(self, feature):\n",
    "        \"\"\"set to true if feature does not end on the left\"\"\"\n",
    "        return False\n",
    "\n",
    "    def compute_feature_open_right(self, feature):\n",
    "        \"\"\"set to true if feature does not end on the right\"\"\"\n",
    "        return False\n",
    "\n",
    "    def compute_feature_label(self, feature):\n",
    "        \"\"\"Remove most feature labels\"\"\"\n",
    "        if feature.type == 'start_codon':\n",
    "            return \"ATG\"\n",
    "        if feature.type == 'TRAMP_probe_tested':\n",
    "            if probe_labels is True:\n",
    "                return feature.qualifiers.get(\"label\")[0]     \n",
    "        else:\n",
    "            pass\n",
    "\n",
    "        #return super().compute_feature_label(feature)\n",
    "    # def compute_feature_min_y_height_of_text_line(self, feature):\n",
    "    #     return 0.1\n",
    "\n",
    "    def compute_feature_fontdict(self, feature):\n",
    "        \"\"\"change label font to arial, size to 10\"\"\"\n",
    "        if feature.type == \"TRAMP_probe_tested\":\n",
    "            #if certain label, align to the right\n",
    "            if feature.qualifiers.get(\"label\")[0] == \"ANAC032#3\":            \n",
    "                return dict(family='sans-serif',size=10, ha='right')\n",
    "            # elif feature.qualifiers.get(\"label\")[0] == \"ANAC032#8\":\n",
    "            #     return dict(family='sans-serif',size=10, ha='left')\n",
    "            if feature.qualifiers.get(\"label\")[0] == \"ANAC032#8\":            \n",
    "                return dict(family='sans-serif',size=10)\n",
    "            else:\n",
    "                return dict(family='sans-serif',size=10)\n",
    "\n",
    "        else:\n",
    "            return dict(family='sans-serif',size=10)\n",
    "        #return dict([('family','sans-serif'),('sans-serif','Arial'),('size',10)])\n",
    "\n",
    "    #make feature_list so that each feature name is only added once if more than one share the same name\n",
    "    #feature_list = []\n",
    "    def compute_feature_legend_text(self, feature):\n",
    "        \"\"\"add legend if feature label has not been added to legend already\"\"\"\n",
    "        if feature.type=='promoter':\n",
    "            pass\n",
    "        # elif feature.type=='exon':\n",
    "        #     pass\n",
    "        # elif feature.type=='intron':\n",
    "        #     pass\n",
    "        # elif feature.type==\"5'UTR\":\n",
    "        #     pass\n",
    "        # elif feature.qualifiers.get(\"label\")[0] in self.feature_list:\n",
    "        #     pass\n",
    "        elif feature.qualifiers.get(\"label\")[0] in feature_list:\n",
    "            pass\n",
    "        else:\n",
    "            \n",
    "            feature_list.append(feature.qualifiers.get(\"label\")[0])\n",
    "                 \n",
    "            #feature_list.append(feature.qualifiers.get(\"label\")[0])\n",
    "            return feature.qualifiers.get(\"label\")[0]     \n",
    "\n",
    "\n",
    "    def compute_filtered_features(self, features):\n",
    "        \"\"\"Do not display the following features\"\"\"\n",
    "        return [\n",
    "            feature for feature in features\n",
    "            if (feature.type != \"TRAMP_probe\")\n",
    "            and (feature.type != \"none\")\n",
    "            and (feature.type != \"DHS\")\n",
    "            and (feature.type != \"misc_feature\")\n",
    "            and (feature.type != \"primer\")\n",
    "            #and (feature.type != \"gene\")\n",
    "            and (feature.type != \"mRNA\")\n",
    "            and (feature.type != \"CDS\")\n",
    "            and (feature.type != \"source\")\n",
    "            and (feature.type != \"misc_RNA\")\n",
    "            and (feature.qualifiers.get(\"label\")[0] != \"ARID5_ARID6\"\n",
    "            and (feature.qualifiers.get(\"label\")[0] != \"ARDI5_ARID6\")\n",
    "            and ('Translation' not in feature.qualifiers.get(\"label\")[0]))\n",
    "            ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [],
   "source": [
    "#code from https://github.com/biopython/biopython/blob/c733feb84e7c48a54921f2c2ace2c57efdd97c8c/Bio/SeqUtils/SeqSlicerWithFeatures.py\n",
    "#this allows slicing of a sequence while keeping truncated annotations\n",
    "\n",
    "\n",
    "# from Bio import SeqIO\n",
    "\n",
    "\n",
    "def move_position(pos, move: int, keep_all: bool, slice_len: int):\n",
    "    tt = type(pos)\n",
    "    if isinstance(pos, (ExactPosition, BeforePosition, AfterPosition)):\n",
    "        if int(pos) + move < 0:\n",
    "            return BeforePosition(0)\n",
    "        elif int(pos) + move > slice_len:\n",
    "            return AfterPosition(slice_len)\n",
    "        else:\n",
    "            return tt(pos + move)\n",
    "    elif isinstance(pos, (BetweenPosition, WithinPosition)):\n",
    "        # This would need update if BetweenPosition or WithinPosition gets obsolete or changed\n",
    "        # attributes accessed because they are not exposed in any other way\n",
    "        diff_left = pos._left + move\n",
    "        diff_right = pos._right + move\n",
    "\n",
    "        if diff_left < 0 and diff_right <= 0:\n",
    "            return BeforePosition(0)\n",
    "        elif diff_left < 0 < diff_right:\n",
    "            if keep_all:\n",
    "                tt(0, 0, pos._right)\n",
    "            else:\n",
    "                return None\n",
    "        elif diff_left > 0 and diff_right > 0:\n",
    "            if diff_left < slice_len and diff_right < slice_len:\n",
    "                return AfterPosition(slice_len)\n",
    "            elif diff_left < slice_len < diff_right:\n",
    "                if keep_all:\n",
    "                    return tt(slice_len, diff_left, slice_len)\n",
    "                else:\n",
    "                    return None\n",
    "            else:\n",
    "                # This should not be reached\n",
    "                raise ValueError\n",
    "        else:\n",
    "            # This should not be reached\n",
    "            raise ValueError\n",
    "    else:\n",
    "        # raise for not known Position classes\n",
    "        raise NotImplementedError('slicing for position {} not implemented'.format(type(pos)))\n",
    "\n",
    "\n",
    "def start_violation_pos(x, sl: slice):\n",
    "    return x.start.position < sl.start < x.end.position\n",
    "\n",
    "\n",
    "def end_violation_pos(x, sl: slice):\n",
    "    return x.start.position < sl.stop < x.end.position\n",
    "\n",
    "\n",
    "def slice_feature(feature: FeatureLocation, sl: slice, keep_all_features: bool, seq_len: int):\n",
    "    if isinstance(feature, CompoundLocation):\n",
    "        of = []\n",
    "        for ff in feature.parts:\n",
    "            # this must be without location boundaries checking\n",
    "            #  - only location objects adjacent to slice boundaries can cross it\n",
    "\n",
    "            # However we need to check whether the location object is fully inside the slice\n",
    "            #  (and discard those which are not)\n",
    "\n",
    "            mps = move_position(ff.start, - sl.start, keep_all_features, seq_len)\n",
    "            mpe = move_position(ff.end, - sl.start, keep_all_features, seq_len)\n",
    "            if mps is not None and mpe is not None:\n",
    "                fl = FeatureLocation(mps, mpe, ff.strand)\n",
    "\n",
    "            # Feature locations outside of slice gets discarded\n",
    "            if fl is not None:\n",
    "                of.append(fl)\n",
    "\n",
    "        return CompoundLocation(of, operator=feature.operator)\n",
    "    else:\n",
    "        if start_violation_pos(feature, sl) or end_violation_pos(feature, sl):\n",
    "\n",
    "            mps = move_position(feature.start, - sl.start, keep_all_features, seq_len)\n",
    "            mpe = move_position(feature.end, - sl.start, keep_all_features, seq_len)\n",
    "\n",
    "            if mps is not None and mpe is not None:\n",
    "\n",
    "                fl = FeatureLocation(mps, mpe, feature.strand)\n",
    "                return fl\n",
    "            else:\n",
    "                return None\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "\n",
    "def slice_sequence_with_features(seq: SeqRecord, sl: slice, keep_all_features: bool = False) -> SeqRecord:\n",
    "    \"\"\"\n",
    "    This is an attempt to make slicing function for Bio.Seq which retains features which are not fully in slice\n",
    "      it is intended for visualization purposes where part of a feature also provides valuable information\n",
    "    # location variants:\n",
    "    - CompoundLocation - Collection of FeatureLocation objects (for joins etc).             ? -> solved\n",
    "    - ExactPosition - Specify the position as being exact.                                  OK\n",
    "    - WithinPosition - Specify a position occurring within some range.                      ? -> solved\n",
    "    - BetweenPosition - Specify a position occurring between a range (OBSOLETE?).           ? -> solved\n",
    "    - BeforePosition - Specify the position as being found before some base.                OK\n",
    "    - AfterPosition - Specify the position as being found after some base.                  OK\n",
    "    - OneOfPosition - Specify a position where the location can be multiple positions.      ? Not is spec ? -> raise\n",
    "    - UnknownPosition - Represents missing information like '?' in UniProt.                 raise\n",
    "    \"\"\"\n",
    "    ns = seq[sl.start:sl.stop]\n",
    "\n",
    "    seq_len = len(ns)\n",
    "    for feat in seq.features:\n",
    "        fl = slice_feature(feat.location, sl, keep_all_features, seq_len)\n",
    "        if fl is not None:\n",
    "            nf = SeqFeature(\n",
    "                fl,\n",
    "                type=feat.type,\n",
    "                id=feat.id,\n",
    "                qualifiers=feat.qualifiers,\n",
    "                ref=feat.ref,\n",
    "                ref_db=feat.ref_db\n",
    "            )\n",
    "            ns.features.append(nf)\n",
    "\n",
    "    return ns\n",
    "\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "\n",
    "#     # to my understanding\n",
    "#     # the BetweenPosition should be OK (not DEPRECATED)\n",
    "#     #    - as the between position is only for (n, n+1) - slicing should only retain full feature\n",
    "#     # the within position (usage restricted)\n",
    "#     #    - the within position must be inspected\n",
    "#     # One-of does not exist in specification\n",
    "#     #    - I do not expect it to see it\n",
    "#     # Unknown position not considered (maybe do some basic testing that presence would not break the function)\n",
    "\n",
    "#     # this fails for unknown reason\n",
    "#     # from Bio import SeqIO\n",
    "#     # import os\n",
    "#     # if os.path.exists('../../Tests/GenBank/NC_005816.gb'):\n",
    "#     #     print('ok')\n",
    "\n",
    "#     # using example record from Biopython Tests\n",
    "#     # a = SeqIO.read('../../Tests/GenBank/NC_005816.gb', format='gb')\n",
    "#     # b = slice_sequence_with_features(a, slice(5933, 5935))\n",
    "#     # b = slice_sequence_with_features(a, slice(5932, 5934))\n",
    "#     # b = slice_sequence_with_features(a, slice(5933, 5934))\n",
    "#     # b = slice_sequence_with_features(a, slice(1370, 1670))\n",
    "#     # b = slice_sequence_with_features(a, slice(1430, 1460))\n",
    "#     # b = slice_sequence_with_features(a, slice(1438, 1619))\n",
    "#     # for f in b.features:\n",
    "#     #     print(f)\n",
    "\n",
    "#     print('a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_record(seqrecord,promoter_name):\n",
    "    \"\"\"Preprocess the biopython record before feeding it into the translator.\"\"\"\n",
    "    #get length of the whole sequence\n",
    "    #seq_length = len(seqrecord.seq)\n",
    "    # #print(f'{seqrecord.id} + {seq_length}')\n",
    "    # #change seqrecord locations\n",
    "    # #if start location is greater than TSS position then the new location is the length of the whole sequence minus the original position\n",
    "    # def convert_location(location, TSS_position, seq_length):\n",
    "    #     \"\"\"convert locations to be relative to TSS position (or another locaton if you choose)\"\"\"\n",
    "    #     if location > TSS_position:\n",
    "    #         new_location = seq_length - location\n",
    "    #     elif location < TSS_position:\n",
    "    #         new_location = -(TSS_position-location)\n",
    "    #     elif location == TSS_position:\n",
    "    #         new_location = 0\n",
    "    #     return new_location\n",
    "    # print(seqrecord)\n",
    "    #shorten the seqrecord to only 1176bp after the TSS\n",
    "    for feature in seqrecord.features:\n",
    "        # if feature.type == 'source':\n",
    "        #     old_start = feature.location.start\n",
    "        #     old_end = feature.location.end\n",
    "            \n",
    "        if feature.type == 'promoter':\n",
    "            #start at beginning of the promoter, end 1176bp after the end of the promoter\n",
    "            start = feature.location.start\n",
    "            promoter_end = feature.location.end\n",
    "            end = promoter_end + 1175\n",
    "    #cycle through features, keeping the index information\n",
    "    #then rename upstream genes that are upstream of the transcription start site (end of promoter)\n",
    "    def rename_upstream_features(TSS_location,feature_name,seqrecord):\n",
    "        \"function to rename features that are upstream of the transcription start site (end of promoter)\"\n",
    "        for idx, feature in enumerate(seqrecord.features):\n",
    "            if feature.type == feature_name:\n",
    "                #if gene is upstream of the TSS, rename feature to gene_upstream\n",
    "                feature_start=feature.location.start\n",
    "                feature_end=feature.location.end            \n",
    "                feature_qualifiers = feature.qualifiers\n",
    "                if feature_start < TSS_location:\n",
    "                    #print(f'renaming upstream gene annotation')\n",
    "                    seqrecord.features[idx]=SeqFeature(FeatureLocation(feature_start,feature_end), type=f\"{feature_name}_upstream\", qualifiers=feature_qualifiers)  \n",
    "    #rename upstream features\n",
    "    rename_upstream_features(promoter_end,'gene',seqrecord)\n",
    "    rename_upstream_features(promoter_end,'exon',seqrecord)\n",
    "    #rename_upstream_features(promoter_end,'misc_RNA',seqrecord)\n",
    "    rename_upstream_features(promoter_end,'mRNA',seqrecord)\n",
    "    rename_upstream_features(promoter_end,'CDS',seqrecord)\n",
    "            \n",
    "    #print(f'start={start}')\n",
    "    #slice the seqrecord using the new start and end, keeping truncated annotations\n",
    "    shortened_seqrecord = slice_sequence_with_features(seqrecord, slice(start, end))\n",
    "    #shortened_seqrecord = seqrecord[start:end]\n",
    "    #add meta data\n",
    "    shortened_seqrecord.id = seqrecord.id + \"_shortened\"\n",
    "    shortened_seqrecord.annotations = seqrecord.annotations\n",
    "    #shorten the annotations start and end locations\n",
    "    keywords = shortened_seqrecord.annotations[\"keywords\"][0]\n",
    "    #print(f'keywords = {keywords}')\n",
    "    old_start_index = int(re.findall(r\"start_index:(\\d+)\",keywords)[0])\n",
    "   # old_end_index = int(re.findall(r\"end_index:(\\d+)\",keywords)[0])\n",
    "    #get difference between promoter start and original genbank fil start\n",
    "\n",
    "\n",
    "    new_start_index = old_start_index+start\n",
    "    new_end_index = old_start_index+end\n",
    "\n",
    "    new_keywords = [f'\"end_index:{new_end_index}\" \"start_index:{new_start_index}\"']\n",
    "\n",
    "    shortened_seqrecord.annotations[\"keywords\"] = new_keywords\n",
    "    #save shortened seq_record named after its ID\n",
    "    SeqIO.write(shortened_seqrecord, f\"../../data/TRAMP/{promoter_name}_shortened.gb\", \"genbank\")\n",
    "\n",
    "    #Ensure that genbank files are reverse complemented so that plots can be aligned to the right in figure\n",
    "    new_seqrecord = shortened_seqrecord.reverse_complement(id=shortened_seqrecord.id +\"_rc\")\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    for feature in new_seqrecord.features:\n",
    "        #change strand to None so that features are rectangular\n",
    "        feature.location.strand = None\n",
    "             \n",
    "        if feature.type == 'TFBS':\n",
    "            #print(feature)\n",
    "            #change sigil to box\n",
    "            feature.qualifiers[\"sigil\"] = 'OCTAGON'\n",
    "            #increase width of TFBS so colour is more visible\n",
    "            TFBS_start = feature.location.start\n",
    "            TFBS_end = feature.location.end\n",
    "            #find middle of TFBS\n",
    "            middle = (TFBS_end-TFBS_start)//2 #floor division creating an integar\n",
    "            \n",
    "            new_start = TFBS_start-8+middle\n",
    "            new_end = TFBS_end-middle+8\n",
    "            feature.location = FeatureLocation(new_start,new_end)\n",
    "            #change name of some TFBSs\n",
    "            if feature.qualifiers.get(\"label\")[0] == 'ANR1_AGL16':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'ANR1*'\n",
    "            elif feature.qualifiers.get(\"label\")[0] == 'ANAC032_NAC002':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'ANAC032*'\n",
    "            elif feature.qualifiers.get(\"label\")[0] == 'ANAC032':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'ANAC032*'\n",
    "            elif feature.qualifiers.get(\"label\")[0] == 'ARF18/9_ARF2':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'ARF9/18*'\n",
    "            elif feature.qualifiers.get(\"label\")[0] == 'ARF9/18':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'ARF9/18*'\n",
    "            elif feature.qualifiers.get(\"label\")[0] == 'NLP7':\n",
    "                feature.qualifiers.get(\"label\")[0] = 'NLP6/7'\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "            \n",
    "            #if feature.type == \"gene_upstream\":\n",
    "\n",
    "        #elif feature.type == \"mRNA_upstream\":\n",
    "           # return 0\n",
    "       # elif feature.type == \"exon_upstream\":\n",
    "        #     return 0\n",
    "        # elif feature.type == \"misc_RNA_upstream\":\n",
    "\n",
    "                \n",
    "         \n",
    "\n",
    "            #if feature was experimentally validated, add qualifier\n",
    "    \n",
    "            \n",
    "    return new_seqrecord\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gb_file_to_seqrecord(promoter_name):\n",
    "    \"\"\"load genbankfile into a seqrecord\"\"\"\n",
    "    #file location\n",
    "    gb_file=f\"../../data/TRAMP/{promoter_name}.gb\"\n",
    "    record = SeqIO.read(gb_file, 'genbank')\n",
    "    #preprocess record\n",
    "    modified_seqrecord = preprocess_record(record,promoter_name)\n",
    "    return modified_seqrecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "metadata": {},
   "outputs": [],
   "source": [
    "def RGB2CMYK(image_file_location, image_file_extension):\n",
    "    \"\"\"convert image from RGB to CMYK colours\"\"\"\n",
    "    # Import image\n",
    "    image = Image.open(image_file_location+image_file_extension)\n",
    "    print(image.mode)\n",
    "    if image.mode == 'RGBA':\n",
    "        print('converting_image')\n",
    "        cmyk_image = image.convert('CMYK')\n",
    "        cmyk_image.save(image_file_location+'_CMYK'+image_file_extension)\n",
    "    #close image\n",
    "    image.close()\n",
    "    # img = plt.imread(image_file_location+image_file_extension)\n",
    "    # #print(image_file_location+image_file_extension)\n",
    "    # # Create float\n",
    "    # bgr = img.astype(float)/255.\n",
    "\n",
    "    # # Extract channels\n",
    "    # with np.errstate(invalid='ignore', divide='ignore'):\n",
    "    #     K = 1 - np.max(bgr, axis=2)\n",
    "    #     C = (1-bgr[...,2] - K)/(1-K)\n",
    "    #     M = (1-bgr[...,1] - K)/(1-K)\n",
    "    #     Y = (1-bgr[...,0] - K)/(1-K)\n",
    "\n",
    "    # # Convert the input BGR image to CMYK colorspace\n",
    "    # CMYK = (np.dstack((C,M,Y,K)) * 255).astype(np.uint8)\n",
    "    # cv2.imwrite(image_file_location+'_CMYK'+image_file_extension, CMYK)\n",
    "\n",
    "    # # Split CMYK channels\n",
    "    # Y, M, C, K = cv2.split(CMYK)\n",
    "\n",
    "    # np.isfinite(C).all()\n",
    "    # np.isfinite(M).all()\n",
    "    # np.isfinite(K).all()\n",
    "    # np.isfinite(Y).all()\n",
    "\n",
    "    # # Save channels\n",
    "    # cv2.imwrite('C:/path/to/C.jpg', C)\n",
    "    # cv2.imwrite('C:/path/to/M.jpg', M)\n",
    "    # cv2.imwrite('C:/path/to/Y.jpg', Y)\n",
    "    # cv2.imwrite('C:/path/to/K.jpg', K)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #create a class specifying feature colours\n",
    "# #make feature_list and colour_dict so that each feature name and colour is only added once to legend if more than one share the same name\n",
    "\n",
    "# feature_list = []\n",
    "# colour_dict = {}\n",
    "# class MyCustomTranslator(BiopythonTranslator):\n",
    "#     \"\"\"Custom translator iplementing the following theme:\n",
    "#     -Colour promoter in pale green\n",
    "#     -colour exons in dark grey\n",
    "#     -colour introns in light grey\n",
    "#     -colour TFs from colour palette\"\"\"\n",
    "\n",
    "#     #import colour blind palette\n",
    "#     #colour palette from mkweb.bcgsc.ca/colorblind/palettes.mhtml\n",
    "#     # df = pd.read_csv(\"colourblind_palette.csv\", header=0)\n",
    "#     # #convert to floats\n",
    "#     # floats=df.divide(255)\n",
    "#     # #make sets of each row to get the red, green blue colours\n",
    "#     # CB_colour_palette = floats.apply(tuple, axis = 1)\n",
    "#     # #make into df\n",
    "#     # df = CB_colour_palette.to_frame()\n",
    "#     # #save file\n",
    "#     # df.to_csv('../../data/TRAMP/colour_list')\n",
    "    \n",
    "#     # #turn into a list of colour sets\n",
    "#     # list_colours = list(CB_colour_palette)\n",
    "#     # colour_list = list_colours\n",
    "#     #colour_list = ['#88CCEE', '#44AA99', '#117733', '#332288', '#DDCC77', '#999933','#CC6677', '#882255', '#AA4499', '#DDDDDD']\n",
    "#     colour_list = ['#2271B2',\n",
    "#     '#3DB7E9',\n",
    "#     '#F748A5',    \n",
    "#     '#d55e00',\n",
    "#     '#e69f00',\n",
    "#     '#f0e442',\n",
    "#     '#000000',]\n",
    "#     #make colour iterator\n",
    "#     colour_iterator=cycle(colour_list)\n",
    "#     #change colour cycle\n",
    "#     #mpl.rcParams['axes.prop_cycle'] = mpl.cycler(color=list_colours)\n",
    "#     #count = -1\n",
    "#     #set colour count\n",
    "\n",
    "\n",
    "#     def compute_feature_color(self, feature):\n",
    "#         \"\"\"set colour of each feature\"\"\"      \n",
    "\n",
    "#         if feature.type == \"promoter\":\n",
    "#             return \"#359B73\"\n",
    "#         elif feature.type == \"exon\":\n",
    "#             return \"darkgrey\"\n",
    "#             #return (169,169,169)\n",
    "#         elif feature.type == \"intron\":\n",
    "#             return \"lightgrey\"  \n",
    "#             #return (211,211,211)\n",
    "#         elif feature.type == \"5'UTR\":\n",
    "#             return \"lightgrey\"\n",
    "#         elif feature.type == \"TFBS\":\n",
    "#             if feature.qualifiers.get(\"label\")[0] in colour_dict.keys():\n",
    "#                 col = colour_dict[feature.qualifiers.get(\"label\")[0]]\n",
    "#             else:\n",
    "#                 col = next(self.colour_iterator)\n",
    "#                 colour_dict[feature.qualifiers.get(\"label\")[0]] = col\n",
    "                        \n",
    "#             return col\n",
    "#         else:\n",
    "#             return \"gold\"\n",
    "\n",
    "#     def compute_feature_box_linewidth(self, feature):\n",
    "#         \"\"\"change shape of features\"\"\"\n",
    "#         return 0\n",
    "\n",
    "#     def compute_feature_label(self, feature):\n",
    "#         \"\"\"Remove feature labels\"\"\"\n",
    "#         #return super().compute_feature_label(feature)\n",
    "        \n",
    "\n",
    "#     def compute_feature_fontdict(self, feature):\n",
    "#         \"\"\"change font to arial, size to 10\"\"\"\n",
    "#         return dict(family='sans-serif',size=10)\n",
    "#         #return dict([('family','sans-serif'),('sans-serif','Arial'),('size',10)])\n",
    "\n",
    "#     #make feature_list so that each feature name is only added once if more than one share the same name\n",
    "#     #feature_list = []\n",
    "#     def compute_feature_legend_text(self, feature):\n",
    "#         \"\"\"add legend if feature label has not been added to legend already\"\"\"\n",
    "#         if feature.type=='promoter':\n",
    "#             pass\n",
    "#         elif feature.type=='exon':\n",
    "#             pass\n",
    "#         elif feature.type=='intron':\n",
    "#             pass\n",
    "#         elif feature.type==\"5'UTR\":\n",
    "#             pass\n",
    "#         # elif feature.qualifiers.get(\"label\")[0] in self.feature_list:\n",
    "#         #     pass\n",
    "#         elif feature.qualifiers.get(\"label\")[0] in feature_list:\n",
    "#             pass\n",
    "#         else:        \n",
    "\n",
    "#             feature_list.append(feature.qualifiers.get(\"label\")[0])\n",
    "                 \n",
    "#             #feature_list.append(feature.qualifiers.get(\"label\")[0])\n",
    "#             return feature.qualifiers.get(\"label\")[0]     \n",
    "\n",
    "\n",
    "#     def compute_filtered_features(self, features):\n",
    "#         \"\"\"Do not display features the following features\"\"\"\n",
    "#         return [\n",
    "#             feature for feature in features\n",
    "#             if (feature.type != \"TRAMP_probe\")\n",
    "#             and (feature.type != \"none\")\n",
    "#             and (feature.type != \"DHS\")\n",
    "#             and (feature.type != \"misc_feature\")\n",
    "#             and (feature.type != \"primer\")\n",
    "#             and (feature.type != \"gene\")\n",
    "#             and (feature.qualifiers.get(\"label\")[0] != \"ARID5_ARID6\"\n",
    "#             and (feature.qualifiers.get(\"label\")[0] != \"ARDI5_ARID6\"))\n",
    "#             ]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def genometrack_plot(file_location):\n",
    "#     \"\"\"make pyGenomeTracks plot\"\"\"\n",
    "#     fig, axs = plt.subplots(2,1,sharex='col')\n",
    "#     region = '1',3000,6000\n",
    "#     chrom_region,start_region,end_region = region\n",
    "#     ax = axs[1]\n",
    "\n",
    "#     track_config = dict(\n",
    "#         file=file_location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def make_plot2(dictionary_of_records_seqrecords,dictionary_of_records_promoternames):\n",
    "#     #length of dictionary\n",
    "#     length_dict = len(dictionary_of_records_promoternames)  \n",
    "#     height = length_dict//2\n",
    "#     fig, axs = plt.subplots(length_dict,1,sharex='none',figsize=(10,height*2))\n",
    "#     def pygenometracks(chrom_region,start_region,end_region,axis):\n",
    "#                 \"\"\"make pygenome track for bigwig files\"\"\"        \n",
    "#                 file_location_roots=\"../../data/ATAC-seq/potter2018/GSE116287_Roots_NaOH_Merged.bw\"\n",
    "#                 file_location_shoots=\"../../data/ATAC-seq/potter2018/GSE116287_Shoots_NaOH_Merged.bw\"\n",
    "#                 track_config_roots = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown')\n",
    "#                 track_config_shoots = dict(file=file_location_shoots,overlay_previous = 'share-y',alpha = 0.5)#alpha is transparency\n",
    "#                 tk_roots = pygtk.BigWigTrack(track_config_roots)\n",
    "#                 tk_shoots = pygtk.BigWigTrack(track_config_shoots)\n",
    "#                 tk_roots.plot(axis,chrom_region,start_region,end_region,)\n",
    "#                 tk_shoots.plot(axis,chrom_region,start_region,end_region,)\n",
    "#             #make graphic records of genes and TFBSs\n",
    "#     for k,v in dictionary_of_records_seqrecords.items():\n",
    "#         #take last character of string, double it as twice as many sequence tracks to include open chromatin atacseq data\n",
    "#         last = int(k[-1])\n",
    "#         #get promoter name\n",
    "#         prom_name = dictionary_of_records_promoternames[k]\n",
    "#         #split on dashes\n",
    "#         #short_prom_name = prom_name.split('-')[0].upper()\n",
    "#        #make_graphic_record(v,short_prom_name,axsRight[last-2]\n",
    "#         #print(v.id)\n",
    "#         #add atacseq track\n",
    "#         #first get chromosome number from sequence ID\n",
    "#         AGI = v.id[v.id.find('(')+1:v.id.find(')')]\n",
    "#         chrom_region = AGI[2]\n",
    "#         #then get start and stop region\n",
    "#         #open genbank file, read third line containing keywords\n",
    "#         gb_file=f\"../../data/TRAMP/{prom_name}.gb\"\n",
    "#         with open(gb_file, \"r\") as f:\n",
    "#             all_lines = f.readlines()\n",
    "#             start_region = int(re.findall(r\"start_index:(\\d+)\",all_lines)[0])\n",
    "#             end_region = int(re.findall(r\"end_index:(\\d+)\",all_lines)[0])\n",
    "\n",
    "#         #keywords = re.findall(r\"KEYWORDS.*\"   \n",
    "        \n",
    "#         print(\"start=\"+str(start_region)+\"end=\"+str(end_region))        \n",
    "#         pygenometracks(chrom_region,start_region,end_region,axs[last-1])\n",
    "#         #add titles\n",
    "#         short_prom_name = prom_name.split('-')[0].upper()\n",
    "#         axs[last-1].set_title(short_prom_name)\n",
    "#         #change xvalues of open chromatin track\n",
    "#         region_length = end_region-start_region\n",
    "#         ax1Ticks = axs[last-1].get_xticks()   \n",
    "#         #ax2Ticks = ax1Ticks\n",
    "#         def tick_function(X,start_region):\n",
    "#             V = X-start_region\n",
    "#             return [\"%.3f\" % z for z in V]\n",
    "#         #axs[last-1].set_xticks(ax2Ticks)\n",
    "#         #ax2.set_xbound(ax1.get_xbound())\n",
    "#         axs[last-1].set_xticklabels(tick_function(ax1Ticks,start_region))\n",
    "\n",
    "#         #plt.xticks(np.arange(end_region-start_region),np.arange(end_region-start_region))\n",
    "#     #make x axes start at 1\n",
    "    \n",
    "#     fig.tight_layout()\n",
    "#     # chrom_region,start_region,end_region = region\n",
    "#     # ax = axs[1]\n",
    "#     # file_location_roots=\"../../data/ATAC-seq/potter2018/GSE116287_Roots_NaOH_Merged.bw\"\n",
    "#     # file_location_shoots=\"../../data/ATAC-seq/potter2018/GSE116287_Shoots_NaOH_Merged.bw\"\n",
    "#     # track_config_roots = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown')\n",
    "#     # track_config_shoots = dict(file=file_location_shoots,overlay_previous = 'share-y',alpha = 0.5)#alpha is transparency\n",
    "#     # tk_roots = pygtk.BigWigTrack(track_config_roots)\n",
    "#     # tk_shoots = pygtk.BigWigTrack(track_config_shoots)\n",
    "#     # tk_roots.plot(axs[0],chrom_region,start_region,end_region,)\n",
    "#     # tk_shoots.plot(axs[0],chrom_region,start_region,end_region,)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def make_plot(dictionary_of_records_seqrecords,dictionary_of_records_promoternames,dir_name,atacseq=True):\n",
    "#     def make_graphic_record(seqrecord, promoter_name, ax,short_annotation=True,title=True):\n",
    "#         \"\"\"make a graphic record object\"\"\"\n",
    "#         #display figure\n",
    "#         graphic_record = MyCustomTranslator().translate_record(seqrecord)\n",
    "#         #graphic_record.labels_spacing = -5\n",
    "#         #set minimum height of annotations\n",
    "#         if short_annotation is True:\n",
    "#             graphic_record.min_y_height_of_text_line = 0.5\n",
    "#         else:\n",
    "#             graphic_record.min_y_height_of_text_line = 0.5\n",
    "#             # graphic_record.labels_spacing = 10\n",
    "\n",
    "#         #set spacing between labels\n",
    "#         #graphic_record.labels_spacing = -5\n",
    "#         #change height to 0 so TFBS can overlap the promoter\n",
    "#         graphic_record.feature_level_height = 0\n",
    "#         #graphic_record = BiopythonTranslator().translate_record(gb_file)\n",
    "#         graphic_record.plot(ax=ax, with_ruler=True,annotate_inline=True,)#,figure_width=10, #strand_in_label_threshold=4,annotate_inline=True so that labels are within the feature\n",
    "#         #add title of promoter\n",
    "#         if title is True:\n",
    "#             ax.title.set_text(promoter_name)\n",
    "#         #return graphic_record\n",
    "#     #set plot parameters\n",
    "#     rcParams['xtick.major.width'] = 2\n",
    "#     rcParams['ytick.major.width'] = 2\n",
    "#     rcParams['font.family'] = 'sans-serif'\n",
    "#     rcParams['font.sans-serif'] = ['Arial']\n",
    "#     #allow font to be edited later in pdf editor\n",
    "#     rcParams ['pdf.fonttype'] = 42\n",
    "\n",
    "#     #make svg text editable\n",
    "#     plt.rcParams['svg.fonttype'] = 'none'\n",
    "    \n",
    "#     #rcParams['axes.linewidth'] = 2\n",
    "#     #rcParams['lines.linewidth'] = 2\n",
    "#     #remove top and right lines\n",
    "#     # rcParams['axes.spines.top'] = False\n",
    "#     # rcParams['axes.spines.right'] = False\n",
    "#     #font size\n",
    "#     rcParams['font.size'] = 11\n",
    "\n",
    "\n",
    "#     #length of dictionary\n",
    "#     length_dict = len(dictionary_of_records_promoternames)\n",
    "#     #print(length_dict)\n",
    "#     #\n",
    "#     #if including atacseq track, include more subfigures\n",
    "#     if atacseq is True:\n",
    "#         #make plot\n",
    "#         height = length_dict//2+length_dict-1 #add length_dict-1 to include empty grids for spacing between pairs\n",
    "#         ### NEED TO MAKE INTO A GRID LIKE HERE: https://stackoverflow.com/questions/51717199/how-to-adjust-space-between-every-second-row-of-subplots-in-matplotlib\n",
    "\n",
    "        \n",
    "\n",
    "#         if length_dict < 4:\n",
    "#             fig = plt.figure(constrained_layout=False,figsize=(12,height+1))\n",
    "#         else:\n",
    "#             fig = plt.figure(constrained_layout=False,figsize=(11,height-5))\n",
    "        \n",
    "#         #make subfigures\n",
    "#         #subfigs = fig.subfigures(1,2, wspace=0.0, width_ratios=[1,5])\n",
    "#         #left legend \n",
    "#         #axsLeft=subfigs[0].subplots(1,1)\n",
    "#         #remove axis\n",
    "#         #axsLeft.axis('off')\n",
    "#         # #right figures\n",
    "#         #create gridspec so that open chromatin is paired with promoter\n",
    "#         n = length_dict # number of double-rows\n",
    "#         m = 2 # number of columns\n",
    "\n",
    "#         t = 0.9 # 1-t == top space \n",
    "#         b = 0.1 # bottom space      (both in figure coordinates)\n",
    "#         if length_dict < 4:\n",
    "#             msp = -0.7\n",
    "#             sp = 0.3\n",
    "\n",
    "#         else:\n",
    "#             msp = -0.3 # minor spacing\n",
    "#             sp = 0.3  # major spacing\n",
    "#         offs=(1+msp)*(t-b)/(2*n+n*msp+(n-1)*sp) # grid offset\n",
    "#         hspace = sp+msp+1 #height space per grid\n",
    "\n",
    "#         gso = GridSpec(n,m, bottom=b+offs, top=t, hspace=hspace,width_ratios=[2,10])\n",
    "#         gse = GridSpec(n,m, bottom=b, top=t-offs, hspace=hspace,width_ratios=[2,10])\n",
    "\n",
    "#         #fig = plt.figure()\n",
    "\n",
    "#         grid = []\n",
    "#         for i in range(n*m):\n",
    "#             grid.append(fig.add_subplot(gso[i]))\n",
    "#             grid.append(fig.add_subplot(gse[i]))\n",
    "        \n",
    "\n",
    "        \n",
    "#         #print(grid)\n",
    "#         #print(len(grid))\n",
    "#         #split plots into two lists - one for left column and one for the right\n",
    "#         #turn off axes for the plots on the left\n",
    "#         axsRight = []\n",
    "#         axsLeft = []\n",
    "#         count=0\n",
    "#         for number in np.arange(len(grid)):\n",
    "#             if count == 0:\n",
    "#                 axsLeft += [grid[number]]\n",
    "#                 grid[number].axis('off')\n",
    "#                 count +=1\n",
    "#             elif count == 1:\n",
    "#                 axsLeft += [grid[number]]\n",
    "#                 count +=1\n",
    "#                 grid[number].axis('off')\n",
    "#             elif count == 2:\n",
    "#                 axsRight += [grid[number]]\n",
    "#                 count +=1\n",
    "#             elif count ==3:\n",
    "#                 axsRight += [grid[number]]\n",
    "#                 count = 0\n",
    "\n",
    "#         #make legend span two plots \n",
    "#         axsLeft[0].set_position(gso[0:3].get_position(fig))\n",
    "#         #axsLeft[1].set_position(gso[0:3].get_position(fig)\n",
    "   \n",
    "                \n",
    "      \n",
    "                \n",
    "#         #move plots closeer to each other\n",
    "#         # for i in range(0,len(axsRight),2):\n",
    "#         #     axsRight[i] = plt.subplot()\n",
    "#         #print(axsRight)\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "#         # num_rows = length_dict*2\n",
    "#         # num_cols = 1\n",
    "#         # row_height = 6\n",
    "#         # space_height = 2\n",
    "#         # num_sep_rows = lambda x: int((x-1)/2)\n",
    "#         # grid = (row_height*num_rows + space_height*num_sep_rows(num_rows), num_cols)\n",
    "#         # #axsRight = subfigs[1].subplots(num_rows,1,sharex=False)\n",
    "#         # #axsRight = subfigs[1]\n",
    "#         # axsRight = []\n",
    "\n",
    "#         # for ind_row in range(num_rows):\n",
    "#         #     for ind_col in range(num_cols):\n",
    "#         #         grid_row = row_height*ind_row + space_height*num_sep_rows(ind_row+1)\n",
    "#         #         grid_col = ind_col\n",
    "\n",
    "#         # axsRight += [plt.subplot2grid(grid, (grid_row, grid_col), rowspan=row_height)]\n",
    "\n",
    "        \n",
    "#         def pygenometracks(chrom_region,start_region,end_region,axis):\n",
    "#             \"\"\"make pygenome track for bigwig files\"\"\"        \n",
    "#             file_location_roots=\"../../data/ATAC-seq/potter2018/GSE116287_Roots_NaOH_Merged.bw\"\n",
    "#             file_location_shoots=\"../../data/ATAC-seq/potter2018/GSE116287_Shoots_NaOH_Merged.bw\"            \n",
    "#             track_config_roots = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',alpha = 0.25,min_value=0, max_value=30)\n",
    "#             track_config_shoots = dict(file=file_location_shoots,overlay_previous = 'share-y',color='teal',alpha = 0.25,min_value=0,max_value=30)\n",
    "            \n",
    "#             #add lines too\n",
    "#             track_config_roots_lines = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',type=\"line:1\",min_value=0, max_value=30)\n",
    "#             track_config_shoots_lines = dict(file=file_location_shoots,overlay_previous = 'share-y',color='teal',type=\"line:1\",min_value=0, max_value=30)\n",
    "#             #alpha is transparency\n",
    "#             tk_roots = pygtk.BigWigTrack(track_config_roots)\n",
    "#             tk_shoots = pygtk.BigWigTrack(track_config_shoots)\n",
    "#             tk_roots_lines = pygtk.BigWigTrack(track_config_roots_lines)\n",
    "#             tk_shoots_lines = pygtk.BigWigTrack(track_config_shoots_lines)\n",
    "#             tk_roots.plot(axis,chrom_region,start_region,end_region,)\n",
    "#             tk_shoots.plot(axis,chrom_region,start_region,end_region,)\n",
    "#             tk_roots_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "#             tk_shoots_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "#         #make graphic records of genes and TFBSs\n",
    "#         for k,v in dictionary_of_records_seqrecords.items():\n",
    "#             #take last character of string, double it as twice as many sequence tracks to include open chromatin atacseq data\n",
    "#             last = int(k[-1])*2+1\n",
    "#             last_chromatin = int(k[-1])*2\n",
    "#             #get promoter name\n",
    "#             prom_name = dictionary_of_records_promoternames[k]\n",
    "#             #split on dashes\n",
    "#             short_prom_name = prom_name.split('-')[0].upper()\n",
    "#             #print(short_prom_name)\n",
    "#             if short_prom_name == \"ANAC032\":\n",
    "#                 #print(ANAC032)\n",
    "#                 make_graphic_record(v,short_prom_name,axsRight[last-2],short_annotation=False, title=False)\n",
    "#             else:\n",
    "#                 make_graphic_record(v,short_prom_name,axsRight[last-2],title=False)\n",
    "        \n",
    "#             #print(v.id)\n",
    "#             #add atacseq track\n",
    "#             #first get chromosome number from sequence ID\n",
    "#             chrom_region = re.findall(r\"chromosome:TAIR10:(\\d)\",v.id)[0]\n",
    "#             #chrom_region = v.id[v.id.find('TAIR10:')+1:v.id.find(':')]\n",
    "#             #print(v)\n",
    "#             #print(chrom_region)\n",
    "#             #chrom_region = AGI[2]\n",
    "#             #then get start and stop region\n",
    "#             #open genbank file, read third line containing keywords\n",
    "#             gb_file=f\"../../data/TRAMP/{prom_name}.gb\"\n",
    "#             with open(gb_file, \"r\") as f:\n",
    "#                 all_lines = f.readlines()\n",
    "#                 for line in all_lines:\n",
    "#                     if re.match(r'KEYWORDS', line):\n",
    "#                         keywords = line                \n",
    "#             start_region = int(re.findall(r\"start_index:(\\d+)\",keywords)[0])\n",
    "#             end_region = int(re.findall(r\"end_index:(\\d+)\",keywords)[0])\n",
    "#             #print(start_region)  \n",
    "#             pygenometracks(chrom_region,start_region,end_region,axsRight[last_chromatin-2])\n",
    "#             #set xlim\n",
    "#             offset_length = 4550-(end_region-start_region)\n",
    "#             axsRight[last_chromatin-2].set_xlim(start_region-offset_length,end_region)\n",
    "#             #get x and y lim\n",
    "#             #axsRight[last_chromatin-2].set_title(short_prom_name,y=0.5, ) #put title to left of axis\n",
    "#             ##setlocation of the title         \n",
    "#             #first transform offset_length to between 0 and 1 for axes location, and offset a little to the left\n",
    "#             trans = (offset_length-25)/4550\n",
    "\n",
    "            \n",
    "#             axsRight[last_chromatin-2].text(x=trans, y=0.125, s=short_prom_name, weight=\"extra bold\", fontsize=10,transform=axsRight[last-2].transAxes, ha='right')#transform=axsRight[last_chromatin-2].transAxes\n",
    "\n",
    "            \n",
    "#             #axsRight[last_chromatin-2].invert_xaxis()\n",
    "#             #change xvalues of open chromatin track      \n",
    "#             #first get x and y values\n",
    "#             #line = axsRight[last_chromatin-2].get_lines()\n",
    "#             #xd = line.get_xdata()\n",
    "#             #yd = line.get_ydata()\n",
    "#             #print(line)\n",
    "#             #ax1Ticks = axsRight[last_chromatin-2].get_xticks()   \n",
    "#            # ax2Ticks = ax1Ticks.copy()\n",
    "#             # def tick_function(X,start_region):\n",
    "#             #     V = X-start_region\n",
    "#             #     return [\"%.3f\" % z for z in V]\n",
    "\n",
    "#             #\n",
    "#             #axs[last-1].set_xticks(ax2Ticks)\n",
    "#             #ax2.set_xbound(ax1.get_xbound())\n",
    "#             #axsRight[last_chromatin-2].set_xticklabels(tick_function(ax2Ticks,start_region))\n",
    "#             #axsRight[last_chromatin-2].set_xscale('function', functions=(forward_function,inverse_function))\n",
    "#             #make x labels integars\n",
    "            \n",
    "\n",
    "#         colour_dict_sorted = {k: v for k,v in sorted(colour_dict.items(), key=lambda item: item[0])}\n",
    "#         #print(colour_dict_sorted)\n",
    "#         handles = []\n",
    "#         labels = []\n",
    "#         for TFBS,colour in colour_dict_sorted.items():\n",
    "#             addition = mpatches.Patch(color=colour)\n",
    "#             #append to handles list\n",
    "#             handles.append(addition)\n",
    "#             labels.append(TFBS)\n",
    "\n",
    "#         #use latex style rendering to allow parts to be bold\n",
    "        \n",
    "#         rc('text', usetex=True)\n",
    "\n",
    "#         #handles = sorted(handles)\n",
    "\n",
    "#         #create TFBS legend and align left\n",
    "#         #append open chromatin to handles\n",
    "#         #first create custom handler for string in legend to add a second title for open chromatin\n",
    "#         #used https://gist.github.com/Raudcu/44b43c7f3f893fe2f4dd900afb740b7f\n",
    "#         class LegendTitle(object):\n",
    "#             def __init__(self, text_props=None):\n",
    "#                 self.text_props = text_props or {}\n",
    "#                 super(LegendTitle, self).__init__()\n",
    "\n",
    "#             def legend_artist(self, legend, orig_handle, fontsize, handlebox):\n",
    "#                 x0, y0 = handlebox.xdescent, handlebox.ydescent\n",
    "#                 title = mtext.Text(x0, y0, r\"\\textbf{{{}}}\".format(orig_handle), **self.text_props)\n",
    "#                 handlebox.add_artist(title)\n",
    "#                 return title\n",
    "# #r\"\\textbf{'+orig_handle+'}\"\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "#         #label titles\n",
    "#         title1 = 'Candidate transcription factor binding sites' \n",
    "#         title2 = 'Gene features'\n",
    "#         title3 = 'Open chromatin peaks'\n",
    "#         #add colour patches\n",
    "#         intron = mpatches.Patch(color='lightgrey')\n",
    "#         exon = mpatches.Patch(color='#635147')\n",
    "#         upstream_mrna = mpatches.Patch(color='#DDCC77')\n",
    "#         root = Line2D([0],[0],color='brown', lw=2)\n",
    "#         shoot = Line2D([0],[0],color='teal',lw=2)\n",
    "\n",
    "#         #insert handles\n",
    "#         handles.insert(0,title1)\n",
    "#         labels.insert(0,'')\n",
    "#         handles.insert(3,title2)\n",
    "#         labels.insert(3,'')\n",
    "#         handles.insert(4,exon)\n",
    "#         labels.insert(4,'Exon')\n",
    "#         handles.insert(5,title3)\n",
    "#         labels.insert(5,'')\n",
    "#         handles.insert(6,root)\n",
    "#         labels.insert(6,'Root')\n",
    "#         #blank label to make layout look nice\n",
    "#         handles.insert(7,'')\n",
    "#         labels.insert(7,'')\n",
    "#         handles.insert(10,'')\n",
    "#         labels.insert(10,'')\n",
    "#         handles.insert(11,intron)\n",
    "#         labels.insert(11,'Intron')\n",
    "#         #blank label to make layout look nice \n",
    "#         handles.insert(12,'')\n",
    "#         labels.insert(12,'')\n",
    "#         handles.insert(13,shoot)\n",
    "#         labels.insert(13,'Shoot')\n",
    "#         #blank label to make layout look nice\n",
    "#         handles.insert(14,'')\n",
    "#         labels.insert(14,'')\n",
    "#         handles.insert(17,'')\n",
    "#         labels.insert(17,'')\n",
    "        \n",
    "#         handles.insert(18,upstream_mrna)\n",
    "#         labels.insert(18,'Upstream transcript')\n",
    "#         handles.insert(19,'')\n",
    "#         labels.insert(19,'')\n",
    "\n",
    "      \n",
    "\n",
    "        \n",
    "#         axsLeft[0].legend(handles=handles,labels=labels, loc='upper right',ncol=3,handler_map={str: LegendTitle({'fontsize': 14})})#handler_map={title: LegendTitle({'fontsize': 14})}\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "#         #axsLeft[0].legend(handles=handles,labels=labels, loc='upper right', title=r\"\\textbf{Transcription factor binding site}\", title_fontsize='14',ncol=2,)#handler_map={title: LegendTitle({'fontsize': 14})}\n",
    "#         # # Add the legend manually to the current Axes.\n",
    "\n",
    "#         #create open chromatin legend below\n",
    "#         #open_chrom_handles = [mpatches.Patch(facecolor='brown', edgecolor='brown', label='Root'),mpatches.Patch(facecolor='green', edgecolor='green', label='Shoot')]\n",
    "#         #axsLeft[4].legend(handles=open_chrom_handles, loc='lower right', title=r\"\\textbf{Open chromatin peaks}\", title_fontsize='14',ncol=2)\n",
    "\n",
    "#         # #turn off latex rendering of text\n",
    "#         rc('text', usetex=False)\n",
    "       \n",
    "#         #change x_lim to flip x axis\n",
    "#         for n in np.arange(length_dict):\n",
    "#             last = n*2+1\n",
    "            \n",
    "#             axsRight[last-2].set_xlim(4550,0)\n",
    "            \n",
    "#             #change font colour of x axis text\n",
    "#             axsRight[last-2].tick_params(axis='x', colors='black')\n",
    "#             #change width of line\n",
    "#             # for axis in ['bottom','right']:\n",
    "#             #     ax.spines[axis].set_linewidth(5)\n",
    "\n",
    "#             x_ticks = np.arange(176, 4550, 500) #start stop and step\n",
    "#             axsRight[last-2].set_xticks(x_ticks)\n",
    "#             fig.canvas.draw()\n",
    "#             # if axsRight[last-2] == axsRight[length_dict*2-1]:\n",
    "#             #     pass\n",
    "#             # else:\n",
    "#             #     #remove xticks\n",
    "#             #     #axsRight[last-2].xaxis.set_ticks_position('none')\n",
    "#             #     #remove axes\n",
    "#             #     ax.axis('off')\n",
    "                \n",
    "#         #remove all axes\n",
    "#         for ax in axsRight:            \n",
    "#             if ax == axsRight[length_dict*2-1]:\n",
    "#                 pass\n",
    "#             else:\n",
    "#                 #ax.xaxis.set_ticks_position('none')\n",
    "#                 ax.axis('off')\n",
    "\n",
    "        \n",
    "#         labels = [item._text for item in axsRight[length_dict*2-1].get_xticklabels()]\n",
    "#         #labels: ['176', '676', '1,176', '1,676', '2,176', '2,676', '3,176', '3,676', '4,176']\n",
    "#         new_labels = []\n",
    "#         for label in labels:\n",
    "#             #remove non-numberic character from string\n",
    "#             label = re.sub(\"[^0-9]\",\"\",label)                \n",
    "#             label = 1176-int(label)\n",
    "#             if label == 0:\n",
    "#                 label = f'{0} (TSS)'\n",
    "\n",
    "#             new_labels.append(label)\n",
    "#         axsRight[length_dict*2-1].set_xticklabels(new_labels)\n",
    "#         #set x axis to be a little closer to the gene\n",
    "#         axsRight[length_dict*2-1].spines['bottom'].set_position(('data',-0.5))\n",
    "#         #increase spacing between bottom plots\n",
    "\n",
    "#         # for k,v in dictionary_of_records_seqrecords.items():\n",
    "#         #     last = int(k[-1])*2+1\n",
    "#         #     last_chromatin = int(k[-1])*2\n",
    "#         #     #get promoter name\n",
    "#         #     prom_name = dictionary_of_records_promoternames[k]\n",
    "#         #     #split on dashes\n",
    "#         #     short_prom_name = prom_name.split('-')[0].upper()\n",
    "\n",
    "#         #chartBox=axsRight[length_dict*2-1].get_position()\n",
    "#         #axsRight[length_dict*2-2].set_position([chartBox.x0,(chartBox.y0-1),chartBox.width,chartBox.height])\n",
    "#         #axsRight[length_dict*2-2].set_position(['data',-0.2,'data','data'])\n",
    "\n",
    "#         fig.subplots_adjust(left=0, bottom=-0.1, right=1, top=0, wspace=-0.5, hspace=-0.1)\n",
    "\n",
    "\n",
    "#     else:\n",
    "#         #make plot\n",
    "#         height = length_dict//2\n",
    "#         fig = plt.figure(constrained_layout=False,figsize=(8,height))\n",
    "#         #make subfigures\n",
    "#         subfigs = fig.subfigures(1,2, wspace=0.0, width_ratios=[1,5])\n",
    "#         #left legend \n",
    "#         axsLeft=subfigs[0].subplots(1,1)\n",
    "#         #remove axis\n",
    "#         axsLeft.axis('off')\n",
    "#         #right figures\n",
    "#         axsRight = subfigs[1].subplots(length_dict,1,sharex=True)\n",
    "#     #move legend to the right\n",
    "#     # box = axsLeft.get_position()\n",
    "#     # box.x0 = box.x0 + 1.8\n",
    "#     # box.x1 = box.x1 + 1.8\n",
    "#     # box.y0 = box.y0 +0.13\n",
    "#     # box.y1 = box.y1 +0.13\n",
    "#     # axsLeft.set_position(box)\n",
    "    \n",
    "#         #make graphic records\n",
    "#         for k,v in dictionary_of_records_seqrecords.items():\n",
    "#             #take last character of string\n",
    "#             last = int(k[-1])\n",
    "#             #get promoter name\n",
    "#             prom_name = dictionary_of_records_promoternames[k]\n",
    "#             #split on dashes\n",
    "#             short_prom_name = prom_name.split('-')[0].upper()\n",
    "#             make_graphic_record(v,short_prom_name,axsRight[last-1])\n",
    "\n",
    "#     #add legend\n",
    "#     #first import colour iterator\n",
    "#     #import colour blind palette\n",
    "#     #colour palette from mkweb.bcgsc.ca/colorblind/palettes.mhtml\n",
    "#    # df = pd.read_csv(\"colourblind_palette.csv\", header=0)\n",
    "#     #convert to floats\n",
    "#    # floats=df.divide(255)\n",
    "#     #make sets of each row to get the red, green blue colours\n",
    "#    # CB_colour_palette = floats.apply(tuple, axis = 1)\n",
    "#     #turn into a list of colour sets\n",
    "#    # list_colours = list(CB_colour_palette)\n",
    "#     #make colour iterator\n",
    "#     #colour_iterator=cycle(list_colours)\n",
    "#     #Use feature_list generated just above the Class MyCustomTranslator(BiopythonTranslator)\n",
    "#     #iterate creating legend from feature list\n",
    "#     #create empty handles list\n",
    "#     #print(colour_dict)\n",
    "#     #sort TFBS names into alphabetical order in colour_dict\n",
    "#         colour_dict_sorted = {k: v for k,v in sorted(colour_dict.items(), key=lambda item: item[0])}\n",
    "#         #print(colour_dict_sorted)\n",
    "#         handles = []\n",
    "#         for TFBS,colour in colour_dict_sorted.items():\n",
    "#             addition = mpatches.Patch(color=colour, label=TFBS)\n",
    "#             #append to handles list\n",
    "#             handles.append(addition)\n",
    "            \n",
    "#         #use latex style rendering to allow parts to be bold\n",
    "#         rc('text', usetex=True)\n",
    "    \n",
    "\n",
    "#         #handles = sorted(handles)\n",
    "        \n",
    "#         #create legend and align left\n",
    "#         axsLeft.legend(handles=handles, loc='upper right', title=r\"\\textbf{Transcription factor binding site}\", title_fontsize='14',ncol=2)\n",
    "      \n",
    "\n",
    "#         #turn off latex rendering of text\n",
    "#         rc('text', usetex=False)\n",
    "\n",
    "\n",
    "#         #change x_lim to flip x axis\n",
    "#         for ax in axsRight:\n",
    "#             ax.set_xlim(4550,0)\n",
    "#             #change font colour of x axis text\n",
    "#             ax.tick_params(axis='x', colors='black')\n",
    "#             #change width of line\n",
    "#             # for axis in ['bottom','right']:\n",
    "#             #     ax.spines[axis].set_linewidth(5)\n",
    "#             if ax == axsRight[length_dict-1]:\n",
    "#                 pass\n",
    "#             else:\n",
    "#                 #remove xticks\n",
    "#                 ax.xaxis.set_ticks_position('none')\n",
    "#                 #remove axes\n",
    "#                 #ax.axis('off')\n",
    "        \n",
    "            \n",
    "\n",
    "\n",
    "#         x_ticks = np.arange(176, 4550, 500) #start stop and step\n",
    "#         plt.xticks(x_ticks)\n",
    "#         # for ax in axsRight:\n",
    "#         #     if ax == axsRight[length_dict-1]:\n",
    "#         #             pass\n",
    "#         #         else:\n",
    "#         #             #remove xticks\n",
    "#         #             ax.set_xticks([])\n",
    "#         #Now change labels of xticks centered around the TSS    \n",
    "#         fig.canvas.draw()\n",
    "#         labels = [item._text for item in axsRight[length_dict-1].get_xticklabels()]\n",
    "#         #labels: ['176', '676', '1,176', '1,676', '2,176', '2,676', '3,176', '3,676', '4,176']\n",
    "#         new_labels = []\n",
    "#         for label in labels:\n",
    "#             #remove non-numberic character from string\n",
    "#             label = re.sub(\"[^0-9]\",\"\",label)                \n",
    "#             label = 1176-int(label)\n",
    "#             if label == 0:\n",
    "#                 label = f'{0} (TSS)'\n",
    "\n",
    "#             new_labels.append(label)\n",
    "#         axsRight[length_dict-1].set_xticklabels(new_labels)\n",
    "#         fig.subplots_adjust(left=0.5, bottom=-0.1, right=0, top=1.2, wspace=0, hspace=0)\n",
    "    \n",
    "#     #print(int(labels[1])+2)\n",
    "#     print(colour_dict)\n",
    "#     #print(colour_iterator)\n",
    "\n",
    "#     #combine graphic records\n",
    "#     #all_records = [graphic_record1,graphic_record2]\n",
    "#     #print(feature_list)\n",
    "#     #create figure legend\n",
    "#     # lines_labels = [graphic_record.get_legend_handles_labels() for graphic_record in axsRight]\n",
    "#     # handles, labels = [sum(l, []) for l in zip(*lines_labels)]\n",
    "#     # fig.legend(handles, labels)\n",
    "\n",
    "#     #subfigs[1].suptitle('Promoters', fontsize='x-large')\n",
    "    \n",
    "#     # #make subplots\n",
    "#     # fig, (ax1,ax2) = plt.subplots(\n",
    "#     #     1,2, figsize=(12,4),gridspec_kw={'width_ratios':[1,4]}\n",
    "#     # )\n",
    "#     #plot Record Map\n",
    "    \n",
    "#     #add legend of last record\n",
    "#     #graphic_record1.plot_legend(ax=axsLeft, loc=1, ncol=1, frameon=True)\n",
    "#     #remove whitespace\n",
    "#     #plt.tight_layout()\n",
    "    \n",
    "#     #add a line at TSS\n",
    "#     #get xtick locations\n",
    "#     xtickslocs = axsRight[length_dict*2-1].get_position()\n",
    "#     xstart = xtickslocs.x0    \n",
    "#     xstop = xtickslocs.x1\n",
    "#     #get length of x axis using figure coordinates    \n",
    "#     xlength=xstop-xstart\n",
    "#     #get tss_location after converting from x axis position to figure position\n",
    "#     tss_location = ((4550-1176)/4550)*xlength+xstart\n",
    "#     #add vertical dashed line at TSS position\n",
    "#     line = Line2D((tss_location,tss_location),(.1,.9), color='black', linewidth=2, linestyle='--')\n",
    "#     fig.add_artist(line)\n",
    "\n",
    "\n",
    "#     #set DPI to 600\n",
    "#     fig.set_dpi(600)\n",
    "#     #fig.tight_layout()\n",
    "#     fig.savefig(f\"{dir_name}/combined.pdf\",bbox_inches='tight')\n",
    "#     fig.savefig(f\"{dir_name}/combined.svg\",bbox_inches='tight')\n",
    "#     fig.savefig(f\"{dir_name}/combined.tiff\",bbox_inches='tight')\n",
    "#     #convert image from RGB to CMYK\n",
    "#     RGB2CMYK(f'{dir_name}/combined', '.tiff')\n",
    "#     # #save figure into memory\n",
    "#     # svg1 = BytesIO()\n",
    "#     # fig.savefig(svg1)\n",
    "#     # #load image into PIL\n",
    "#     # svg2 = Image.open(svg1)\n",
    "#     # #save as TIFF\n",
    "#     # svg2.save(f\"../../data/TRAMP/plots/combined2.tiff\")\n",
    "#     # svg1.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 627,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_plot(dictionary_of_records_seqrecords,dictionary_of_records_promoternames,dir_name,openchromatin=True,dnaseI=False):\n",
    "    def make_graphic_record(seqrecord, promoter_name, ax,short_annotation=True,title=True):\n",
    "        \"\"\"make a graphic record object\"\"\"\n",
    "        #display figure\n",
    "        graphic_record = MyCustomTranslator().translate_record(seqrecord)\n",
    "        #graphic_record.labels_spacing = -5\n",
    "        #set minimum height of annotations\n",
    "        if short_annotation is True:\n",
    "            graphic_record.min_y_height_of_text_line = 0.5\n",
    "        else:\n",
    "            graphic_record.min_y_height_of_text_line = 0.5\n",
    "            # graphic_record.labels_spacing = 10\n",
    "\n",
    "        #set spacing between labels\n",
    "        #graphic_record.labels_spacing = -5\n",
    "        #change height to 0 so TFBS can overlap the promoter\n",
    "        graphic_record.feature_level_height = 0\n",
    "        #graphic_record = BiopythonTranslator().translate_record(gb_file)\n",
    "        graphic_record.plot(ax=ax, with_ruler=True,annotate_inline=True,)#,figure_width=10, #strand_in_label_threshold=4,annotate_inline=True so that labels are within the feature\n",
    "        #add title of promoter\n",
    "        if title is True:\n",
    "            ax.title.set_text(promoter_name)\n",
    "        #return graphic_record\n",
    "    #set plot parameters\n",
    "    rcParams['xtick.major.width'] = 2\n",
    "    rcParams['ytick.major.width'] = 2\n",
    "    rcParams['font.family'] = 'sans-serif'\n",
    "    rcParams['font.sans-serif'] = ['Arial']\n",
    "    #allow font to be edited later in pdf editor\n",
    "    rcParams ['pdf.fonttype'] = 42\n",
    "\n",
    "    #make svg text editable\n",
    "    rcParams['svg.fonttype'] = 'none'\n",
    "    \n",
    "    #rcParams['axes.linewidth'] = 2\n",
    "    #rcParams['lines.linewidth'] = 2\n",
    "    #remove top and right lines\n",
    "    # rcParams['axes.spines.top'] = False\n",
    "    # rcParams['axes.spines.right'] = False\n",
    "    #font size\n",
    "    rcParams['font.size'] = 11\n",
    "\n",
    "\n",
    "\n",
    "    #length of dictionary\n",
    "    length_dict = len(dictionary_of_records_promoternames)\n",
    "    #print(length_dict)\n",
    "    #\n",
    "    #if including openchromatin (atacseq by default) track, include more subfigures\n",
    "    if openchromatin is True:\n",
    "        #make plot\n",
    "        height = length_dict//2+length_dict-1 #add length_dict-1 to include empty grids for spacing between pairs\n",
    "        ### NEED TO MAKE INTO A GRID LIKE HERE: https://stackoverflow.com/questions/51717199/how-to-adjust-space-between-every-second-row-of-subplots-in-matplotlib\n",
    "\n",
    "        \n",
    "\n",
    "        if length_dict < 4:\n",
    "            fig = plt.figure(constrained_layout=False,figsize=(12,height+1))\n",
    "        else:\n",
    "            fig = plt.figure(constrained_layout=False,figsize=(11,height-5))\n",
    "        \n",
    "        #make subfigures\n",
    "        #subfigs = fig.subfigures(1,2, wspace=0.0, width_ratios=[1,5])\n",
    "        #left legend \n",
    "        #axsLeft=subfigs[0].subplots(1,1)\n",
    "        #remove axis\n",
    "        #axsLeft.axis('off')\n",
    "        # #right figures\n",
    "        #create gridspec so that open chromatin is paired with promoter\n",
    "        n = length_dict # number of double-rows\n",
    "        m = 2 # number of columns\n",
    "\n",
    "        t = 0.9 # 1-t == top space \n",
    "        b = 0.1 # bottom space      (both in figure coordinates)\n",
    "        if length_dict < 4:\n",
    "            msp = -0.7\n",
    "            sp = 0.3\n",
    "\n",
    "        else:\n",
    "            msp = -0.3 # minor spacing\n",
    "            sp = 0.3  # major spacing\n",
    "        offs=(1+msp)*(t-b)/(2*n+n*msp+(n-1)*sp) # grid offset\n",
    "        hspace = sp+msp+1 #height space per grid\n",
    "\n",
    "        gso = GridSpec(n,m, bottom=b+offs, top=t, hspace=hspace,width_ratios=[2,10])\n",
    "        gse = GridSpec(n,m, bottom=b, top=t-offs, hspace=hspace,width_ratios=[2,10])\n",
    "\n",
    "        #fig = plt.figure()\n",
    "\n",
    "        grid = []\n",
    "        for i in range(n*m):\n",
    "            grid.append(fig.add_subplot(gso[i]))\n",
    "            grid.append(fig.add_subplot(gse[i]))\n",
    "        \n",
    "\n",
    "        \n",
    "        #print(grid)\n",
    "        #print(len(grid))\n",
    "        #split plots into two lists - one for left column and one for the right\n",
    "        #turn off axes for the plots on the left\n",
    "        axsRight = []\n",
    "        axsLeft = []\n",
    "        count=0\n",
    "        for number in np.arange(len(grid)):\n",
    "            if count == 0:\n",
    "                axsLeft += [grid[number]]\n",
    "                grid[number].axis('off')\n",
    "                count +=1\n",
    "            elif count == 1:\n",
    "                axsLeft += [grid[number]]\n",
    "                count +=1\n",
    "                grid[number].axis('off')\n",
    "            elif count == 2:\n",
    "                axsRight += [grid[number]]\n",
    "                count +=1\n",
    "            elif count ==3:\n",
    "                axsRight += [grid[number]]\n",
    "                count = 0\n",
    "\n",
    "        #make legend span two plots \n",
    "        axsLeft[0].set_position(gso[0:3].get_position(fig))\n",
    "        #axsLeft[1].set_position(gso[0:3].get_position(fig)\n",
    "   \n",
    "                \n",
    "      \n",
    "                \n",
    "        #move plots closeer to each other\n",
    "        # for i in range(0,len(axsRight),2):\n",
    "        #     axsRight[i] = plt.subplot()\n",
    "        #print(axsRight)\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "        # num_rows = length_dict*2\n",
    "        # num_cols = 1\n",
    "        # row_height = 6\n",
    "        # space_height = 2\n",
    "        # num_sep_rows = lambda x: int((x-1)/2)\n",
    "        # grid = (row_height*num_rows + space_height*num_sep_rows(num_rows), num_cols)\n",
    "        # #axsRight = subfigs[1].subplots(num_rows,1,sharex=False)\n",
    "        # #axsRight = subfigs[1]\n",
    "        # axsRight = []\n",
    "\n",
    "        # for ind_row in range(num_rows):\n",
    "        #     for ind_col in range(num_cols):\n",
    "        #         grid_row = row_height*ind_row + space_height*num_sep_rows(ind_row+1)\n",
    "        #         grid_col = ind_col\n",
    "\n",
    "        # axsRight += [plt.subplot2grid(grid, (grid_row, grid_col), rowspan=row_height)]\n",
    "\n",
    "        if dnaseI==False:\n",
    "            def pygenometracks(chrom_region,start_region,end_region,axis):\n",
    "                \"\"\"make pygenome track for bigwig files\"\"\"        \n",
    "                file_location_roots=\"../../data/ATAC-seq/potter2018/GSE116287_Roots_NaOH_Merged.bw\"\n",
    "                file_location_shoots=\"../../data/ATAC-seq/potter2018/GSE116287_Shoots_NaOH_Merged.bw\"            \n",
    "                track_config_roots = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',alpha = 0.25,min_value=0, max_value=30)\n",
    "                track_config_shoots = dict(file=file_location_shoots,overlay_previous = 'share-y',color='teal',alpha = 0.25,min_value=0,max_value=30)\n",
    "                \n",
    "                #add lines too\n",
    "                track_config_roots_lines = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',type=\"line:1\",min_value=0, max_value=30)\n",
    "                track_config_shoots_lines = dict(file=file_location_shoots,overlay_previous = 'share-y',color='teal',type=\"line:1\",min_value=0, max_value=30)\n",
    "                #alpha is transparency\n",
    "                tk_roots = pygtk.BigWigTrack(track_config_roots)\n",
    "                tk_shoots = pygtk.BigWigTrack(track_config_shoots)\n",
    "                tk_roots_lines = pygtk.BigWigTrack(track_config_roots_lines)\n",
    "                tk_shoots_lines = pygtk.BigWigTrack(track_config_shoots_lines)\n",
    "                tk_roots.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_shoots.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_roots_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_shoots_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "            #make graphic records of genes and TFBSs\n",
    "            for k,v in dictionary_of_records_seqrecords.items():\n",
    "                #take last character of string, double it as twice as many sequence tracks to include open chromatin atacseq data\n",
    "                last = int(k[-1])*2+1\n",
    "                last_chromatin = int(k[-1])*2\n",
    "                #get promoter name\n",
    "                prom_name = dictionary_of_records_promoternames[k]\n",
    "                #split on dashes\n",
    "                short_prom_name = prom_name.split('-')[0].upper()\n",
    "                #print(short_prom_name)\n",
    "                if short_prom_name == \"ANAC032\":\n",
    "                    #print(ANAC032)\n",
    "                    make_graphic_record(v,short_prom_name,axsRight[last-2],short_annotation=False, title=False)\n",
    "                else:\n",
    "                    make_graphic_record(v,short_prom_name,axsRight[last-2],title=False)\n",
    "            \n",
    "                #print(v.id)\n",
    "                #add atacseq track\n",
    "                #first get chromosome number from sequence ID\n",
    "                chrom_region = re.findall(r\"chromosome:TAIR10:(\\d)\",v.id)[0]\n",
    "                #chrom_region = v.id[v.id.find('TAIR10:')+1:v.id.find(':')]\n",
    "                #print(v)\n",
    "                #print(chrom_region)\n",
    "                #chrom_region = AGI[2]\n",
    "                #then get start and stop region\n",
    "                #open genbank file, read third line containing keywords\n",
    "                gb_file=f\"../../data/TRAMP/{prom_name}_shortened.gb\"\n",
    "                with open(gb_file, \"r\") as f:\n",
    "                    all_lines = f.readlines()\n",
    "                    for line in all_lines:\n",
    "                        if re.match(r'KEYWORDS', line):\n",
    "                            keywords = line                \n",
    "                start_region = int(re.findall(r\"start_index:(\\d+)\",keywords)[0])\n",
    "                end_region = int(re.findall(r\"end_index:(\\d+)\",keywords)[0])\n",
    "                #print(start_region)  \n",
    "                pygenometracks(chrom_region,start_region,end_region,axsRight[last_chromatin-2])\n",
    "                #set xlim\n",
    "                offset_length = 4550-(end_region-start_region)\n",
    "                axsRight[last_chromatin-2].set_xlim(start_region-offset_length,end_region)\n",
    "                #get x and y lim\n",
    "                #axsRight[last_chromatin-2].set_title(short_prom_name,y=0.5, ) #put title to left of axis\n",
    "                ##setlocation of the title         \n",
    "                #first transform offset_length to between 0 and 1 for axes location, and offset a little to the left\n",
    "                trans = (offset_length-25)/4550\n",
    "\n",
    "                \n",
    "                axsRight[last_chromatin-2].text(x=trans, y=0.125, s=short_prom_name, weight=\"extra bold\", fontsize=10,transform=axsRight[last-2].transAxes, ha='right')#transform=axsRight[last_chromatin-2].transAxes\n",
    "\n",
    "                #add black border round each gene\n",
    "                #add black border round each gene\n",
    "                #rect = plt.Rectangle((xstart,0.5),0.5,0.1,fill=False, color='black', linewidth=2)\n",
    "                xtickslocs = axsRight[length_dict*2-1].get_position()\n",
    "            #need to fix the addition of a rectangle here\n",
    "            xstart = xtickslocs.x0    \n",
    "            xstop = xtickslocs.x1\n",
    "            #get length of x axis using figure coordinates    \n",
    "            xlength=xstop-xstart\n",
    "            #get tss_location after converting from x axis position to figure position\n",
    "            tss_location = ((4550-1176)/4550)*xlength+xstart\n",
    "            start=(2000/4550)*xlength+xstart\n",
    "            width=start-xstop\n",
    "            #add vertical dashed line at TSS position\n",
    "            line = Line2D((tss_location,tss_location),(width,.9), color='black', linewidth=2, linestyle='--')\n",
    "            fig.add_artist(line)\n",
    "        \n",
    "\n",
    "                rect = plt.Rectangle(\n",
    "                    # (lower left corner coordinates), width, height                    \n",
    "                    (offset_length/4550,0),(end_region-start_region),0.05,fill=False, color='black', linewidth=2)\n",
    "                fig.add_artist(rect)\n",
    "                # rect=axsRight[last_chromatin-2].add_patch(rect)\n",
    "                # rect.set_clip_on(False)\n",
    "               # rect = plt.Rectangle(\n",
    "                    # (lower left corner coordinates), width, height                    \n",
    "             #       (offset_length,0),(end_region-start_region),0.05,fill=False, color='black', linewidth=2)\n",
    "                \n",
    "\n",
    "                \n",
    "                #axsRight[last_chromatin-2].invert_xaxis()\n",
    "                #change xvalues of open chromatin track      \n",
    "                #first get x and y values\n",
    "                #line = axsRight[last_chromatin-2].get_lines()\n",
    "                #xd = line.get_xdata()\n",
    "                #yd = line.get_ydata()\n",
    "                #print(line)\n",
    "                #ax1Ticks = axsRight[last_chromatin-2].get_xticks()   \n",
    "            # ax2Ticks = ax1Ticks.copy()\n",
    "                # def tick_function(X,start_region):\n",
    "                #     V = X-start_region\n",
    "                #     return [\"%.3f\" % z for z in V]\n",
    "\n",
    "                #\n",
    "                #axs[last-1].set_xticks(ax2Ticks)\n",
    "                #ax2.set_xbound(ax1.get_xbound())\n",
    "                #axsRight[last_chromatin-2].set_xticklabels(tick_function(ax2Ticks,start_region))\n",
    "                #axsRight[last_chromatin-2].set_xscale('function', functions=(forward_function,inverse_function))\n",
    "                #make x labels integars\n",
    "        elif dnaseI==True:\n",
    "            def pygenometracks(chrom_region,start_region,end_region,axis):\n",
    "                \"\"\"make pygenome track for bigwig files\"\"\"        \n",
    "                file_location_roots=\"../../data/DNase-seq/sullivan2014/sullivan2014_root_merged.bw\"\n",
    "                file_location_seedling=\"../../data/DNase-seq/sullivan2014/sullivan2014_seedling_merged.bw\"\n",
    "                track_config_seedling = dict(file=file_location_seedling,overlay_previous = 'share-y',color='teal',alpha = 0.25,min_value=0,max_value=2000,height=0.05)       \n",
    "                track_config_roots = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',alpha = 0.25,min_value=0, max_value=2000,height=0.05)\n",
    "                \n",
    "                \n",
    "                #add lines too\n",
    "                track_config_seedling_lines = dict(file=file_location_seedling,overlay_previous = 'share-y',color='teal',type=\"line:1\",min_value=0, max_value=2000,height=0.05)\n",
    "                track_config_roots_lines = dict(file=file_location_roots,overlay_previous = 'share-y',color='brown',type=\"line:1\",min_value=0, max_value=2000,height=0.05)\n",
    "                \n",
    "                #alpha is transparency\n",
    "                tk_seedling = pygtk.BigWigTrack(track_config_seedling)\n",
    "                tk_roots = pygtk.BigWigTrack(track_config_roots)\n",
    "                tk_seedling_lines = pygtk.BigWigTrack(track_config_seedling_lines)\n",
    "                tk_roots_lines = pygtk.BigWigTrack(track_config_roots_lines)\n",
    "                tk_seedling.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_roots.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_seedling_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "                tk_roots_lines.plot(axis,chrom_region,start_region,end_region,)\n",
    "                \n",
    "            #make graphic records of genes and TFBSs\n",
    "            for k,v in dictionary_of_records_seqrecords.items():\n",
    "                #take last character of string, double it as twice as many sequence tracks to include open chromatin atacseq data\n",
    "                last = int(k[-1])*2+1\n",
    "                last_chromatin = int(k[-1])*2\n",
    "                #get promoter name\n",
    "                prom_name = dictionary_of_records_promoternames[k]\n",
    "                #split on dashes\n",
    "                short_prom_name = prom_name.split('-')[0].upper()\n",
    "                #print(short_prom_name)\n",
    "                if short_prom_name == \"ANAC032\":\n",
    "                    #print(ANAC032)\n",
    "                    make_graphic_record(v,short_prom_name,axsRight[last-2],short_annotation=False, title=False)\n",
    "                else:\n",
    "                    make_graphic_record(v,short_prom_name,axsRight[last-2],title=False)\n",
    "            \n",
    "                #print(v.id)\n",
    "                #add atacseq track\n",
    "                #first get chromosome number from sequence ID\n",
    "                chrom_region = re.findall(r\"chromosome:TAIR10:(\\d)\",v.id)[0]\n",
    "                #chrom_region = v.id[v.id.find('TAIR10:')+1:v.id.find(':')]\n",
    "                #print(v)\n",
    "                #print(chrom_region)\n",
    "                #chrom_region = AGI[2]\n",
    "                #then get start and stop region\n",
    "                #open genbank file, read third line containing keywords\n",
    "                gb_file=f\"../../data/TRAMP/{prom_name}_shortened.gb\"\n",
    "                with open(gb_file, \"r\") as f:\n",
    "                    all_lines = f.readlines()\n",
    "                    for line in all_lines:\n",
    "                        if re.match(r'KEYWORDS', line):\n",
    "                            keywords = line                \n",
    "                start_region = int(re.findall(r\"start_index:(\\d+)\",keywords)[0])\n",
    "                end_region = int(re.findall(r\"end_index:(\\d+)\",keywords)[0])\n",
    "                #print(start_region)  \n",
    "                pygenometracks(chrom_region,start_region,end_region,axsRight[last_chromatin-2])\n",
    "                #set xlim\n",
    "                offset_length = 4550-(end_region-start_region)\n",
    "                axsRight[last_chromatin-2].set_xlim(start_region-offset_length,end_region)\n",
    "                #get x and y lim\n",
    "                #axsRight[last_chromatin-2].set_title(short_prom_name,y=0.5, ) #put title to left of axis\n",
    "                ##setlocation of the title         \n",
    "                #first transform offset_length to between 0 and 1 for axes location, and offset a little to the left\n",
    "                trans = (offset_length-25)/4550\n",
    "\n",
    "                \n",
    "                axsRight[last_chromatin-2].text(x=trans, y=0.125, s=short_prom_name, weight=\"extra bold\", fontsize=10,transform=axsRight[last-2].transAxes, ha='right')#transform=axsRight[last_chromatin-2].transAxes\n",
    "\n",
    "                \n",
    "\n",
    "\n",
    "        colour_dict_sorted = {k: v for k,v in sorted(colour_dict.items(), key=lambda item: item[0])}\n",
    "        #print(colour_dict_sorted)\n",
    "        handles = []\n",
    "        labels = []\n",
    "        for TFBS,colour in colour_dict_sorted.items():\n",
    "            addition = mpatches.Patch(color=colour)\n",
    "            #append to handles list\n",
    "            handles.append(addition)\n",
    "            labels.append(TFBS)\n",
    "\n",
    "        #use latex style rendering to allow parts to be bold\n",
    "        \n",
    "        rc('text', usetex=True)\n",
    "\n",
    "        #handles = sorted(handles)\n",
    "\n",
    "        #create TFBS legend and align left\n",
    "        #append open chromatin to handles\n",
    "        #first create custom handler for string in legend to add a second title for open chromatin\n",
    "        #used https://gist.github.com/Raudcu/44b43c7f3f893fe2f4dd900afb740b7f\n",
    "        class LegendTitle(object):\n",
    "            def __init__(self, text_props=None):\n",
    "                self.text_props = text_props or {}\n",
    "                super(LegendTitle, self).__init__()\n",
    "\n",
    "            def legend_artist(self, legend, orig_handle, fontsize, handlebox):\n",
    "                x0, y0 = handlebox.xdescent, handlebox.ydescent\n",
    "                title = mtext.Text(x0, y0, r\"\\textbf{{{}}}\".format(orig_handle), **self.text_props)\n",
    "                handlebox.add_artist(title)\n",
    "                return title\n",
    "#r\"\\textbf{'+orig_handle+'}\"\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "        #label titles\n",
    "        title1 = 'Candidate transcription factor binding sites' \n",
    "        title2 = 'Gene features'\n",
    "        title3 = 'Open chromatin peaks'\n",
    "        #add colour patches\n",
    "        intron = mpatches.Patch(color='lightgrey')\n",
    "        exon = mpatches.Patch(color='#635147')\n",
    "        upstream_mrna = mpatches.Patch(color='#DDCC77')\n",
    "        if dnaseI==False:\n",
    "            root = Line2D([0],[0],color='brown', lw=2)\n",
    "            shoot = Line2D([0],[0],color='teal',lw=2)\n",
    "\n",
    "            #insert handles\n",
    "            handles.insert(0,title1)\n",
    "            labels.insert(0,'')\n",
    "            handles.insert(3,title2)\n",
    "            labels.insert(3,'')\n",
    "            handles.insert(4,exon)\n",
    "            labels.insert(4,'Exon')\n",
    "            handles.insert(5,title3)\n",
    "            labels.insert(5,'')\n",
    "            handles.insert(6,root)\n",
    "            labels.insert(6,'Root')\n",
    "            #blank label to make layout look nice\n",
    "            handles.insert(7,'')\n",
    "            labels.insert(7,'')\n",
    "            handles.insert(10,'')\n",
    "            labels.insert(10,'')\n",
    "            handles.insert(11,intron)\n",
    "            labels.insert(11,'Intron')\n",
    "            #blank label to make layout look nice \n",
    "            handles.insert(12,'')\n",
    "            labels.insert(12,'')\n",
    "            handles.insert(13,shoot)\n",
    "            labels.insert(13,'Shoot')\n",
    "            #blank label to make layout look nice\n",
    "            handles.insert(14,'')\n",
    "            labels.insert(14,'')\n",
    "            handles.insert(17,'')\n",
    "            labels.insert(17,'')\n",
    "\n",
    "            handles.insert(18,upstream_mrna)\n",
    "            labels.insert(18,'Upstream transcript')\n",
    "            handles.insert(19,'')\n",
    "            labels.insert(19,'')\n",
    "        if dnaseI==True:\n",
    "            root = Line2D([0],[0],color='brown', lw=2)\n",
    "            shoot = Line2D([0],[0],color='teal',lw=2)\n",
    "\n",
    "            #insert handles\n",
    "            handles.insert(0,title1)\n",
    "            labels.insert(0,'')\n",
    "            handles.insert(3,title2)\n",
    "            labels.insert(3,'')\n",
    "            handles.insert(4,exon)\n",
    "            labels.insert(4,'Exon')\n",
    "            handles.insert(5,title3)\n",
    "            labels.insert(5,'')\n",
    "            handles.insert(6,root)\n",
    "            labels.insert(6,'Root')\n",
    "            #blank label to make layout look nice\n",
    "            handles.insert(7,'')\n",
    "            labels.insert(7,'')\n",
    "            handles.insert(10,'')\n",
    "            labels.insert(10,'')\n",
    "            handles.insert(11,intron)\n",
    "            labels.insert(11,'Intron')\n",
    "            #blank label to make layout look nice \n",
    "            handles.insert(12,'')\n",
    "            labels.insert(12,'')\n",
    "            handles.insert(13,shoot)\n",
    "            labels.insert(13,'Seedling')\n",
    "            #blank label to make layout look nice\n",
    "            handles.insert(14,'')\n",
    "            labels.insert(14,'')\n",
    "            handles.insert(17,'')\n",
    "            labels.insert(17,'')\n",
    "\n",
    "            handles.insert(18,upstream_mrna)\n",
    "            labels.insert(18,'Upstream transcript')\n",
    "            handles.insert(19,'')\n",
    "            labels.insert(19,'')\n",
    "\n",
    "\n",
    "      \n",
    "\n",
    "        \n",
    "        axsLeft[0].legend(handles=handles,labels=labels, loc='upper right',ncol=3,handler_map={str: LegendTitle({'fontsize': 14})})#handler_map={title: LegendTitle({'fontsize': 14})}\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        #axsLeft[0].legend(handles=handles,labels=labels, loc='upper right', title=r\"\\textbf{Transcription factor binding site}\", title_fontsize='14',ncol=2,)#handler_map={title: LegendTitle({'fontsize': 14})}\n",
    "        # # Add the legend manually to the current Axes.\n",
    "\n",
    "        #create open chromatin legend below\n",
    "        #open_chrom_handles = [mpatches.Patch(facecolor='brown', edgecolor='brown', label='Root'),mpatches.Patch(facecolor='green', edgecolor='green', label='Shoot')]\n",
    "        #axsLeft[4].legend(handles=open_chrom_handles, loc='lower right', title=r\"\\textbf{Open chromatin peaks}\", title_fontsize='14',ncol=2)\n",
    "\n",
    "        # #turn off latex rendering of text\n",
    "        rc('text', usetex=False)\n",
    "       \n",
    "        #change x_lim to flip x axis\n",
    "        for n in np.arange(length_dict):\n",
    "            last = n*2+1\n",
    "            \n",
    "            axsRight[last-2].set_xlim(4550,0)\n",
    "            \n",
    "            #change font colour of x axis text\n",
    "            axsRight[last-2].tick_params(axis='x', colors='black')\n",
    "            #change width of line\n",
    "            # for axis in ['bottom','right']:\n",
    "            #     ax.spines[axis].set_linewidth(5)\n",
    "\n",
    "            x_ticks = np.arange(176, 4550, 500) #start stop and step\n",
    "            axsRight[last-2].set_xticks(x_ticks)\n",
    "            fig.canvas.draw()\n",
    "            # if axsRight[last-2] == axsRight[length_dict*2-1]:\n",
    "            #     pass\n",
    "            # else:\n",
    "            #     #remove xticks\n",
    "            #     #axsRight[last-2].xaxis.set_ticks_position('none')\n",
    "            #     #remove axes\n",
    "            #     ax.axis('off')\n",
    "                \n",
    "        #remove all axes\n",
    "        for ax in axsRight:            \n",
    "            if ax == axsRight[length_dict*2-1]:\n",
    "                pass\n",
    "            else:\n",
    "                #ax.xaxis.set_ticks_position('none')\n",
    "                ax.axis('off')\n",
    "\n",
    "        \n",
    "        labels = [item._text for item in axsRight[length_dict*2-1].get_xticklabels()]\n",
    "        #labels: ['176', '676', '1,176', '1,676', '2,176', '2,676', '3,176', '3,676', '4,176']\n",
    "        new_labels = []\n",
    "        for label in labels:\n",
    "            #remove non-numberic character from string\n",
    "            label = re.sub(\"[^0-9]\",\"\",label)                \n",
    "            label = 1176-int(label)\n",
    "            if label == 0:\n",
    "                label = f'{0} (TSS)'\n",
    "\n",
    "            new_labels.append(label)\n",
    "        axsRight[length_dict*2-1].set_xticklabels(new_labels)\n",
    "        #set x axis to be a little closer to the gene\n",
    "        axsRight[length_dict*2-1].spines['bottom'].set_position(('data',-0.5))\n",
    "        \n",
    "        #increase spacing between bottom plots\n",
    "\n",
    "        # for k,v in dictionary_of_records_seqrecords.items():\n",
    "        #     last = int(k[-1])*2+1\n",
    "        #     last_chromatin = int(k[-1])*2\n",
    "        #     #get promoter name\n",
    "        #     prom_name = dictionary_of_records_promoternames[k]\n",
    "        #     #split on dashes\n",
    "        #     short_prom_name = prom_name.split('-')[0].upper()\n",
    "\n",
    "        #chartBox=axsRight[length_dict*2-1].get_position()\n",
    "        #axsRight[length_dict*2-2].set_position([chartBox.x0,(chartBox.y0-1),chartBox.width,chartBox.height])\n",
    "        #axsRight[length_dict*2-2].set_position(['data',-0.2,'data','data'])\n",
    "\n",
    "        fig.subplots_adjust(left=0, bottom=-0.1, right=1, top=0, wspace=-0.5, hspace=-0.1)\n",
    "\n",
    "\n",
    "    else:\n",
    "        #make plot\n",
    "        height = length_dict//2\n",
    "        fig = plt.figure(constrained_layout=False,figsize=(8,height))\n",
    "        #make subfigures\n",
    "        subfigs = fig.subfigures(1,2, wspace=0.0, width_ratios=[1,5])\n",
    "        #left legend \n",
    "        axsLeft=subfigs[0].subplots(1,1)\n",
    "        #remove axis\n",
    "        axsLeft.axis('off')\n",
    "        #right figures\n",
    "        axsRight = subfigs[1].subplots(length_dict,1,sharex=True)\n",
    "    #move legend to the right\n",
    "    # box = axsLeft.get_position()\n",
    "    # box.x0 = box.x0 + 1.8\n",
    "    # box.x1 = box.x1 + 1.8\n",
    "    # box.y0 = box.y0 +0.13\n",
    "    # box.y1 = box.y1 +0.13\n",
    "    # axsLeft.set_position(box)\n",
    "    \n",
    "        #make graphic records\n",
    "        for k,v in dictionary_of_records_seqrecords.items():\n",
    "            #take last character of string\n",
    "            last = int(k[-1])\n",
    "            #get promoter name\n",
    "            prom_name = dictionary_of_records_promoternames[k]\n",
    "            #split on dashes\n",
    "            short_prom_name = prom_name.split('-')[0].upper()\n",
    "            make_graphic_record(v,short_prom_name,axsRight[last-1])\n",
    "\n",
    "    #add legend\n",
    "    #first import colour iterator\n",
    "    #import colour blind palette\n",
    "    #colour palette from mkweb.bcgsc.ca/colorblind/palettes.mhtml\n",
    "   # df = pd.read_csv(\"colourblind_palette.csv\", header=0)\n",
    "    #convert to floats\n",
    "   # floats=df.divide(255)\n",
    "    #make sets of each row to get the red, green blue colours\n",
    "   # CB_colour_palette = floats.apply(tuple, axis = 1)\n",
    "    #turn into a list of colour sets\n",
    "   # list_colours = list(CB_colour_palette)\n",
    "    #make colour iterator\n",
    "    #colour_iterator=cycle(list_colours)\n",
    "    #Use feature_list generated just above the Class MyCustomTranslator(BiopythonTranslator)\n",
    "    #iterate creating legend from feature list\n",
    "    #create empty handles list\n",
    "    #print(colour_dict)\n",
    "    #sort TFBS names into alphabetical order in colour_dict\n",
    "        colour_dict_sorted = {k: v for k,v in sorted(colour_dict.items(), key=lambda item: item[0])}\n",
    "        #print(colour_dict_sorted)\n",
    "        handles = []\n",
    "        for TFBS,colour in colour_dict_sorted.items():\n",
    "            addition = mpatches.Patch(color=colour, label=TFBS)\n",
    "            #append to handles list\n",
    "            handles.append(addition)\n",
    "            \n",
    "        #use latex style rendering to allow parts to be bold\n",
    "        rc('text', usetex=True)\n",
    "    \n",
    "\n",
    "        #handles = sorted(handles)\n",
    "        \n",
    "        #create legend and align left\n",
    "        axsLeft.legend(handles=handles, loc='upper right', title=r\"\\textbf{Transcription factor binding site}\", title_fontsize='14',ncol=2)\n",
    "      \n",
    "\n",
    "        #turn off latex rendering of text\n",
    "        rc('text', usetex=False)\n",
    "\n",
    "\n",
    "        #change x_lim to flip x axis\n",
    "        for ax in axsRight:\n",
    "            ax.set_xlim(4550,0)\n",
    "            #change font colour of x axis text\n",
    "            ax.tick_params(axis='x', colors='black')\n",
    "            #change width of line\n",
    "            # for axis in ['bottom','right']:\n",
    "            #     ax.spines[axis].set_linewidth(5)\n",
    "            if ax == axsRight[length_dict-1]:\n",
    "                pass\n",
    "            else:\n",
    "                #remove xticks\n",
    "                ax.xaxis.set_ticks_position('none')\n",
    "                #remove axes\n",
    "                #ax.axis('off')\n",
    "        \n",
    "            \n",
    "\n",
    "\n",
    "        x_ticks = np.arange(176, 4550, 500) #start stop and step\n",
    "        plt.xticks(x_ticks)\n",
    "        # for ax in axsRight:\n",
    "        #     if ax == axsRight[length_dict-1]:\n",
    "        #             pass\n",
    "        #         else:\n",
    "        #             #remove xticks\n",
    "        #             ax.set_xticks([])\n",
    "        #Now change labels of xticks centered around the TSS    \n",
    "        fig.canvas.draw()\n",
    "        labels = [item._text for item in axsRight[length_dict-1].get_xticklabels()]\n",
    "        #labels: ['176', '676', '1,176', '1,676', '2,176', '2,676', '3,176', '3,676', '4,176']\n",
    "        new_labels = []\n",
    "        for label in labels:\n",
    "            #remove non-numberic character from string\n",
    "            label = re.sub(\"[^0-9]\",\"\",label)                \n",
    "            label = 1176-int(label)\n",
    "            if label == 0:\n",
    "                label = f'{0} (TSS)'\n",
    "\n",
    "            new_labels.append(label)\n",
    "        axsRight[length_dict-1].set_xticklabels(new_labels)\n",
    "        fig.subplots_adjust(left=0.5, bottom=-0.1, right=0, top=1.2, wspace=0, hspace=0)\n",
    "    \n",
    "    #print(int(labels[1])+2)\n",
    "    print(colour_dict)\n",
    "    #print(colour_iterator)\n",
    "\n",
    "    #combine graphic records\n",
    "    #all_records = [graphic_record1,graphic_record2]\n",
    "    #print(feature_list)\n",
    "    #create figure legend\n",
    "    # lines_labels = [graphic_record.get_legend_handles_labels() for graphic_record in axsRight]\n",
    "    # handles, labels = [sum(l, []) for l in zip(*lines_labels)]\n",
    "    # fig.legend(handles, labels)\n",
    "\n",
    "    #subfigs[1].suptitle('Promoters', fontsize='x-large')\n",
    "    \n",
    "    # #make subplots\n",
    "    # fig, (ax1,ax2) = plt.subplots(\n",
    "    #     1,2, figsize=(12,4),gridspec_kw={'width_ratios':[1,4]}\n",
    "    # )\n",
    "    #plot Record Map\n",
    "    \n",
    "    #add legend of last record\n",
    "    #graphic_record1.plot_legend(ax=axsLeft, loc=1, ncol=1, frameon=True)\n",
    "    #remove whitespace\n",
    "    #plt.tight_layout()\n",
    "    \n",
    "    #add a line at TSS\n",
    "    #get xtick locations\n",
    "    xtickslocs = axsRight[length_dict*2-1].get_position()\n",
    "    xstart = xtickslocs.x0    \n",
    "    xstop = xtickslocs.x1\n",
    "    #get length of x axis using figure coordinates    \n",
    "    xlength=xstop-xstart\n",
    "    #get tss_location after converting from x axis position to figure position\n",
    "    tss_location = ((4550-1176)/4550)*xlength+xstart\n",
    "    start=(2000/4550)*xlength+xstart\n",
    "    width=start-xstop\n",
    "    #add vertical dashed line at TSS position\n",
    "    line = Line2D((tss_location,tss_location),(width,.9), color='black', linewidth=2, linestyle='--')\n",
    "    fig.add_artist(line)\n",
    "    #add black border round each gene\n",
    "    rect = plt.Rectangle((xstart,0.5),0.5,0.1,fill=False, color='black', linewidth=2)\n",
    "    fig.add_artist(rect)\n",
    "\n",
    "\n",
    "    #set DPI to 600\n",
    "    fig.set_dpi(600)\n",
    "    #fig.tight_layout()\n",
    "    if dnaseI==False:\n",
    "        fig.savefig(f\"{dir_name}/combined.pdf\",bbox_inches='tight')\n",
    "        fig.savefig(f\"{dir_name}/combined.svg\",bbox_inches='tight')\n",
    "        fig.savefig(f\"{dir_name}/combined.tiff\",bbox_inches='tight')\n",
    "    elif dnaseI==True:\n",
    "        fig.savefig(f\"{dir_name}/combined_dnaseI.pdf\",bbox_inches='tight')\n",
    "        fig.savefig(f\"{dir_name}/combined_dnaseI.svg\",bbox_inches='tight')\n",
    "        fig.savefig(f\"{dir_name}/combined_dnaseI.tiff\",bbox_inches='tight')\n",
    "    #convert image from RGB to CMYK\n",
    "    RGB2CMYK(f'{dir_name}/combined', '.tiff')\n",
    "    # #save figure into memory\n",
    "    # svg1 = BytesIO()\n",
    "    # fig.savefig(svg1)\n",
    "    # #load image into PIL\n",
    "    # svg2 = Image.open(svg1)\n",
    "    # #save as TIFF\n",
    "    # svg2.save(f\"../../data/TRAMP/plots/combined2.tiff\")\n",
    "    # svg1.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 628,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I need to add the following function in to fix annotation boxes being detached (see https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer/issues/42_)\n",
    "# This problem is a bit complicated and may need refactoring, but redefining the below function before plotting at least connects the boxes until a proper solution is implemented:\n",
    "\n",
    "# def new_determine_annotation_height(levels):\n",
    "#     return 1\n",
    "# record.determine_annotation_height = new_determine_annotation_height\n",
    "\n",
    "#Also I need to add an option to exclude TRAMP probe labels\n",
    "\n",
    "#Also look into DREB26 - the exon is not annotated in the plot. The 5'UTR is actually labelled as exon in the ensembl plant annotation - investigate this\n",
    "#In fact check all of the genes, some non-coding exons are mislablled as introns eg. in ARF18\n",
    "#Also need to write in script to cut the gb file sequence short so that all genes are the same length so align exactly on the TSS - done\n",
    "#fix TFBS in promoter plot - shorten a longer GB file. (Make a narrower version).  - done\n",
    "# #Make sure TFBS DREB26 and nlp7 colours are correct. Also allow the legend to be editable. Make the promoter name in bold. \n",
    "#use code from here: https://github.com/biopython/biopython/pull/1771/commits/c733feb84e7c48a54921f2c2ace2c57efdd97c8c to allow slicing while keeping truncated anontations\n",
    "#need to add lines round select features to improve the figure visually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 629,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(args):\n",
    "    # parse arguments\n",
    "    #args = parse_args(args)\n",
    "    #dependent_variable = \"GC_content\"\n",
    "\n",
    "    # make directory for the plots to be exported to\n",
    "    dirName = f\"../../data/TRAMP/plots\"\n",
    "    try:\n",
    "        # Create target Directory\n",
    "        os.mkdir(dirName)\n",
    "        print(\"Directory \", dirName, \" created\")\n",
    "    except FileExistsError:\n",
    "        print(\"Directory \", dirName, \" already exists\")\n",
    "    \n",
    "    promoter_names=dict([('promoter_name8',\"nlp7-at4g24020_ensembl_plant\"),\n",
    "    ('promoter_name7',\"nlp6-at1g64530_ensembl_plant\"),\n",
    "    ('promoter_name9',\"tga1-at5g65210_ensembl_plant\"),\n",
    "    ('promoter_name4',\"arf18-at3g61830_ensembl_plant\"),\n",
    "    ('promoter_name3',\"arf9-at4g23980_ensembl_plant\"),\n",
    "    ('promoter_name1',\"anac032-at1g77450_ensembl_plant\"),\n",
    "    ('promoter_name2',\"anr1-at2g14210_ensembl_plant\"),\n",
    "    ('promoter_name5',\"dreb26-at1g21910_ensembl_plant\"),\n",
    "    ('promoter_name6',\"nir1-at2g15620_ensembl_plant\")])\n",
    "    #sort promoter names dictionary in alphabetical order\n",
    "    promoter_names = {k: v for k,v in sorted(promoter_names.items(), key=lambda item: item[0])}\n",
    "    #load seqrecords\n",
    "    #create empty dictionary\n",
    "    seqrecords = {}\n",
    "    for k,v in promoter_names.items():             \n",
    "        #add to new dictionary of seqrecords\n",
    "        seqrecords[k] = gb_file_to_seqrecord(v)\n",
    "    \n",
    "    #make plot using dictionary\n",
    "    #make_plot(seqrecords,promoter_names)\n",
    "    #make_plot(seqrecords,promoter_names,dirName)\n",
    "    #make another with the union atacseq bed file from Alex Mason\n",
    "    make_plot(seqrecords,promoter_names,dirName,openchromatin=True,dnaseI=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 630,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #figure for SAB 2021 poster\n",
    "# def main(args):\n",
    "#     # parse arguments\n",
    "#     #args = parse_args(args)\n",
    "#     #dependent_variable = \"GC_content\"\n",
    "\n",
    "#     # make directory for the plots to be exported to\n",
    "#     dirName = f\"../../data/TRAMP/plots/SAB2021\"\n",
    "#     try:\n",
    "#         # Create target Directory\n",
    "#         os.mkdir(dirName)\n",
    "#         print(\"Directory \", dirName, \" created\")\n",
    "#     except FileExistsError:\n",
    "#         print(\"Directory \", dirName, \" already exists\")\n",
    "    \n",
    "#     promoter_names=dict([('promoter_name2',\"nlp7-at4g24020_ensembl_plant\"),\n",
    "#     #('promoter_name7',\"nlp6-at1g64530_ensembl_plant\"),\n",
    "#     #('promoter_name9',\"tga1-at5g65210_ensembl_plant\"),\n",
    "#     #('promoter_name4',\"arf18-at3g61830_ensembl_plant\"),\n",
    "#     #('promoter_name3',\"arf9-at4g23980_ensembl_plant\"),\n",
    "#     ('promoter_name1',\"anac032-at1g77450_ensembl_plant\")])\n",
    "#     #('promoter_name2',\"anr1-at2g14210_ensembl_plant\"),\n",
    "#     #('promoter_name5',\"dreb26-at1g21910_ensembl_plant\"),\n",
    "# #('promoter_name6',\"nir1-at2g15620_ensembl_plant\")])\n",
    "#     #sort promoter names dictionary in alphabetical order\n",
    "#     promoter_names = {k: v for k,v in sorted(promoter_names.items(), key=lambda item: item[0])}\n",
    "#     #load seqrecords\n",
    "#     #create empty dictionary\n",
    "#     seqrecords = {}\n",
    "#     for k,v in promoter_names.items():             \n",
    "#         #add to new dictionary of seqrecords\n",
    "#         seqrecords[k] = gb_file_to_seqrecord(v)\n",
    "    \n",
    "#     #make plot using dictionary\n",
    "#     #make_plot(seqrecords,promoter_names)\n",
    "#     make_plot(seqrecords,promoter_names,dirName)\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 631,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/TRAMP/plots  already exists\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/witham/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:213: BiopythonParserWarning: Over indented TRAMP_probe_tested feature?\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ARF9/18*': '#2271B2', 'ANR1*': '#3DB7E9', 'TGA1': '#F748A5', 'ANAC032*': '#d55e00', 'NLP6/7': '#e69f00', 'DREB26': '#228833'}\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Image size of 29681771x6515 pixels is too large. It must be less than 2^16 in each direction.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_18975/1069595384.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mimport\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/tmp/ipykernel_18975/2229301685.py\u001b[0m in \u001b[0;36mmain\u001b[0;34m(args)\u001b[0m\n\u001b[1;32m     35\u001b[0m     \u001b[0;31m#make_plot(seqrecords,promoter_names,dirName)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;31m#make another with the union atacseq bed file from Alex Mason\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m     \u001b[0mmake_plot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseqrecords\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpromoter_names\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdirName\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mopenchromatin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdnaseI\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_18975/923533636.py\u001b[0m in \u001b[0;36mmake_plot\u001b[0;34m(dictionary_of_records_seqrecords, dictionary_of_records_promoternames, dir_name, openchromatin, dnaseI)\u001b[0m\n\u001b[1;32m    691\u001b[0m         \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dir_name}/combined.pdf\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbbox_inches\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'tight'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    692\u001b[0m         \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dir_name}/combined.svg\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbbox_inches\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'tight'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 693\u001b[0;31m         \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dir_name}/combined.tiff\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbbox_inches\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'tight'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    694\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mdnaseI\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    695\u001b[0m         \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msavefig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dir_name}/combined_dnaseI.pdf\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbbox_inches\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'tight'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/figure.py\u001b[0m in \u001b[0;36msavefig\u001b[0;34m(self, fname, transparent, **kwargs)\u001b[0m\n\u001b[1;32m   3013\u001b[0m                 \u001b[0mpatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_edgecolor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'none'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3014\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3015\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcanvas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprint_figure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3016\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3017\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtransparent\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backend_bases.py\u001b[0m in \u001b[0;36mprint_figure\u001b[0;34m(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)\u001b[0m\n\u001b[1;32m   2253\u001b[0m                 \u001b[0;31m# force the figure dpi to 72), so we need to set it again here.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2254\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0mcbook\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setattr_cm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdpi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2255\u001b[0;31m                     result = print_method(\n\u001b[0m\u001b[1;32m   2256\u001b[0m                         \u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2257\u001b[0m                         \u001b[0mfacecolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfacecolor\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backend_bases.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m   1667\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1668\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1669\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1670\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1671\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mprint_tif\u001b[0;34m(self, filename_or_obj, pil_kwargs)\u001b[0m\n\u001b[1;32m    589\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0m_check_savefig_extra_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    590\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mprint_tif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilename_or_obj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpil_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 591\u001b[0;31m         \u001b[0mFigureCanvasAgg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdraw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    592\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpil_kwargs\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    593\u001b[0m             \u001b[0mpil_kwargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mdraw\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    399\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdraw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    400\u001b[0m         \u001b[0;31m# docstring inherited\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 401\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrenderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_renderer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcleared\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    402\u001b[0m         \u001b[0;31m# Acquire a lock on the shared font cache.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    403\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mRendererAgg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlock\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mget_renderer\u001b[0;34m(self, cleared)\u001b[0m\n\u001b[1;32m    415\u001b[0m                           and getattr(self, \"_lastKey\", None) == key)\n\u001b[1;32m    416\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mreuse_renderer\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 417\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrenderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRendererAgg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    418\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lastKey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    419\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mcleared\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, width, height, dpi)\u001b[0m\n\u001b[1;32m     89\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwidth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwidth\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mheight\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 91\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_renderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_RendererAgg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwidth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mheight\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     92\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_filter_renderers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Image size of 29681771x6515 pixels is too large. It must be less than 2^16 in each direction."
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Image size of 29681771x6515 pixels is too large. It must be less than 2^16 in each direction.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/IPython/core/formatters.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    339\u001b[0m                 \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    340\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 341\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mprinter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    342\u001b[0m             \u001b[0;31m# Finally look for special method names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    343\u001b[0m             \u001b[0mmethod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_real_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprint_method\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/IPython/core/pylabtools.py\u001b[0m in \u001b[0;36mprint_figure\u001b[0;34m(fig, fmt, bbox_inches, base64, **kwargs)\u001b[0m\n\u001b[1;32m    149\u001b[0m         \u001b[0mFigureCanvasBase\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    150\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 151\u001b[0;31m     \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcanvas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprint_figure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbytes_io\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    152\u001b[0m     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbytes_io\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetvalue\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mfmt\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'svg'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backend_bases.py\u001b[0m in \u001b[0;36mprint_figure\u001b[0;34m(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)\u001b[0m\n\u001b[1;32m   2253\u001b[0m                 \u001b[0;31m# force the figure dpi to 72), so we need to set it again here.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2254\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0mcbook\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setattr_cm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdpi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2255\u001b[0;31m                     result = print_method(\n\u001b[0m\u001b[1;32m   2256\u001b[0m                         \u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2257\u001b[0m                         \u001b[0mfacecolor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfacecolor\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backend_bases.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m   1667\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1668\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1669\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1670\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1671\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mprint_png\u001b[0;34m(self, filename_or_obj, metadata, pil_kwargs, *args)\u001b[0m\n\u001b[1;32m    506\u001b[0m             \u001b[0;34m*\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mincluding\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mdefault\u001b[0m \u001b[0;34m'Software'\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    507\u001b[0m         \"\"\"\n\u001b[0;32m--> 508\u001b[0;31m         \u001b[0mFigureCanvasAgg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdraw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    509\u001b[0m         mpl.image.imsave(\n\u001b[1;32m    510\u001b[0m             \u001b[0mfilename_or_obj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuffer_rgba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"png\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morigin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"upper\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mdraw\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    399\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdraw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    400\u001b[0m         \u001b[0;31m# docstring inherited\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 401\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrenderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_renderer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcleared\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    402\u001b[0m         \u001b[0;31m# Acquire a lock on the shared font cache.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    403\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mRendererAgg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlock\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36mget_renderer\u001b[0;34m(self, cleared)\u001b[0m\n\u001b[1;32m    415\u001b[0m                           and getattr(self, \"_lastKey\", None) == key)\n\u001b[1;32m    416\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mreuse_renderer\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 417\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrenderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRendererAgg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    418\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lastKey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    419\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mcleared\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/gene_feature_plot/lib/python3.9/site-packages/matplotlib/backends/backend_agg.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, width, height, dpi)\u001b[0m\n\u001b[1;32m     89\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwidth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwidth\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mheight\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 91\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_renderer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_RendererAgg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwidth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mheight\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     92\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_filter_renderers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Image size of 29681771x6515 pixels is too large. It must be less than 2^16 in each direction."
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 6600x4200 with 36 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    import sys\n",
    "\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2edcec37c9f2626182329974fec9906c797ef12ccf33c4b458ba0845e5da6c77"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('gene_feature_plot': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
